---
title: "Methylation_R_scripts"
output: html_document
date: "2025-03-12"
---

```{r setup, include=FALSE, message=FALSE, warning=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

# Libraries
```{r, message=FALSE, warning=FALSE}
library(ggplot2)
library(dplyr)
library(data.table)
library(tidyr)
library(phyloseq)
library(stringr)
library(eulerr)
library(vegan)
```

# Colors
```{r}
cols <- c(
"NA" = "#FF0000", 
"Citrobacter_B koseri" = "#3cb44b",
"Pseudomonas_E putida" = "#ffe119",
"Bordatella avium" = "#4363d8",
"Microvirga lotononidis" = "#000000",
"Paracoccus denitrificans" = "#911eb4",
"Comamonas testosteroni_C" = "#46f0f0",
"Pseudomonas_E chlororaphis" = "#f032e6",
"Morganella morganii" = "#bcf60c",
"Cupriavidus oxalaticus" = "#fabebe",
"Sphingobium yanoikuyae" = "#008080",
"Aeromonas caviae" = "#e6beff",
"Agrobacterium tumefaciens" = "#9a6324",
"Stenotrophomonas maltophilia" = "#fffac8",
"Sphingobacterium spiritivorum" = "#800000",
"Kluyvera intermedia" = "#aaffc3")
```

# Sylph synthetic community
```{r}
# Load data
Sylph_HAMBI_merged <- read.delim("~/Desktop/Git/Methylation/R/Methylation_R/RFiles/Sylph_HAMBI_merged_genus_full.txt", row.names=1)

# Tax table
tax_table0 <- data.frame("rownames" = rownames(Sylph_HAMBI_merged))
tax_table <- separate(tax_table0, rownames, 
                       into = c("domain", "phylum", "class", "order", "family", "genus", "species", "strain"), sep = "\\|",
                       fill = "right", remove = TRUE)
tax_table$domain <- gsub("d__", "", tax_table$domain)
tax_table$phylum <- gsub("p__", "", tax_table$phylum)
tax_table$class <- gsub("c__", "", tax_table$class)
tax_table$order <- gsub("o__", "", tax_table$order)
tax_table$family <- gsub("f__", "", tax_table$family)
tax_table$genus <- gsub("g__", "", tax_table$genus)
tax_table$species <- gsub("s__", "", tax_table$species)
tax_table$strain <- gsub("t__", "", tax_table$strain)

# Otu table
OTU_table = Sylph_HAMBI_merged
rownames(OTU_table) <- 1:nrow(Sylph_HAMBI_merged)
rownames(OTU_table) <- paste0("sp", rownames(OTU_table))

# Sample data
metadata <- data.frame("sample" = colnames(OTU_table))
rownames(metadata) <- colnames(OTU_table)
meta <- sample_data(metadata)

# Phyloseq
OTU = otu_table(OTU_table, taxa_are_rows = TRUE)
tax_table_matrix <- as.matrix(tax_table)
TAX = tax_table(tax_table_matrix)

PHY <- phyloseq(OTU, TAX, meta)

PHY_genus <- tax_glom(PHY, taxrank = "genus")
```

# Plot
```{r}
cols <- c("NA" = "#FF0000", "Citrobacter_B" = "#3cb44b", "Pseudomonas_E" = "#f032e6", "Bordatella" = "#4363d8", "Microvirga" = "#000000", "Paracoccus" = "#911eb4", "Comamonas" = "#46f0f0", "Morganella" = "#bcf60c", "Cupriavidus" = "#fabebe", "Sphingobium" = "#008080", "Aeromonas" = "#e6beff", "Agrobacterium" = "#9a6324", "Stenotrophomonas" = "#fffac8", "Sphingobacterium" = "#800000", "Kluyvera" = "#aaffc3", "Acinetobacter" = "navyblue", "Azorhizobium" = "#3294c1", "Azospirillum" = "#D35400", "Chromobacterium" = "grey", "Elizabethkingia" = "#A3E4D7", "Enterococcus" = "orange", "Escherichia" = "#cfb189", "Hafnia" = "#29522F", "Listeria" = "#D1D19B", "Phyllobacteria" = "#74A172", "Staphylococcus" = "#F56782")

# Get total relative abundance for each taxon across all samples
PHY_genus_total_abund <- taxa_sums(PHY_genus)

# Filter
#PHY_genus_filtered <- prune_taxa(PHY_genus_total_abund > 0.1, PHY_genus)
PHY_genus_filtered = PHY_genus

# Function to remove border color
plot_bar_2 <-  function (PHY_genus_filtered, x = "sample", y = "Abundance", fill = NULL, title = NULL, facet_grid = NULL, border_color = NA) 
{
  mdf = psmelt(PHY_genus_filtered)
  p = ggplot(mdf, aes_string(x = x, y = y, fill = fill))
  p = p + geom_bar(stat = "identity", position = "stack",  color = border_color)
  p = p + theme(axis.text.x = element_text(angle = -90, hjust = 0))
  if (!is.null(facet_grid)) {
    p <- p + facet_grid(facet_grid)
  }
  if (!is.null(title)) {
    p <- p + ggtitle(title)
  }
  return(p)
}

p <- plot_bar_2(PHY_genus_filtered, fill = "genus") + 
  geom_bar(stat = "identity") + theme_minimal() + 
  labs(title = "Bacterial composition (genus)",
       subtitle = "Synthetic community",
       y = "Relative abundance") +
  scale_y_continuous(expand = c(0,0)) +
  scale_x_discrete(labels = c(
    "bcAd1023T" = "bcAd1023\n0.2 µg/ml TC, 15 °C",
    "bcAd1037T" = "bcAd1037\n20 µg/ml SMZ, 25 °C",
    "bcAd1039T" = "bcAd1039\n20 µg/ml SMZ, 25 °C",
    "bcAd1046T" = "bcAd1046\n0.02 µg/ml TC, 37 °C",
    "bcAd1063T" = "bcAd1063\nno ab, 37 °C")) +
  theme(
        text = element_text(family = "Times"),
        legend.text = element_text(face = "italic", size = 14, margin = margin(l = 7)),
        legend.title = element_blank(),
        axis.title.x = element_blank(),
        axis.title.y = element_text(size = 14, face = "bold"),
        axis.text.y = element_text(size = 12),
        axis.text.x = element_text(size = 12, hjust = 1, vjust = 1, angle = 35, face = "bold"),
        plot.title = element_text(size = 20),
        plot.subtitle = element_text(size = 14)) + 
  scale_fill_manual(values=c(cols))
p

ggsave(filename = "plots/HAMBI_sylph.png",
       width = 8, height = 6, dpi = 300, units = "in", device="png", scale = 1)

ggsave(filename = "plots/HAMBI_sylph.svg", 
       width = 8, height = 6, units = "in")
```

# Sylph wastewater community
```{r}
# Load data
Sylph_HiFi_merged <- read.delim("~/Desktop/Git/Methylation/R/Methylation_R/RFiles/Sylph_HiFi_merged_genus_full.txt", row.names=1)

# Tax table
tax_table0 <- data.frame("rownames" = rownames(Sylph_HiFi_merged))
tax_table <- separate(tax_table0, rownames, 
                       into = c("domain", "phylum", "class", "order", "family", "genus", "species", "strain"),
                      sep = "\\|", fill = "right", remove = TRUE)
tax_table$domain <- gsub("d__", "", tax_table$domain)
tax_table$phylum <- gsub("p__", "", tax_table$phylum)
tax_table$class <- gsub("c__", "", tax_table$class)
tax_table$order <- gsub("o__", "", tax_table$order)
tax_table$family <- gsub("f__", "", tax_table$family)
tax_table$genus <- gsub("g__", "", tax_table$genus)
tax_table$species <- gsub("s__", "", tax_table$species)
tax_table$strain <- gsub("t__", "", tax_table$strain)

# Otu table
OTU_table = Sylph_HiFi_merged
rownames(OTU_table) <- 1:nrow(Sylph_HiFi_merged)
rownames(OTU_table) <- paste0("sp", rownames(OTU_table))

# Sample data
metadata <- data.frame("sample_type" = rep(c("effluent", "influent", "sludge"), each = 3),
                       "sample" = colnames(OTU_table))
rownames(metadata) <- colnames(OTU_table)
meta <- sample_data(metadata)

# Phyloseq
OTU = otu_table(OTU_table, taxa_are_rows = TRUE)
tax_table_matrix <- as.matrix(tax_table)
TAX = tax_table(tax_table_matrix)

PHY <- phyloseq(OTU, TAX, meta)

PHY_genus <- tax_glom(PHY, taxrank = "genus")
```

# Plot
```{r}
cols <- c("#F5F5DC", "#34495E", "#5d7b65", "#c9f7d1", "#D35400", "#F1C40F", "#a40500", "#c9f063", "#06CC97",
          "#E6E3E3", "#cfb189", "#3294c1", "#c1a3c0", "#7FDBFF", "#FF6B6B", "black", "navyblue",
          "#808000","#1ABC9C", "#584691", "#1cb9be", "#f6b178", "#1f8d28", "#092fdb", "#fed7d1", "orange", "coral",
          "#fe7fca", "#6698f5", "#FFB6B0", "#6f18c1", "#49c118", "grey", "#faec81",
          "#35f2ef")

# Get total relative abundance for each taxon across all samples
PHY_genus_total_abund <- taxa_sums(PHY_genus)

PHY_genus_filtered <- prune_taxa(PHY_genus_total_abund > 0.5, PHY_genus)
# What does it mean? 1216 genera are filtered out and we are left with 14.966 % of the genus diversity.
# 1430 - 1216 = 214
# Still the sample sums are (so only rare genera were dropped):
#    EFF1    EFF2    EFF3    INF1    INF2    INF3    SLU1    SLU2    SLU3 
# 86.8107 86.7874 90.4764 88.3669 88.0146 86.6902 90.3468 90.3934 91.7258 

#TopNOTUs <- names(sort(taxa_sums(PHY_genus), TRUE)[1:20])
#PHY_filtered <- prune_species(TopNOTUs, PHY_genus)

# Reorder sample_type with custom levels
sample_data(PHY_genus_filtered)$sample_type <- factor(sample_data(PHY_genus_filtered)$sample_type,
                                       levels = c("influent", "effluent", "sludge"))

# Function to remove border color
plot_bar_2 <-  function (PHY_genus_filtered, x = "sample", y = "Abundance", fill = NULL, title = NULL, facet_grid = NULL, border_color = NA) 
{
  mdf = psmelt(PHY_genus_filtered)
  p = ggplot(mdf, aes_string(x = x, y = y, fill = fill))
  p = p + geom_bar(stat = "identity", position = "stack",  color = border_color)
  p = p + theme(axis.text.x = element_text(angle = -90, hjust = 0))
  if (!is.null(facet_grid)) {
    p <- p + facet_grid(facet_grid)
  }
  if (!is.null(title)) {
    p <- p + ggtitle(title)
  }
  return(p)
}

p <- plot_bar_2(PHY_genus_filtered, fill = "phylum") + 
  geom_bar(stat = "identity") + theme_minimal() + 
  labs(title = "Bacterial composition (phylum)",
       subtitle = "Wastewater community (genera > 0.5 %)",
       y = "Relative abundance") +
  scale_y_continuous(expand = c(0,0), breaks = seq(0, 100, by = 25)) +
  theme(
        text = element_text(family = "Times"),
        legend.text = element_text(face = "italic", size = 14, margin = margin(l = 7)),
        legend.title = element_blank(),
        axis.title.x = element_blank(),
        axis.title.y = element_text(size = 14, face = "bold"),
        axis.text.y = element_text(size = 12),
        axis.text.x = element_text(size = 12, face = "bold"),
        plot.title = element_text(size = 20),
        plot.subtitle = element_text(size = 14)) + 
  scale_fill_manual(values=c(cols)) +
  facet_grid(~sample_type, scales = "free_x", space = "free_x", labeller = labeller(sample_type = function(x) rep("", length(x))))
p

#ggsave(filename = "plots/WW_sylph_>0.5_percent.png",
#       width = 10, height = 6, dpi = 300, units = "in", device="png", scale = 1)

#ggsave(filename = "plots/WW_sylph_>0.5_percent.svg", 
#       width = 10, height = 6, units = "in")
```

# Sylph wastewater community
## Bacteroidales
```{r}
cols <- c("#F5F5DC", "#34495E", "#5d7b65", "#c9f7d1", "#D35400", "#a40500", "#c9f063", "#E74C3C",
          "#A3E4D7", "#cfb189", "#3294c1", "#c1a3c0", "#7FDBFF", "#FF6B6B", "black",
          "#808000","#1ABC9C", "#584691", "#1cb9be", "#1f8d28", "#092fdb", "#fed7d1",
          "#fe7fca", "#6698f5", "#FFB6B0", "#F1C40F", "#542C42", "#6f18c1", "#49c118", "grey", "#faec81", "#35f2ef")

PHY_Bact <- subset_taxa(PHY_genus, order == "Bacteroidales")

#PHY_Bact_rel_total_abund <- taxa_sums(PHY_Bact_rel)
#PHY_Bact_filtered <- prune_taxa(PHY_Bact_rel_total_abund > 0.01, PHY_Bact_rel)

# Reorder sample_type with custom levels
sample_data(PHY_Bact)$sample_type <- factor(sample_data(PHY_Bact)$sample_type,
                                       levels = c("influent", "effluent", "sludge"))

# Function to remove border color
plot_bar_2 <-  function (PHY_Bact, x = "sample", y = "Abundance", fill = NULL, title = NULL, facet_grid = NULL, border_color = NA) 
{
  mdf = psmelt(PHY_Bact)
  p = ggplot(mdf, aes_string(x = x, y = y, fill = fill))
  p = p + geom_bar(stat = "identity", position = "stack",  color = border_color)
  p = p + theme(axis.text.x = element_text(angle = -90, hjust = 0))
  if (!is.null(facet_grid)) {
    p <- p + facet_grid(facet_grid)
  }
  if (!is.null(title)) {
    p <- p + ggtitle(title)
  }
  return(p)
}

p <- plot_bar_2(PHY_Bact, fill = "family") + 
  geom_bar(stat = "identity") + theme_minimal() + 
  labs(title = "Bacterial composition (family)",
       subtitle = "Wastewater community (order: Bacteroidales)",
       y = "Relative abundance") +
  scale_y_continuous(expand = c(0,0)) +
    theme(
        text = element_text(family = "Times"),
        legend.text = element_text(face = "italic", size = 14, margin = margin(l = 7)),
        legend.title = element_blank(),
        axis.title.x = element_blank(),
        axis.title.y = element_text(size = 14, face = "bold"),
        axis.text.y = element_text(size = 12),
        axis.text.x = element_text(size = 12, face = "bold"),
        plot.title = element_text(size = 20),
        plot.subtitle = element_text(size = 14)) + 
  scale_fill_manual(values=c(cols)) + 
  facet_grid(~sample_type, scales = "free_x", space = "free_x", labeller = labeller(sample_type = function(x) rep("", length(x))))
p

ggsave(filename = "plots/WW_sylph_Bacteroidales_family.png",
       width = 10, height = 6, dpi = 300, units = "in", device="png", scale = 1)

ggsave(filename = "plots/WW_sylph_Bacteroidales_family.svg", 
       width = 10, height = 6, units = "in")
```

# Sylph wastewater community
## Acinetobacter
```{r}
cols <- c("#548058")

PHY_Bact <- subset_taxa(PHY_genus, genus == "Acinetobacter")

#PHY_Bact_rel_total_abund <- taxa_sums(PHY_Bact_rel)
#PHY_Bact_filtered <- prune_taxa(PHY_Bact_rel_total_abund > 0.01, PHY_Bact_rel)

# Reorder sample_type with custom levels
sample_data(PHY_Bact)$sample_type <- factor(sample_data(PHY_Bact)$sample_type,
                                       levels = c("influent", "effluent", "sludge"))

# Function to remove border color
plot_bar_2 <-  function (PHY_Bact, x = "sample", y = "Abundance", fill = NULL, title = NULL, facet_grid = NULL, border_color = NA) 
{
  mdf = psmelt(PHY_Bact)
  p = ggplot(mdf, aes_string(x = x, y = y, fill = fill))
  p = p + geom_bar(stat = "identity", position = "stack",  color = border_color)
  p = p + theme(axis.text.x = element_text(angle = -90, hjust = 0))
  if (!is.null(facet_grid)) {
    p <- p + facet_grid(facet_grid)
  }
  if (!is.null(title)) {
    p <- p + ggtitle(title)
  }
  return(p)
}

p <- plot_bar_2(PHY_Bact, fill = "genus") + 
  geom_bar(stat = "identity") + theme_minimal() + 
  labs(title = "Bacterial composition (genus)",
       subtitle = expression("Wastewater community (genus: " * italic("Acinetobacter")*")"),
       y = "Relative abundance") +
  scale_y_continuous(expand = c(0,0)) +
    theme(
        text = element_text(family = "Times"),
        legend.position = "none",
        axis.title.x = element_blank(),
        axis.title.y = element_text(size = 14, face = "bold"),
        axis.text.y = element_text(size = 12),
        axis.text.x = element_text(size = 12, face = "bold"),
        plot.title = element_text(size = 20),
        plot.subtitle = element_text(size = 14)) + 
  scale_fill_manual(values=c(cols)) + 
  facet_grid(~sample_type, scales = "free_x", space = "free_x", labeller = labeller(sample_type = function(x) rep("", length(x))))
p

ggsave(filename = "plots/WW_sylph_Acinetobacter.png",
       width = 10, height = 6, dpi = 300, units = "in", device="png", scale = 1)

ggsave(filename = "plots/WW_sylph_Acinetobacter.svg", 
       width = 10, height = 6, units = "in")
```

# Ordination on bacterial composition
```{r}
# Perform PCoA
ord_PHY_genus <- ordinate(PHY_genus, method = "PCoA", distance = "bray")

cols <- c("#17b2c2", "#47ecfd", "#29b3fc", "#f9e30a", "#fcf0a0", "#eac51e","#7c4107","#c89b6d", "#cc7442")

# Plot the PCoA
p <- plot_ordination(PHY_genus, ord_PHY_genus, color = "sample") + 
  geom_point(size = 4) +
  theme_minimal() +
  labs(title = "PCoA of Taxonomic Diversity (genus)",
       subtitle = "Bray-Curtis") +
  theme(legend.title = element_blank(),
        legend.text = element_text(size = 14),
        text = element_text(family = "Times"),
        plot.title = element_text(size = 20),
        plot.subtitle = element_text(size = 14),
        axis.text = element_text(size = 14),
        axis.title = element_text(size = 14, face = "bold")) +
        scale_color_manual(values = c(cols)) + coord_fixed()
p

ggsave(filename = "plots/WW_sylph_ord.png",
       width = 10, height = 6, dpi = 300, units = "in", device="png", scale = 1)

ggsave(filename = "plots/WW_sylph_ord.svg", 
       width = 10, height = 6, units = "in")
```

## PERMANOVA
```{r}
# Calculate Bray-Curtis distance matrix
bray_dist <- phyloseq::distance(PHY_genus, method = "bray")

# Extract sample metadata
metadata <- as(sample_data(PHY_genus), "data.frame")

# Run PERMANOVA
permanova_result <- adonis2(bray_dist ~ sample_type, data = metadata)

# View results
print(permanova_result)
```

# HiFi MAG Pipeline results
```{r}
# Load data
hifi_MAG_pipeline_results <- read.delim("~/Desktop/Git/Methylation/R/Methylation_R/RFiles/hifi_MAG_pipeline_results.txt", row.names=1)

# List High-quality MAGs
HQ_MAGs <- hifi_MAG_pipeline_results %>%
  filter(Completeness >= 90 & Contamination <= 5)

# Filter those that have ARGs
HQ_MAGs_with_ARGs <- HQ_MAGs %>%
  filter(ARG != "NA")

# Count HQ MAGs/sample
HQ_MAG_count_sample <- HQ_MAGs %>% count(Sample)

# Count HQ MAGs/taxa
HQ_MAG_count_species <- HQ_MAGs %>% count(species)
HQ_MAG_count_genus <- HQ_MAGs %>% count(genus)
HQ_MAG_count_class <- HQ_MAGs %>% count(class)
HQ_MAG_count_phylum <- HQ_MAGs %>% count(phylum)

# Count HQ MAGs/taxa/Sample
HQ_MAG_count_species_sample <- HQ_MAGs %>% count(species, Sample)
HQ_MAG_count_genus_sample <- HQ_MAGs %>% count(genus, Sample)
HQ_MAG_count_class_sample <- HQ_MAGs %>% count(class, Sample)
HQ_MAG_count_phylum_sample <- HQ_MAGs %>% count(phylum, Sample)

# Get statistics
hifi_MAG_pipeline_stats <- hifi_MAG_pipeline_results %>%
  summarize(mean_contigs = mean(Contig_Number),
            mean_completeness = mean(Completeness),
            mean_contamination = mean(Contamination),
            max_completeness = max(Completeness),
            min_completeness = min(Completeness),
            max_contamination = max(Contamination),
            min_contamination = min(Contamination))

# Filter those that have ARGs (all MAGs)
MAGs_with_ARGs <- hifi_MAG_pipeline_results %>%
  filter(ARG != "NA")

# MAGs with any ARGs
MAGs_with_ARGs <- hifi_MAG_pipeline_results %>%
  filter(ARG != "NA")

# erm(F) hosts
MAGs_with_erm <- hifi_MAG_pipeline_results %>%
  filter(grepl("erm", ARG))

# Count MAGs/sample
MAG_count_sample <- hifi_MAG_pipeline_results %>% count(Sample)

# Count MAGs/taxa
MAG_count_species <- hifi_MAG_pipeline_results %>% count(species)
MAG_count_genus <- hifi_MAG_pipeline_results %>% count(genus)
MAG_count_class <- hifi_MAG_pipeline_results %>% count(class)
MAG_count_phylum <- hifi_MAG_pipeline_results %>% count(phylum)

# Count MAGs/taxa/Sample
MAG_count_species_sample <- hifi_MAG_pipeline_results %>% count(species, Sample)
MAG_count_genus_sample <- hifi_MAG_pipeline_results %>% count(genus, Sample)
MAG_count_class_sample <- hifi_MAG_pipeline_results %>% count(class, Sample)
MAG_count_phylum_sample <- hifi_MAG_pipeline_results %>% count(phylum, Sample)
```

# Plot HiFi MAG Pipeline results
## All
##### huom, luvut pitää mahd päivittää
```{r}
cols <- c("#17b2c2", "#47ecfd", "#29b3fc", "#f9e30a", "#fcf0a0", "#eac51e","#7c4107","#c89b6d", "#cc7442")

# MAGs/taxa/Sample
## Class
p <- ggplot(MAG_count_class_sample, aes(x = class, y = n, fill = Sample)) + 
  geom_bar(stat = "identity") + scale_fill_manual(values = c(cols)) + theme_minimal() +
  labs(y = "Number of MAGs", title = "MAGs by HiFi MAG Pipeline (class)") +
  scale_y_continuous(expand = c(0,0), limits = c(0,67)) +
  theme(
        text = element_text(family = "Times"),
        legend.text = element_text(size = 14),
        legend.title = element_blank(),
        axis.title.x = element_blank(),
        axis.title.y = element_text(size = 14, face = "bold"),
        axis.text.y = element_text(size = 14),
        axis.text.x = element_text(size = 13, angle = 60, hjust = 1.009),
        plot.title = element_text(size = 20))
p

ggsave(filename = "plots/HiFi_MAGs_class.png",
       width = 10, height = 6, dpi = 300, units = "in", device="png", scale = 1)

ggsave(filename = "plots/HiFi_MAGs_class.svg", 
       width = 10, height = 6, units = "in")


# Other taxa levels

## Phylum
p <- ggplot(MAG_count_phylum_sample, aes(x = phylum, y = n, fill = Sample)) + 
  geom_bar(stat = "identity") + scale_fill_manual(values = c(cols)) + theme_minimal() +
  ggtitle("MAGs by HiFi MAG Pipeline (phylum)") +
  scale_y_continuous(expand = c(0,0)
                     #, limits = c(0,50)
                     ) +
  theme(
        text = element_text(family = "Palatino"),
        legend.text = element_text(size = 20),
        legend.title = element_blank(),
        axis.title = element_blank(),
        axis.text.y = element_text(size = 30),
        axis.text.x = element_text(size = 16, angle = 65, hjust = 0.95),
        plot.title = element_text(size = 30))
p

## Genus
p <- ggplot(MAG_count_genus_sample, aes(x = genus, y = n, fill = Sample)) + 
  geom_bar(stat = "identity") + scale_fill_manual(values = c(cols)) + theme_minimal() +
  ggtitle("MAGs by HiFi MAG Pipeline (genus)") +
  scale_y_continuous(expand = c(0,0)
                     #, limits = c(0,45)
                     ) +
  theme(
        text = element_text(family = "Palatino"),
        legend.text = element_text(size = 20),
        legend.title = element_blank(),
        axis.title = element_blank(),
        axis.text.y = element_text(size = 20),
        axis.text.x = element_text(size = 8, angle = 90, face = "italic", vjust = 0.5, hjust=1),
        plot.title = element_text(size = 20))
p

# MAGs/taxa/Sample
## Species
p <- ggplot(MAG_count_species_sample, aes(x = species, y = n, fill = Sample)) + 
  geom_bar(stat = "identity") + scale_fill_manual(values = c(cols)) + theme_minimal() +
  ggtitle("MAGs by HiFi MAG Pipeline (species)") +
  scale_y_continuous(expand = c(0,0)
                     #, limits = c(0,50)
                     ) +
  theme(
        text = element_text(family = "Palatino"),
        legend.text = element_text(size = 20),
        legend.title = element_blank(),
        axis.title = element_blank(),
        axis.text.y = element_text(size = 20),
        axis.text.x = element_text(size = 8, angle = 90, face = "italic", vjust = 0.5, hjust=1),
        plot.title = element_text(size = 20))
p
```

# Summary of HiFi MAG Pipeline results
###### muokkaa tätä kokonaisuudessaan, tsekkaa luvut  #########
```{r}
cols <- c("EFF1" = "#17b2c2", "EFF2" = "#47ecfd", "EFF3" = "#29b3fc",
          "INF1" = "#f9e30a", "INF2" = "#fcf0a0", "INF3" = "#eac51e",
          "SLU1" = "#e38236", "SLU2" = "#fec191", "SLU3" = "#daa187")

# Add ARGs
counts_MAGs_with_ARGs <- MAGs_with_ARGs %>% count(Sample)
View(counts_MAGs_with_ARGs)
colnames(counts_MAGs_with_ARGs) <- c("Sample", "ARGs")

# Match
x <- match(MAG_count_sample$Sample, counts_MAGs_with_ARGs$Sample)
counts_MAGs_with_ARGs <- counts_MAGs_with_ARGs[x,]

# Bind
df <- cbind(MAG_count_sample, "ARGs" = counts_MAGs_with_ARGs$ARGs)
df[is.na(df)] <- ""

# reorder
df$Sample <- factor(df$Sample, levels = c("INF1", "INF2", "INF3", "EFF1", "EFF2", "EFF3", "SLU1", "SLU2", "SLU3"))

# Plot
p <- ggplot(df, aes(x = Sample, y = n, fill = Sample, label = ARGs)) + 
  geom_bar(stat = "identity", alpha = 0.75) + theme_minimal() +
  geom_text(size = 10, position = position_stack(vjust = .5), color = "black", family = "Times") +
  labs(title = "Number of MAGs by HiFi MAG Pipeline", y = "Number of MAGs") + theme_minimal() +
  scale_y_continuous(expand = c(0,0)) +
  theme(
        text = element_text(family = "Times"),
        legend.text = element_text(size = 14),
        legend.title = element_blank(),
        axis.title.x = element_blank(),
        axis.title.y = element_text(size = 14, face = "bold"),
        axis.text.y = element_text(size = 14),
        axis.text.x = element_text(size = 14, face = "bold"),
        plot.title = element_text(size = 20)) + scale_fill_manual(values = c(cols))
p

ggsave(filename = "plots/Pipeline_MAGs.png",
       width = 10, height = 6, dpi = 300, units = "in", device="png", scale = 1)

ggsave(filename = "plots/Pipeline_MAGs.svg", 
       width = 10, height = 6, units = "in")
```

# HiFi MAG Pipeline results vs. methylation-based MAGs
```{r}
ARG_MAGs_counts <- read.delim("~/Desktop/Git/Methylation/R/Methylation_R/RFiles/ARG_MAG_counts.txt")
ARG_MAGs_counts$Methylation_based_analysis <- as.numeric(ARG_MAGs_counts$Methylation_based_analysis)
ARG_MAGs_counts$HiFi_MAG_Pipeline <- as.numeric(ARG_MAGs_counts$HiFi_MAG_Pipeline)

ARG_MAGs_counts_long <- melt(setDT(ARG_MAGs_counts), id.vars = c("sample"), variable.name = "Methylation_based_analysis")
ARG_MAGs_counts_long$sample_type <- ARG_MAGs_counts_long$sample
ARG_MAGs_counts_long$sample_type <- gsub("[0-9]", "", ARG_MAGs_counts_long$sample_type)
ARG_MAGs_counts_long$sample_type <- gsub("INF", "influent", ARG_MAGs_counts_long$sample_type)
ARG_MAGs_counts_long$sample_type <- gsub("EFF", "effluent", ARG_MAGs_counts_long$sample_type)
ARG_MAGs_counts_long$sample_type <- gsub("SLU", "dried sludge", ARG_MAGs_counts_long$sample_type)

colnames(ARG_MAGs_counts_long) <- c("sample", "Method", "count", "sample_type")

df <- ARG_MAGs_counts_long[1:18,]
df$count <- as.integer(df$count)

df$sample <- factor(df$sample, levels=c("INF1", "INF2", "INF3",
                                          "EFF1", "EFF2", "EFF3",
                                          "SLU1", "SLU2", "SLU3"))

p <- ggplot(df, aes(x = sample, y = count, fill = interaction(sample, Method))) + 
  geom_bar(stat = "identity", position = "dodge") + theme_bw() +
  scale_y_continuous(labels = scales::number_format(accuracy = 1), expand = c(0,0), limits = c(0,10.75)) +
  scale_fill_manual(values = c(
        EFF1.Methylation_based_analysis = "#29b3fc", 
        EFF2.Methylation_based_analysis = "#29b3fc", 
        EFF3.Methylation_based_analysis = "#29b3fc",
        EFF1.HiFi_MAG_Pipeline = "#b7e0f6", EFF2.HiFi_MAG_Pipeline = "#b7e0f6", EFF3.HiFi_MAG_Pipeline = "#b7e0f6",
        INF1.Methylation_based_analysis = "#f9e30a",
        INF2.Methylation_based_analysis = "#f9e30a",
        INF3.Methylation_based_analysis = "#f9e30a",
        INF1.HiFi_MAG_Pipeline = "#f9f5de", INF2.HiFi_MAG_Pipeline = "#f9f5de", INF3.HiFi_MAG_Pipeline = "#f9f5de",
        SLU1.Methylation_based_analysis = "#daa187",
        SLU2.Methylation_based_analysis = "#daa187",
        SLU3.Methylation_based_analysis = "#daa187",
        SLU1.HiFi_MAG_Pipeline = "#ffdcc0", SLU2.HiFi_MAG_Pipeline = "#ffdcc0", SLU3.HiFi_MAG_Pipeline = "#ffdcc0")) +
  labs(title = "Number of MAGs with ARGs", y = "Number of MAGs") + theme_minimal() +
  theme(text = element_text(family = "Times"),
        legend.position = "none",
        strip.text = element_text(size = 16, face = "bold"),
    strip.placement = "outside",  # Puts strip labels outside panels
    strip.placement.x = "top",    # Ensures strips are at the top
    strip.text.x = element_text(margin = margin(b = 10)),  # Add space below strip label
    panel.spacing = unit(1.2, "lines")   ,
        axis.title = element_blank(),
        axis.text.y = element_text(size = 16),
        axis.text.x = element_text(size = 16),
        #axis.text.x = element_text(size = 24, angle = 90, hjust = 1, color = "white"),
        plot.title = element_text(size = 30)) + 
  facet_grid(~factor(sample_type, levels=c("influent", "effluent", "dried sludge")),
             space = "free", scales = "free", switch = "x")
p

ggsave(filename = "plots/ARG_MAGs_bar_plot.png",
       width = 10, height = 6, dpi = 300, units = "in", device="png", scale = 1)

ggsave(filename = "plots/ARG_MAGs_bar_plot.svg",
       width = 10, height = 6, dpi = 300, units = "in", device="svg", scale = 1)
```


# HiFi MAG Pipeline results vs. methylation-based MAGs with fARGene results
```{r}
ARG_MAGs_counts <- read.delim("~/Desktop/Git/Methylation/R/Methylation_R/RFiles/fARG_MAG_counts.txt")
ARG_MAGs_counts$Methylation_based_analysis <- as.numeric(ARG_MAGs_counts$Methylation_based_analysis)
ARG_MAGs_counts$HiFi_MAG_Pipeline <- as.numeric(ARG_MAGs_counts$HiFi_MAG_Pipeline)

ARG_MAGs_counts_long <- melt(setDT(ARG_MAGs_counts), id.vars = c("sample"), variable.name = "Methylation_based_analysis")
ARG_MAGs_counts_long$sample_type <- ARG_MAGs_counts_long$sample
ARG_MAGs_counts_long$sample_type <- gsub("[0-9]", "", ARG_MAGs_counts_long$sample_type)
ARG_MAGs_counts_long$sample_type <- gsub("INF", "influent", ARG_MAGs_counts_long$sample_type)
ARG_MAGs_counts_long$sample_type <- gsub("EFF", "effluent", ARG_MAGs_counts_long$sample_type)
ARG_MAGs_counts_long$sample_type <- gsub("SLU", "dried sludge", ARG_MAGs_counts_long$sample_type)

colnames(ARG_MAGs_counts_long) <- c("sample", "Method", "count", "sample_type")

df <- ARG_MAGs_counts_long[1:18,]
df$count <- as.integer(df$count)

df$sample <- factor(df$sample, levels=c("INF1", "INF2", "INF3",
                                          "EFF1", "EFF2", "EFF3",
                                          "SLU1", "SLU2", "SLU3"))

p <- ggplot(df, aes(x = sample, y = count, fill = interaction(sample, Method))) + 
  geom_bar(stat = "identity", position = "dodge") + theme_bw() +
  scale_y_continuous(labels = scales::number_format(accuracy = 1), expand = c(0,0), limits = c(0,10.75)) +
  scale_fill_manual(values = c(
        EFF1.Methylation_based_analysis = "#29b3fc", 
        EFF2.Methylation_based_analysis = "#29b3fc", 
        EFF3.Methylation_based_analysis = "#29b3fc",
        EFF1.HiFi_MAG_Pipeline = "#b7e0f6", EFF2.HiFi_MAG_Pipeline = "#b7e0f6", EFF3.HiFi_MAG_Pipeline = "#b7e0f6",
        INF1.Methylation_based_analysis = "#f9e30a",
        INF2.Methylation_based_analysis = "#f9e30a",
        INF3.Methylation_based_analysis = "#f9e30a",
        INF1.HiFi_MAG_Pipeline = "#f9f5de", INF2.HiFi_MAG_Pipeline = "#f9f5de", INF3.HiFi_MAG_Pipeline = "#f9f5de",
        SLU1.Methylation_based_analysis = "#daa187",
        SLU2.Methylation_based_analysis = "#daa187",
        SLU3.Methylation_based_analysis = "#daa187",
        SLU1.HiFi_MAG_Pipeline = "#ffdcc0", SLU2.HiFi_MAG_Pipeline = "#ffdcc0", SLU3.HiFi_MAG_Pipeline = "#ffdcc0")) +
  labs(title = "Number of MAGs with Latent ARGs (fARGene)", y = "Number of MAGs") + theme_minimal() +
  theme(text = element_text(family = "Times"),
        legend.position = "none",
        strip.text = element_text(size = 16, face = "bold"),
    strip.placement = "outside",  # Puts strip labels outside panels
    strip.placement.x = "top",    # Ensures strips are at the top
    strip.text.x = element_text(margin = margin(b = 10)),  # Add space below strip label
    panel.spacing = unit(1.2, "lines")   ,
        axis.title = element_blank(),
        axis.text.y = element_text(size = 16),
        axis.text.x = element_text(size = 16),
        #axis.text.x = element_text(size = 24, angle = 90, hjust = 1, color = "white"),
        plot.title = element_text(size = 30)) + 
  facet_grid(~factor(sample_type, levels=c("influent", "effluent", "dried sludge")),
             space = "free", scales = "free", switch = "x")
p

ggsave(filename = "plots/fARG_MAGs_bar_plot.png",
       width = 10, height = 6, dpi = 300, units = "in", device="png", scale = 1)

ggsave(filename = "plots/fARG_MAGs_bar_plot.svg",
       width = 10, height = 6, dpi = 300, units = "in", device="svg", scale = 1)
```




# Convert to long format for UMAP
## EFF1
```{r}
EFF1_HiFi_MAGs_contigs <- read.table("~/Desktop/Git/Methylation/R/Methylation_R/RFiles/EFF1_HiFi_MAGs_contigs.txt", sep = "\t", header = TRUE, stringsAsFactors = FALSE)

# Convert to long format
EFF1_HiFi_MAGs_contigs_long <- EFF1_HiFi_MAGs_contigs %>%
  separate_rows(Contig_Names, sep = ",\\s*")  # splits by comma and optional space

# Reorder columns
EFF1_HiFi_MAGs_contigs_long <- EFF1_HiFi_MAGs_contigs_long[c("Contig_Names", "Name")]

# Save
write.table(EFF1_HiFi_MAGs_contigs_long, "RFiles/EFF1_HiFi_MAGs_contigs_long.csv", sep = ",", row.names = FALSE, quote = FALSE)
```

## SLU1
```{r}
SLU1_HiFi_MAGs_contigs <- read.table("~/Desktop/Git/Methylation/R/Methylation_R/RFiles/SLU1_HiFi_MAGs_contigs.txt", sep = "\t", header = TRUE, stringsAsFactors = FALSE)

# Convert to long format
SLU1_HiFi_MAGs_contigs_long <- SLU1_HiFi_MAGs_contigs %>%
  separate_rows(Contig_Names, sep = ",\\s*")  # splits by comma and optional space

# Reorder columns
SLU1_HiFi_MAGs_contigs_long <- SLU1_HiFi_MAGs_contigs_long[c("Contig_Names", "Name")]

# Save
write.table(SLU1_HiFi_MAGs_contigs_long, "RFiles/SLU1_HiFi_MAGs_contigs_long.csv", sep = ",", row.names = FALSE, quote = FALSE)
```

# fARGene results, REMOVE THE "_model"
```{r}
fargene_counts <- read.delim("~/Desktop/Git/Methylation/R/Methylation_R/RFiles/fargene_counts.tsv", row.names=1)

rownames(fargene_counts) <- str_remove(rownames(fargene_counts), ".hmm")
fargene_counts <- select(fargene_counts, -c("SUM"))

fargene_counts_t <- data.frame(t(fargene_counts))
fargene_counts_t$type <- c("EFF", "EFF", "EFF", "INF", "INF", "INF", "SLU", "SLU", "SLU")
fargene_counts_t$sample <- c(rownames(fargene_counts_t))

fargene <- melt(setDT(fargene_counts_t), id.vars = c("type", "sample"), variable.name = "class")

order <- c("INF", "EFF", "SLU")
fargene$type <- factor(fargene$type, levels = order)
fargene = fargene[order(fargene$type), ]

fargene$sample <- factor(fargene$sample, levels = unique(fargene$sample))


cols <- c("#fabefa", "#d86950", "#ef360c", "#a27faf", "#c308a4", "#f7b2a5",
          "#04c60a", "#1e7e21", "#779e78", "black", "#66e4e4", "#25a5a5", "#e8db16",
          "#1656e8", "navyblue", "#86a4eb", "#5b48d8", "#146eb4",  "#6f87f3", "#85baec", "#04bdfe", "#0d68c2")

p <- ggplot(fargene, aes(fill=class, y=value, x=sample)) + 
    geom_bar(stat="identity") +
    scale_y_continuous(expand = c(0,0)) +
    scale_fill_manual(values = c(cols)) +
    labs(title = "Latent ARGs by fARGene", y = "count") +
    theme_minimal() + theme( 
        text = element_text(family = "Times"),
        axis.title.x = element_blank(),
        axis.title.y = element_blank(),
        axis.text.y = element_text(size = 12),
        axis.text.x = element_text(size = 12, face = "bold"),
        plot.title = element_text(size = 20),
        plot.subtitle = element_text(size = 14)) + 
  scale_fill_manual(values=c(cols))
p

ggsave(filename = "plots/WW_fARGene.png",
       width = 10, height = 6, dpi = 300, units = "in", device="png", scale = 1)

ggsave(filename = "plots/WW_fARGene.svg",
       width = 10, height = 6, dpi = 300, units = "in", device="svg", scale = 1)
```









































# Plot HiFi MAG Pipeline results
## High-quality
```{r}
# colors
cols <- c("EFF1" = "#17b2c2", "EFF2" = "#47ecfd", "EFF3" = "#29b3fc",
            "INF1" = "#f9e30a", "INF2" = "#fcf0a0", "INF3" = "#eac51e",
            "SLU1" = "#e38236", "SLU2" = "#fec191", "SLU3" = "#daa187")

# HQ MAGs/taxa/Sample
## Phylum
p <- ggplot(HQ_MAG_count_phylum_sample, aes(x = phylum, y = n, fill = Sample)) + 
  geom_bar(stat = "identity") + scale_fill_manual(values = c(cols)) + theme_minimal() +
  ggtitle("High-quality MAGs by HiFi MAG Pipeline (phylum)") +
  scale_y_continuous(expand = c(0,0), limits = c(0,45)) +
  theme(
        text = element_text(family = "Palatino"),
        legend.text = element_text(size = 30),
        legend.title = element_blank(),
        axis.title = element_blank(),
        axis.text.y = element_text(size = 30),
        axis.text.x = element_text(size = 16, angle = 55, hjust = 0.95),
        plot.title = element_text(size = 30))
p

ggsave(filename = "plots/HQ_MAGs_phylum.png",
       width = 16, height = 13, dpi = 300, units = "in", device="png", scale = 1)

## Class
p <- ggplot(HQ_MAG_count_class_sample, aes(x = class, y = n, fill = Sample)) + 
  geom_bar(stat = "identity") + scale_fill_manual(values = c(cols)) + theme_minimal() +
  ggtitle("High-quality MAGs by HiFi MAG Pipeline (class)") +
  scale_y_continuous(expand = c(0,0), limits = c(0,45)) +
  theme(
        text = element_text(family = "Palatino"),
        legend.text = element_text(size = 30),
        legend.title = element_blank(),
        axis.title = element_blank(),
        axis.text.y = element_text(size = 30),
        axis.text.x = element_text(size = 16, angle = 55, hjust = 0.95),
        plot.title = element_text(size = 30))
p

ggsave(filename = "plots/HQ_MAGs_class.png",
       width = 16, height = 13, dpi = 300, units = "in", device="png", scale = 1)

## Genus
p <- ggplot(HQ_MAG_count_genus_sample, aes(x = genus, y = n, fill = Sample)) + 
  geom_bar(stat = "identity") + scale_fill_manual(values = c(cols)) + theme_minimal() +
  ggtitle("High-quality MAGs by HiFi MAG Pipeline (genus)") +
  scale_y_continuous(expand = c(0,0), limits = c(0,45)) +
  theme(
        text = element_text(family = "Palatino"),
        legend.text = element_text(size = 20),
        legend.title = element_blank(),
        axis.title = element_blank(),
        axis.text.y = element_text(size = 10),
        axis.text.x = element_text(size = 10, angle = 65, hjust = 0.95, face = "italic"),
        plot.title = element_text(size = 20))
p

ggsave(filename = "plots/HQ_MAGs_genus.png",
       width = 16, height = 13, dpi = 300, units = "in", device="png", scale = 1)

## Species
p <- ggplot(HQ_MAG_count_species_sample, aes(x = species, y = n, fill = Sample)) + 
  geom_bar(stat = "identity") + scale_fill_manual(values = c(cols)) + theme_minimal() +
  ggtitle("High-quality MAGs by HiFi MAG Pipeline (species)") +
  scale_y_continuous(expand = c(0,0), limits = c(0,45)) +
  theme(
        text = element_text(family = "Palatino"),
        legend.text = element_text(size = 20),
        legend.title = element_blank(),
        axis.title = element_blank(),
        axis.text.y = element_text(size = 10),
        axis.text.x = element_text(size = 10, angle = 65, hjust = 0.95, face = "italic"),
        plot.title = element_text(size = 20))
p

ggsave(filename = "plots/HQ_MAGs_species.png",
       width = 16, height = 13, dpi = 300, units = "in", device="png", scale = 1)
```

# Pie chart
```{r}
MAG_count_phylum_sample_INF <- MAG_count_phylum_sample[grepl("INF", MAG_count_phylum_sample$Sample), ]
MAG_count_phylum_sample_EFF <- MAG_count_phylum_sample[grepl("EFF", MAG_count_phylum_sample$Sample), ]
MAG_count_phylum_sample_SLU <- MAG_count_phylum_sample[grepl("SLU", MAG_count_phylum_sample$Sample), ]

# Colors
cols <- c("Acidobacteriota"	= "#e6194b","Actinobacteriota" = "#3cb44b","Armatimonadota"	= "#ffe119","Bacteroidota"	= "#808000", 
          "Bdellovibrionota" = "#7b68ee", "Cyanobacteria" = "#1f78b4",
          "Caldisericota"	= "#f28500","Campylobacterota"	= "#911eb4","Chloroflexota"	= "#46f0f0","Cloacimonadota"	= "#f032e6",
          "Nitrospirota" = "#ae7e8e", "Omnitrophota" = "#beebec",
          "Desulfobacterota"	= "#5e4fa2","Desulfobacterota_G"	= "#000080","Firmicutes"	= "#abdda4","Firmicutes_A"	= "#66c2a5",
          "Firmicutes_B"	= "#33a02c","Firmicutes_G"	= "#aaffc3","Halobacteriota"	= "#e6beff","Hydrogenedentota"	= "#e6ab02",
          "JAFGOL01"	= "#a6761d","Methanobacteriota_B"	= "#666666","Micrarchaeota"	= "#f0027f","Myxococcota"	= "#808080",
          "Nanoarchaeota"	= "#ffd8b1","Patescibacteria"	= "#fffac8","Planctomycetota"	= "#a6cee3","Proteobacteria"	= "#1e90ff",
          "Ratteibacteria" =	"#800000","Thermoplasmatota" =	"black","Thermotogota"	= "#fb9a99","Verrucomicrobiota" =	"#fabebe")


p <- ggplot(MAG_count_phylum_sample_INF, aes(x = "", y = n, fill = phylum)) +
#p <- ggplot(MAG_count_phylum_sample_EFF, aes(x = "", y = n, fill = phylum)) +
#p <- ggplot(MAG_count_phylum_sample_SLU, aes(x = "", y = n, fill = phylum)) +
  geom_col() + scale_fill_manual(values = c(cols)) +
  ggtitle("Influent") +
  #ggtitle("Effluent") +
  #ggtitle("Sludge") +
  theme_minimal() +
  theme(text = element_text(family = "Palatino"),
        panel.background = element_blank(),
        plot.background = element_blank(),
        panel.grid = element_blank(),
        axis.title = element_blank(),
        axis.text = element_blank(),
        plot.title = element_text(size = 30, hjust = 0.5, vjust = -0.5),
        legend.position = "bottom",
        legend.text = element_text(size = 20, face = "italic"),
        legend.title = element_blank()) +
  coord_polar(theta = "y")
p

ggsave(filename = "plots/MAGs_pie_INF.png",
       width = 16, height = 13, dpi = 300, units = "in", device="png", scale = 1)

ggsave(filename = "plots/MAGs_pie_EFF.png",
       width = 16, height = 13, dpi = 300, units = "in", device="png", scale = 1)

ggsave(filename = "plots/MAGs_pie_SLU.png",
       width = 16, height = 13, dpi = 300, units = "in", device="png", scale = 1)
```

# Kaiju, WW data
```{r}
WW_kaiju_summary <- read.delim("~/Desktop/Git/Methylation/R/Methylation_R/RFiles/WW_kaiju_summary.tsv")

# Into numeric
WW_kaiju_summary$percent <- as.numeric(WW_kaiju_summary$percent)

# Top % 1?
WW_kaiju_summary_filt <- WW_kaiju_summary %>% filter(percent > 1)
#WW_kaiju_summary_filt <- WW_kaiju_summary %>% filter(percent > 2)
```

# Plot
```{r}
unique(WW_kaiju_summary_filt$taxon_name)
WW_kaiju_summary_filt[WW_kaiju_summary_filt == "cannot"] <- "unknown"
unique(WW_kaiju_summary_filt$taxon_name)

cols <- c("#FF6B6B", "#1ABC9C", "#000080", "#F5F5DC", "#34495E", "#6C5CE7", "#3294c1", "#B8E986", "#5d7b65",
          "#7FDBFF", "#800080", "#a40500", "#87F28A", "#F1C40F", "#9B59B6", "#1cb9be","#E74C3C", 
          "#FFB6B0", "#8C510A", "#808000", "#00FFFF", "#FF00FF",
          "#2980B9", "#D35400", "#f6b178", "black", "grey", "#605c5a")

WW_kaiju_summary_filt$taxon_name <- factor(WW_kaiju_summary_filt$taxon_name, 
                                        levels=c("Methylotenera","Rhodoferax","Pseudomonas","Acidovorax","Acinetobacter",
                                          "Hyphomicrobium","Flavobacterium","Arcobacter","Aliarcobacter","Lactococcus","Aeromonas",
                                          "Bacteroides","Moraxella","Blautia","Faecalibacterium","Phocaeicola","Candidatus Cloacimonas",
                                          "Anaerolinea","Methanothrix","Petrimonas","Brevefilum","Proteiniphilum","Pelolinea",
                                          "Syntrophus","Tenuifilum","Viruses","unknown","unclassified"))

# reorder samples
WW_kaiju_summary_filt$type <- WW_kaiju_summary_filt$sample
WW_kaiju_summary_filt$type <- str_sub(WW_kaiju_summary_filt$type, end = -2)
order <- c("INF", "EFF", "SLU")
WW_kaiju_summary_filt$type <- factor(WW_kaiju_summary_filt$type, levels = order)
WW_kaiju_summary_filt = WW_kaiju_summary_filt[order(WW_kaiju_summary_filt$type), ]
rownames(WW_kaiju_summary_filt) <- 1:nrow(WW_kaiju_summary_filt)

WW_kaiju_summary_filt$sample <- factor(WW_kaiju_summary_filt$sample, levels = unique(WW_kaiju_summary_filt$sample))

p <- ggplot(WW_kaiju_summary_filt, aes(x = sample, y = percent, fill = taxon_name)) + 
  geom_bar(stat = "identity") + theme_minimal() + 
  ggtitle("Bacterial composition, WW data,\nAbundance > 1 % (Kaiju)") +
  scale_y_continuous(expand = c(0,0), limits = c(0,100)) +
  theme(
        text = element_text(family = "Palatino"),
        legend.text = element_text(face = "italic", size = 20),
        legend.title = element_blank(),
        axis.title = element_blank(),
        strip.text = element_text(size = 18),
        axis.text.y = element_text(size = 30),
        axis.text.x = element_text(size = 20),
        plot.title = element_text(size = 30)) + 
  scale_fill_manual(values=c(cols)) +
  facet_grid(~type, scales = "free_x", space = "free_x", labeller = labeller(type = function(x) rep("", length(x))))
p

#ggsave(filename = "plots/WW_kaiju.png",
#       width = 16, height = 13, dpi = 300, units = "in", device="png", scale = 1)
```
# Bacteroidota
```{r}
WW_kaiju_summary_Bacteroidota <- WW_kaiju_summary[grepl("acteroid", WW_kaiju_summary$taxon_name), ]

# reorder samples
WW_kaiju_summary_Bacteroidota$type <- WW_kaiju_summary_Bacteroidota$sample
WW_kaiju_summary_Bacteroidota$type <- str_sub(WW_kaiju_summary_Bacteroidota$type, end = -2)
order <- c("INF", "EFF", "SLU")
WW_kaiju_summary_Bacteroidota$type <- factor(WW_kaiju_summary_Bacteroidota$type, levels = order)
WW_kaiju_summary_Bacteroidota = WW_kaiju_summary_Bacteroidota[order(WW_kaiju_summary_Bacteroidota$type), ]
rownames(WW_kaiju_summary_Bacteroidota) <- 1:nrow(WW_kaiju_summary_Bacteroidota)


WW_kaiju_summary_Bacteroidota$sample <- factor(WW_kaiju_summary_Bacteroidota$sample, levels = unique(WW_kaiju_summary_Bacteroidota$sample))

p <- ggplot(WW_kaiju_summary_Bacteroidota, aes(x = sample, y = reads, fill = taxon_name)) + 
  geom_bar(stat = "identity") + theme_minimal() + 
  ggtitle("") +
  scale_y_continuous(expand = c(0,0)) +
  theme(
        text = element_text(family = "Palatino"),
        legend.text = element_text(face = "italic", size = 20),
        legend.title = element_blank(),
        axis.title = element_blank(),
        strip.text = element_text(size = 18),
        axis.text.y = element_text(size = 30),
        axis.text.x = element_text(size = 20),
        plot.title = element_text(size = 30)) + 
  scale_fill_manual(values=c(cols)) +
  facet_grid(~type, scales = "free_x", space = "free_x", labeller = labeller(type = function(x) rep("", length(x))))
p
```

# Sylph
```{r}
# Load data
Sylph_HiFi_merged <- read.delim("~/Desktop/Git/Methylation/R/Methylation_R/RFiles/Sylph_HiFi_merged.txt", row.names=1)

# Tax table
tax_table0 <- data.frame("rownames" = rownames(Sylph_HiFi_merged))
tax_table <- separate(tax_table0, rownames, 
                       into = c("domain", "phylum", "class", "order", "family", "genus", "species", "strain"), sep = "\\|",
                       fill = "right", remove = TRUE)
tax_table$domain <- gsub("d__", "", tax_table$domain)
tax_table$phylum <- gsub("p__", "", tax_table$phylum)
tax_table$class <- gsub("c__", "", tax_table$class)
tax_table$order <- gsub("o__", "", tax_table$order)
tax_table$family <- gsub("f__", "", tax_table$family)
tax_table$genus <- gsub("g__", "", tax_table$genus)
tax_table$species <- gsub("s__", "", tax_table$species)
tax_table$strain <- gsub("t__", "", tax_table$strain)

# Otu table
OTU_table = Sylph_HiFi_merged
rownames(OTU_table) <- 1:nrow(Sylph_HiFi_merged)
rownames(OTU_table) <- paste0("sp", rownames(OTU_table))

# Sample data
metadata <- data.frame("sample_type" = rep(c("effluent", "influent", "sludge"), each = 3),
                       "sample" = colnames(OTU_table))
rownames(metadata) <- colnames(OTU_table)
meta <- sample_data(metadata)

# Phyloseq
OTU = otu_table(OTU_table, taxa_are_rows = TRUE)
tax_table_matrix <- as.matrix(tax_table)
TAX = tax_table(tax_table_matrix)

PHY <- phyloseq(OTU, TAX, meta)

# Subset
#PHY_bac <- subset_taxa(PHY, domain == "Bacteria")
#PHY_bac_genus <- tax_glom(PHY_bac, taxrank = "genus")
#PHY_bac_species <- tax_glom(PHY_bac, taxrank = "species")
```

# Plot ordination
```{r}
# Perform PCoA
ord_PHY_bac_genus <- ordinate(PHY_bac_genus, method = "PCoA", distance = "bray")

cols <- c("#17b2c2", "#47ecfd", "#29b3fc", "#f9e30a", "#fcf0a0", "#eac51e","#7c4107","#c89b6d", "#cc7442")

# Plot the PCoA
p <- plot_ordination(PHY_bac_genus, ord_PHY_bac_genus, color = "sample") + 
  geom_point(size = 6) +
  theme_minimal() +
  theme(legend.title = element_blank(),
        legend.text = element_text(size = 24),
        text = element_text(family = "Palatino"),
        plot.title = element_text(size = 30),
        axis.text = element_text(size = 24),
        axis.title = element_text(size = 24, face = "bold")) +
  ggtitle("PCoA of Bray-Curtis Distance") + scale_color_manual(values = c(cols))
p

ggsave(filename = "plots/WW_sylph_ord.png",
       width = 16, height = 13, dpi = 300, units = "in", device="png", scale = 1)
```

# Plot bars, top15
```{r}
cols <- c("#34495E","#87F28A", "#F1C40F", "#9B59B6", "#1ABC9C","#E74C3C", 
          "#FFB6B0", "#8C510A", "#808000", "#00FFFF", "#FF00FF", "#E67E22", "#16A085",
          "#2980B9", "#D35400", "#f6b178", "grey", "#605c5a", "black")

top_n <- 15

PHY_bac_genus_rel <- transform_sample_counts(PHY_bac_genus, function(x) x / sum(x))

# Get total abundance for each taxon
taxa_sums_total <- taxa_sums(PHY_bac_genus_rel)
top_taxa <- names(sort(taxa_sums_total, decreasing = TRUE))[1:top_n]

# Subset the phyloseq object to keep only those taxa
PHY_bac_genus_rel_top <- prune_taxa(top_taxa, PHY_bac_genus_rel)

p <- plot_bar(PHY_bac_genus_rel_top, fill = "genus") + 
  geom_bar(stat = "identity") + theme_minimal() + 
  ggtitle("15 most abundant genera, WW dataset (Sylph)") +
  scale_y_continuous(expand = c(0,0)) +
  theme(
        text = element_text(family = "Palatino"),
        legend.text = element_text(face = "italic", size = 30),
        legend.title = element_blank(),
        axis.title = element_blank(),
        strip.text = element_text(size = 18),
        axis.text.y = element_text(size = 30),
        axis.text.x = element_blank(),
        plot.title = element_text(size = 30)) + 
  scale_fill_manual(values=c(cols))
p

ggsave(filename = "plots/WW_sylph_top15.png",
       width = 16, height = 13, dpi = 300, units = "in", device="png", scale = 1)
```

# Try those > 10 % 
```{r}
cols <- c("#F5F5DC", "#34495E", "#5d7b65", "#c9f7d1", "#D35400", "#a40500", "#c9f063", "#E74C3C",
          "#A3E4D7", "#cfb189", "#3294c1", "#c1a3c0", "#7FDBFF", "#FF6B6B",
          "#808000","#1ABC9C", "#584691", "#1cb9be", "#1f8d28", "#808000",
          "#fe7fca", "#6698f5", "#FFB6B0", "#F1C40F", "#f6b178")

PHY_bac_genus_rel <- transform_sample_counts(PHY_bac_genus, function(x) x / sum(x))

# Get total relative abundance for each taxon across all samples
PHY_bac_genus_rel_total_abund <- taxa_sums(PHY_bac_genus_rel)

PHY_bac_genus_rel_filtered <- prune_taxa(PHY_bac_genus_rel_total_abund > 0.1, PHY_bac_genus_rel)

# Reorder sample_type with custom levels
sample_data(PHY_bac_genus_rel_filtered)$sample_type <- factor(sample_data(PHY_bac_genus_rel_filtered)$sample_type,
                                       levels = c("influent", "effluent", "sludge"))

# Function to remove border color
plot_bar_2 <-  function (PHY_bac_genus_rel_filtered, x = "sample", y = "Abundance", fill = NULL, title = NULL, facet_grid = NULL, border_color = NA) 
{
  mdf = psmelt(PHY_bac_genus_rel_filtered)
  p = ggplot(mdf, aes_string(x = x, y = y, fill = fill))
  p = p + geom_bar(stat = "identity", position = "stack",  color = border_color)
  p = p + theme(axis.text.x = element_text(angle = -90, hjust = 0))
  if (!is.null(facet_grid)) {
    p <- p + facet_grid(facet_grid)
  }
  if (!is.null(title)) {
    p <- p + ggtitle(title)
  }
  return(p)
}

p <- plot_bar_2(PHY_bac_genus_rel_filtered, fill = "genus") + 
  geom_bar(stat = "identity") + theme_minimal() + 
  ggtitle("Bacterial composition, WW data,\nAbundance > 10 % (Sylph)") +
  scale_y_continuous(expand = c(0,0), limits = c(0,1)) +
  theme(
        text = element_text(family = "Palatino"),
        legend.text = element_text(face = "italic", size = 20),
        legend.title = element_blank(),
        axis.title = element_blank(),
        strip.text = element_text(size = 18),
        axis.text.y = element_text(size = 20),
        axis.text.x = element_text(size = 20),
        plot.title = element_text(size = 30)) + 
  scale_fill_manual(values=c(cols)) + 
  facet_grid(~sample_type, scales = "free_x", space = "free_x", labeller = labeller(sample_type = function(x) rep("", length(x))))
p

ggsave(filename = "plots/WW_sylph_>10.png",
       width = 16, height = 13, dpi = 300, units = "in", device="png", scale = 1)
```

# Sylph ----- try again with genus data only
```{r}
# Load data
Sylph_HiFi_merged <- read.delim("~/Desktop/Git/Methylation/R/Methylation_R/RFiles/Sylph_HiFi_merged_genus_full.txt", row.names=1)

# Tax table
tax_table0 <- data.frame("rownames" = rownames(Sylph_HiFi_merged))
tax_table <- separate(tax_table0, rownames, 
                       into = c("domain", "phylum", "class", "order", "family", "genus", "species", "strain"), sep = "\\|",
                       fill = "right", remove = TRUE)
tax_table$domain <- gsub("d__", "", tax_table$domain)
tax_table$phylum <- gsub("p__", "", tax_table$phylum)
tax_table$class <- gsub("c__", "", tax_table$class)
tax_table$order <- gsub("o__", "", tax_table$order)
tax_table$family <- gsub("f__", "", tax_table$family)
tax_table$genus <- gsub("g__", "", tax_table$genus)
tax_table$species <- gsub("s__", "", tax_table$species)
tax_table$strain <- gsub("t__", "", tax_table$strain)

# Otu table
OTU_table = Sylph_HiFi_merged
rownames(OTU_table) <- 1:nrow(Sylph_HiFi_merged)
rownames(OTU_table) <- paste0("sp", rownames(OTU_table))

# Sample data
metadata <- data.frame("sample_type" = rep(c("effluent", "influent", "sludge"), each = 3),
                       "sample" = colnames(OTU_table))
rownames(metadata) <- colnames(OTU_table)
meta <- sample_data(metadata)

# Phyloseq
OTU = otu_table(OTU_table, taxa_are_rows = TRUE)
tax_table_matrix <- as.matrix(tax_table)
TAX = tax_table(tax_table_matrix)

PHY <- phyloseq(OTU, TAX, meta)

PHY_genus <- tax_glom(PHY, taxrank = "genus")
```
# Plot
```{r}
cols <- c("#F5F5DC", "#34495E", "#5d7b65", "#c9f7d1", "#D35400", "#F1C40F", "#a40500", "#c9f063", "#E74C3C",
          "#A3E4D7", "#cfb189", "#3294c1", "#c1a3c0", "#7FDBFF", "#FF6B6B", "black", "navyblue",
          "#808000","#1ABC9C", "#584691", "#1cb9be", "#f6b178", "#1f8d28", "#092fdb", "#fed7d1", "orange", "coral",
          "#fe7fca", "#6698f5", "#FFB6B0", "#6f18c1", "#49c118", "grey", "#faec81",
          "#35f2ef")

# Get total relative abundance for each taxon across all samples
PHY_genus_total_abund <- taxa_sums(PHY_genus)

PHY_genus_filtered <- prune_taxa(PHY_genus_total_abund > 0.5, PHY_genus)

#TopNOTUs <- names(sort(taxa_sums(PHY_genus), TRUE)[1:20])
#PHY_filtered <- prune_species(TopNOTUs, PHY_genus)

# Reorder sample_type with custom levels
sample_data(PHY_genus_filtered)$sample_type <- factor(sample_data(PHY_genus_filtered)$sample_type,
                                       levels = c("influent", "effluent", "sludge"))

# Function to remove border color
plot_bar_2 <-  function (PHY_genus_filtered, x = "sample", y = "Abundance", fill = NULL, title = NULL, facet_grid = NULL, border_color = NA) 
{
  mdf = psmelt(PHY_genus_filtered)
  p = ggplot(mdf, aes_string(x = x, y = y, fill = fill))
  p = p + geom_bar(stat = "identity", position = "stack",  color = border_color)
  p = p + theme(axis.text.x = element_text(angle = -90, hjust = 0))
  if (!is.null(facet_grid)) {
    p <- p + facet_grid(facet_grid)
  }
  if (!is.null(title)) {
    p <- p + ggtitle(title)
  }
  return(p)
}

p <- plot_bar_2(PHY_genus_filtered, fill = "phylum") + 
  geom_bar(stat = "identity") + theme_minimal() + 
  ggtitle("Bacterial composition - phylum, WW data,\nAbundance > 10 % (Sylph)") +
  scale_y_continuous(expand = c(0,0)) +
  theme(
        text = element_text(family = "Palatino"),
        legend.text = element_text(face = "italic", size = 20),
        legend.title = element_blank(),
        axis.title = element_blank(),
        strip.text = element_text(size = 18),
        axis.text.y = element_text(size = 20),
        axis.text.x = element_text(size = 20),
        plot.title = element_text(size = 30)) + 
  scale_fill_manual(values=c(cols)) + 
  facet_grid(~sample_type, scales = "free_x", space = "free_x", labeller = labeller(sample_type = function(x) rep("", length(x))))
p

ggsave(filename = "plots/WW_sylph_>5_percent.png",
       width = 16, height = 13, dpi = 300, units = "in", device="png", scale = 1)
```



# Only Bacteroidota
```{r}
cols <- c("#F5F5DC", "#34495E", "#5d7b65", "#c9f7d1", "#D35400", "#a40500", "#c9f063", "#E74C3C",
          "#A3E4D7", "#cfb189", "#3294c1", "#c1a3c0", "#7FDBFF", "#FF6B6B", "black",
          "#808000","#1ABC9C", "#584691", "#1cb9be", "#1f8d28", "#092fdb", "#fed7d1",
          "#fe7fca", "#6698f5", "#FFB6B0", "#F1C40F", "#f6b178", "#6f18c1", "#49c118", "grey", "#faec81", "#35f2ef")

PHY_Bact <- subset_taxa(PHY_genus, order == "Bacteroidales")

#PHY_Bact_rel_total_abund <- taxa_sums(PHY_Bact_rel)
#PHY_Bact_filtered <- prune_taxa(PHY_Bact_rel_total_abund > 0.01, PHY_Bact_rel)

# Reorder sample_type with custom levels
sample_data(PHY_Bact)$sample_type <- factor(sample_data(PHY_Bact)$sample_type,
                                       levels = c("influent", "effluent", "sludge"))

# Function to remove border color
plot_bar_2 <-  function (PHY_Bact, x = "sample", y = "Abundance", fill = NULL, title = NULL, facet_grid = NULL, border_color = NA) 
{
  mdf = psmelt(PHY_Bact)
  p = ggplot(mdf, aes_string(x = x, y = y, fill = fill))
  p = p + geom_bar(stat = "identity", position = "stack",  color = border_color)
  p = p + theme(axis.text.x = element_text(angle = -90, hjust = 0))
  if (!is.null(facet_grid)) {
    p <- p + facet_grid(facet_grid)
  }
  if (!is.null(title)) {
    p <- p + ggtitle(title)
  }
  return(p)
}

p <- plot_bar_2(PHY_Bact, fill = "family") + 
  geom_bar(stat = "identity") + theme_minimal() + 
  ggtitle("Bacteroidales order - family , WW data (Sylph)") +
  scale_y_continuous(expand = c(0,0)) +
  theme(
        text = element_text(family = "Palatino"),
        legend.text = element_text(face = "italic", size = 20),
        legend.title = element_blank(),
        axis.title = element_blank(),
        strip.text = element_text(size = 18),
        axis.text.y = element_text(size = 20),
        axis.text.x = element_text(size = 20),
        plot.title = element_text(size = 30)) + 
  scale_fill_manual(values=c(cols)) + 
  facet_grid(~sample_type, scales = "free_x", space = "free_x", labeller = labeller(sample_type = function(x) rep("", length(x))))
p

ggsave(filename = "plots/WW_sylph_Bacteroidales_family.png",
       width = 16, height = 13, dpi = 300, units = "in", device="png", scale = 1)
```

## fARGene, REMOVE THE "_model"
```{r}
fargene_counts <- read.delim("~/Desktop/Git/Methylation/R/Methylation_R/RFiles/fargene_counts.tsv", row.names=1)

rownames(fargene_counts) <- str_remove(rownames(fargene_counts), ".hmm")
fargene_counts <- select(fargene_counts, -c("SUM"))

fargene_counts_t <- data.frame(t(fargene_counts))
fargene_counts_t$type <- c("EFF", "EFF", "EFF", "INF", "INF", "INF", "SLU", "SLU", "SLU")
fargene_counts_t$sample <- c(rownames(fargene_counts_t))

fargene <- melt(setDT(fargene_counts_t), id.vars = c("type", "sample"), variable.name = "class")

order <- c("INF", "EFF", "SLU")
fargene$type <- factor(fargene$type, levels = order)
fargene = fargene[order(fargene$type), ]

fargene$sample <- factor(fargene$sample, levels = unique(fargene$sample))


cols <- c("#fabefa", "#d86950", "#ef360c", "#a27faf", "#c308a4", "#f7b2a5",
          "#04c60a", "#1e7e21", "#779e78", "black", "#66e4e4", "#25a5a5", "#e8db16",
          "#1656e8", "navyblue", "#86a4eb", "#5b48d8", "#146eb4",  "#6f87f3", "#85baec", "#04bdfe", "#0d68c2")

p <- ggplot(fargene, aes(fill=class, y=value, x=sample)) + 
    geom_bar(stat="identity") +
    scale_y_continuous(expand = c(0,0)) +
    scale_fill_manual(values = c(cols)) +
    theme_minimal() +
    theme(text = element_text(family = "Palatino"),
          legend.title = element_blank(),
          legend.text = element_text(size = 18),
          strip.text = element_blank(),
          axis.title = element_blank(),
          plot.title = element_text(size = 30, face = "bold", hjust = 0.5),
          axis.text = element_text(size = 20)) + 
  ggtitle("fARGene results") + facet_grid(~type, scale = "free", space = "free")
p

ggsave(filename = "plots/WW_fARGene.png",
       width = 16, height = 13, dpi = 300, units = "in", device="png", scale = 1)
```


# Number of ARG MAGs by methylation & HiFi MAG Pipeline
## Venn Diagram
```{r}
# https://jolars.github.io/eulerr/reference/plot.euler.html
ARG_MAGs <- read.delim("~/Desktop/Git/Methylation/R/Methylation_R/RFiles/ARG_MAGs.txt")
ARG_MAGs$Methylation_based_analysis <- as.numeric(ARG_MAGs$Methylation_based_analysis)
ARG_MAGs$HiFi_MAG_Pipeline <- as.numeric(ARG_MAGs$HiFi_MAG_Pipeline)

INF_MAGs <- subset(ARG_MAGs, sample_type == "influent")
EFF_MAGs <- subset(ARG_MAGs, sample_type == "effluent")
SLU_MAGs <- subset(ARG_MAGs, sample_type == "dried sludge")

fit <- euler(c("A" = sum(INF_MAGs$Methylation_based_analysis),
               "B" = sum(INF_MAGs$HiFi_MAG_Pipeline), "A&B" = sum(rowSums(INF_MAGs[,2:3]) > 1)))

png("plots/INF_venn.png",width=3.25,height=3.25,units="in",res=1500)
plot(fit, main = "Influent",
     fills = list(fill = c("#f9e30a", "#fcf0a0")), 
     alpha = 0.5, edges = FALSE, quantities = TRUE,
     labels = list(font = 0.75, cex = 0.60, cex.lab=0.5, 
                   labels = c("Methylation based analysis", "HiFi MAG Pipeline")))
dev.off()

fit <- euler(c("A" = sum(EFF_MAGs$Methylation_based_analysis),
               "B" = sum(EFF_MAGs$HiFi_MAG_Pipeline), "A&B" = sum(rowSums(EFF_MAGs[,2:3]) > 1)))

png("plots/EFF_venn.png",width=3.25,height=3.25,units="in",res=1500)
plot(fit, main = "Effluent",
     fills = list(fill = c("#29b3fc", "#47ecfd")), 
     alpha = 0.5, edges = FALSE, quantities = TRUE,
     labels = list(font = 0.75, cex = 0.60, cex.lab=0.5, 
                   labels = c("Methylation based analysis", "HiFi MAG Pipeline")))
dev.off()

fit <- euler(c("A" = sum(SLU_MAGs$Methylation_based_analysis),
               "B" = sum(SLU_MAGs$HiFi_MAG_Pipeline), "A&B" = sum(rowSums(SLU_MAGs[,2:3]) > 1)))

png("plots/SLU_venn.png",width=3.25,height=3.25,units="in",res=1500)
plot(fit, main = "Dried sludge",
     fills = list(fill = c("#e38236", "#fec191")), 
     alpha = 0.5, edges = FALSE, quantities = TRUE,
     labels = list(font = 0.75, cex = 0.60, cex.lab=0.5, 
                   labels = c("Methylation based analysis", "HiFi MAG Pipeline")))
dev.off()

#plot(euler(ARG_MAGs, by = list(sample_type)), legend = TRUE)
```
# Number of ARG MAGs by methylation & HiFi MAG Pipeline
## Barplot
```{r}
ARG_MAGs_counts <- read.delim("~/Desktop/Git/Methylation/R/Methylation_R/RFiles/ARG_MAG_counts.txt")
ARG_MAGs_counts$Methylation_based_analysis <- as.numeric(ARG_MAGs_counts$Methylation_based_analysis)
ARG_MAGs_counts$HiFi_MAG_Pipeline <- as.numeric(ARG_MAGs_counts$HiFi_MAG_Pipeline)

ARG_MAGs_counts_long <- melt(setDT(ARG_MAGs_counts), id.vars = c("sample"), variable.name = "Methylation_based_analysis")
ARG_MAGs_counts_long$sample_type <- ARG_MAGs_counts_long$sample
ARG_MAGs_counts_long$sample_type <- gsub("[0-9]", "", ARG_MAGs_counts_long$sample_type)
ARG_MAGs_counts_long$sample_type <- gsub("INF", "influent", ARG_MAGs_counts_long$sample_type)
ARG_MAGs_counts_long$sample_type <- gsub("EFF", "effluent", ARG_MAGs_counts_long$sample_type)
ARG_MAGs_counts_long$sample_type <- gsub("SLU", "dried sludge", ARG_MAGs_counts_long$sample_type)

colnames(ARG_MAGs_counts_long) <- c("sample", "Method", "count", "sample_type")

df <- ARG_MAGs_counts_long[1:18,]
df$count <- as.integer(df$count)

df$sample <- factor(df$sample, levels=c("INF1", "INF2", "INF3",
                                          "EFF1", "EFF2", "EFF3",
                                          "SLU1", "SLU2", "SLU3"))

p <- ggplot(df, aes(x = sample, y = count, fill = interaction(sample, Method))) + 
  geom_bar(stat = "identity", position = "dodge") + theme_bw() +
  scale_y_continuous(labels = scales::number_format(accuracy = 1), expand = c(0,0), limits = c(0,10.75)) +
  scale_fill_manual(values = c(
        EFF1.Methylation_based_analysis = "#29b3fc", 
        EFF2.Methylation_based_analysis = "#29b3fc", 
        EFF3.Methylation_based_analysis = "#29b3fc",
        EFF1.HiFi_MAG_Pipeline = "#b7e0f6", EFF2.HiFi_MAG_Pipeline = "#b7e0f6", EFF3.HiFi_MAG_Pipeline = "#b7e0f6",
        INF1.Methylation_based_analysis = "#f9e30a",
        INF2.Methylation_based_analysis = "#f9e30a",
        INF3.Methylation_based_analysis = "#f9e30a",
        INF1.HiFi_MAG_Pipeline = "#f9f5de", INF2.HiFi_MAG_Pipeline = "#f9f5de", INF3.HiFi_MAG_Pipeline = "#f9f5de",
        SLU1.Methylation_based_analysis = "#daa187",
        SLU2.Methylation_based_analysis = "#daa187",
        SLU3.Methylation_based_analysis = "#daa187",
        SLU1.HiFi_MAG_Pipeline = "#ffdcc0", SLU2.HiFi_MAG_Pipeline = "#ffdcc0", SLU3.HiFi_MAG_Pipeline = "#ffdcc0")) +
  ggtitle("Number of bins with ARGs") +  theme_minimal() +
  theme(text = element_text(family = "Palatino"),
        legend.position = "none",
        strip.text = element_text(size = 34, face = "bold"),
        axis.title = element_blank(),
        axis.text.y = element_text(size = 28),
        axis.text.x = element_text(size = 36, angle = 90, hjust = 1, color = "white"),
        plot.title = element_text(size = 40, vjust = 0.5, hjust = 0.5)) + 
  facet_grid(~factor(sample_type, levels=c("influent", "effluent", "dried sludge")), space = "free", scales = "free")
p

ggsave(filename = "plots/ARG_MAGs_bar_plot.png",
       width = 16, height = 13, dpi = 300, units = "in", device="png", scale = 1)

ggsave(filename = "plots/ARG_MAGs_bar_plot.svg",
       width = 16, height = 13, dpi = 300, units = "in", device="svg", scale = 1)
```



# Analysis of Kaiju results (Synthetic community)
```{r}
kaiju_summary <- read.delim("~/Desktop/Git/Methylation/R/Methylation_R/RFiles/kaiju_summary.tsv")

# Remove extra column headers
kaiju_summary <- kaiju_summary[- grep("sample", kaiju_summary$sample),]

# Into numeric
kaiju_summary$percent <- as.numeric(kaiju_summary$percent)

# Top % 1?
kaiju_summary_filt <- kaiju_summary %>% filter(percent > 1)

# Do some editing for plotting
kaiju_summary_filt$sample <- gsub("bcAd1023T--bcAd1023T", "bcAd1023T", kaiju_summary_filt$sample)
kaiju_summary_filt$sample <- gsub("bcAd1037T--bcAd1037T", "bcAd1037T", kaiju_summary_filt$sample)
kaiju_summary_filt$sample <- gsub("bcAd1039T--bcAd1039T", "bcAd1039T", kaiju_summary_filt$sample)
kaiju_summary_filt$sample <- gsub("bcAd1046T--bcAd1046T", "bcAd1046T", kaiju_summary_filt$sample)
kaiju_summary_filt$sample <- gsub("bcAd1063T--bcAd1063T", "bcAd1063T", kaiju_summary_filt$sample)

# Add sample info
kaiju_summary_filt$sample_info <- ifelse(kaiju_summary_filt$sample == "bcAd1023T", "High", "Low")

kaiju_summary_filt <- kaiju_summary_filt %>%
  mutate(sample_info = case_when(
    sample == "bcAd1023T" ~ "Tetracycline: 0.2, 15 ªC, rep-5 (bcAd1023T)",
    sample == "bcAd1037T" ~ "Sulfonamide: 20, 25 ªC, rep-3 (bcAd1037T)",
    sample == "bcAd1039T" ~ "Sulfonamide: 20, 25 ªC, rep-4 (bcAd1039T)",
    sample == "bcAd1046T" ~ "Tetracycline: 0.02, 37 ªC, rep-3 (bcAd1046T)",
    sample == "bcAd1063T" ~ "no ab, 37 ªC, rep-3 (bcAd1063T)",
  ))
kaiju_summary_filt$sample_info = factor(kaiju_summary_filt$sample_info, 
                                        levels=c("Tetracycline: 0.2, 15 ªC, rep-5 (bcAd1023T)",
                                                  "Sulfonamide: 20, 25 ªC, rep-3 (bcAd1037T)",
                                                  "Sulfonamide: 20, 25 ªC, rep-4 (bcAd1039T)",
                                                  "Tetracycline: 0.02, 37 ªC, rep-3 (bcAd1046T)",
                                                  "no ab, 37 ªC, rep-3 (bcAd1063T)"))
```
## Plot
```{r}
cols <- c("#34495E", "#800080", "#3bccf7", "#9ACD32", "#8B4513", "#00FF00", "#E74C3C", "black", "#DAA520", "#800000", "#000080", "#4682B4", "#008080", "grey")

kaiju_summary_filt[kaiju_summary_filt == "cannot be assigned to a (non-viral) genus"] <- "unknown"

kaiju_summary_filt$taxon_name <- factor(kaiju_summary_filt$taxon_name, 
                                        levels=c("Acinetobacter", "Aeromonas", "Agrobacterium",
                                                 "Azorhizobium", "Azospirillum", "Citrobacter",
                                                 "Elizabethkingia", "Enterococcus", "Kluyvera",
                                                 "Paracoccus", "Pseudomonas", "Sphingobacterium",
                                                 "Sphingobium", "unknown"))

p <- ggplot(kaiju_summary_filt, aes(x = sample, y = percent, fill = taxon_name)) + 
  geom_bar(stat = "identity", alpha = 0.85) + theme_minimal() + 
  ggtitle("Bacterial composition, HAMBI data,\nAbundance > 1 % (Kaiju)") +
  scale_y_continuous(expand = c(0,0), limits = c(0,100)) +
  theme(
        text = element_text(family = "Palatino"),
        legend.text = element_text(size = 20, face = "italic"),
        legend.title = element_blank(),
        axis.title = element_blank(),
        strip.text = element_text(size = 18),
        axis.text.y = element_text(size = 30),
        axis.text.x = element_blank(),
        plot.title = element_text(size = 30, face = "bold")) + 
  scale_fill_manual(values=c(cols)) + 
  facet_grid(~sample_info, space = "free", scale = "free", 
             labeller = label_wrap_gen(width=16), switch = "both")
p

ggsave(filename = "plots/HAMBI_kaiju.png",
       width = 16, height = 13, dpi = 300, units = "in", device="png", scale = 1)
```


# Feature Importance
```{r}
feature_importance <- read.delim("~/Desktop/Git/Methylation/R/Methylation_R/RFiles/feature_importance.txt")
df <- feature_importance[rownames(feature_importance) != "All", ]
df$Features <- as.numeric(df$Features)

value_h <- feature_importance[grep("All", feature_importance$Features), ]
value_h = print(value_h["Accuracy.Test"][ ,])

value_v <- 150

p <- ggplot(df, aes(x=Features, y=Accuracy.Test)) + ggtitle("Feature Importance") +
  geom_line() + theme_bw() +
  theme(text = element_text(family = "Palatino"),
    plot.title = element_text(size = 34),
    axis.title = element_text(size = 34),
    axis.text = element_text(size = 28))
p1 <- p + geom_hline(yintercept=value_h, linetype="dashed", color = "red", linewidth = 2) +
  geom_text(aes(230, value_h, label = "All features", vjust = - 1),
            color = "red", family = "Palatino", size = 10)
p2 <- p1 + geom_vline(xintercept=value_v, linetype="dotdash", color = "blue", linewidth = 1)
p2

ggsave(filename = "plots/feature_importance.png",
       width = 16, height = 13, dpi = 300, units = "in", device="png", scale = 1)
```

# Summary of HiFi MAG Pipeline results
```{r}
colnames(MAG_count_sample) <- c("Sample", "other_MAGs")
colnames(HQ_MAG_count_sample) <- c("Sample", "HQ_MAGs")

MAG_counts <- cbind(MAG_count_sample, "HQ_MAGs" = HQ_MAG_count_sample$HQ_MAGs)

long_MAG_counts <- melt(setDT(MAG_counts), id.vars = c("Sample"), variable.name = "MAGs")
colnames(long_MAG_counts) <- c("sample", "MAGs", "count")

# Colors
#cols <- c("#87bbda", "#0540a0")

cols <- c("#17b2c2", "#47ecfd", "#29b3fc", "#f9e30a", "#fcf0a0", "#eac51e","#e38236","#fec191", "#daa187")

#cols <- c("EFF1" = "#17b2c2", "EFF2" = "#47ecfd", "EFF3" = "#29b3fc","INF1" = "#f9e30a", "INF2" = "#fcf0a0", "INF3" = "#eac51e","SLU1" = "#e38236", "SLU2" = "#fec191", "SLU3" = "#daa187")

# Reorder
long_MAG_counts$sample <- factor(long_MAG_counts$sample,
                                 levels=c("INF1", "INF2", "INF3",
                                          "EFF1", "EFF2", "EFF3",
                                          "SLU1", "SLU2", "SLU3"))
# Add ARGs
counts_MAGs_with_ARGs <- MAGs_with_ARGs %>% count(Sample)
View(counts_MAGs_with_ARGs)

counts_HQ_MAGs_with_ARGs <- HQ_MAGs_with_ARGs %>% count(Sample)
View(counts_HQ_MAGs_with_ARGs)

ARGs <- data.frame(sample = c("EFF1","EFF2","EFF3", "INF1","INF2","INF3","SLU1","SLU2","SLU3",
                              "EFF1","EFF2","EFF3","INF1","INF2","INF3","SLU1","SLU2","SLU3"),
                   MAGs = c("other_MAGs", "other_MAGs", "other_MAGs", "other_MAGs", "other_MAGs",
                            "other_MAGs", "other_MAGs", "other_MAGs", "other_MAGs",
                            "HQ_MAGs", "HQ_MAGs", "HQ_MAGs", "HQ_MAGs", "HQ_MAGs", "HQ_MAGs", "HQ_MAGs", "HQ_MAGs", "HQ_MAGs"),
                   ARGs = c(0,1,0,0,0,0,3,2,1,
                            0,1,0,0,0,0,0,2,2))

long_MAG_counts$ARGs <- as.numeric(ARGs$ARGs)

long_MAG_counts$ARGs[long_MAG_counts$ARGs == 0] <- " "

# Plot
p <- ggplot(long_MAG_counts, aes(x = sample, y = count, fill = MAGs, label = ARGs)) + 
  geom_bar(stat = "identity", alpha = 0.5) + scale_fill_manual(values = c(cols)) + theme_minimal() +
  geom_text(size = 10, position = position_stack(vjust = .5), color = "red", family = "Courier",) +
  ggtitle("Number of MAGs by HiFi MAG Pipeline") +  theme_minimal() +
  scale_y_continuous(expand = c(0,0)) +
  theme(
        text = element_text(family = "Palatino"),
        legend.text = element_text(size = 30),
        legend.title = element_blank(),
        axis.title = element_blank(),
        axis.text.y = element_text(size = 30),
        axis.text.x = element_text(size = 24, face = "bold"),
        plot.title = element_text(size = 30))
p


p <- ggplot(long_MAG_counts, aes(x = sample, y = count, fill = interaction(sample, MAGs), label = ARGs)) + 
  geom_bar(stat = "identity") +
  geom_text(size = 10, position = position_stack(vjust = .5), color = "red", family = "Palatino") +
  theme_minimal() +
  scale_fill_manual(values = c(EFF1.HQ_MAGs = "#17b2c2", EFF2.HQ_MAGs = "#47ecfd", EFF3.HQ_MAGs = "#29b3fc",
                               EFF1.other_MAGs = "#98e6ee", EFF2.other_MAGs = "#bff8fe", EFF3.other_MAGs = "#b7e0f6",
                               INF1.HQ_MAGs = "#f9e30a", INF2.HQ_MAGs = "#fcf0a0", INF3.HQ_MAGs = "#eac51e",
                               INF1.other_MAGs = "#fef174", INF2.other_MAGs = "#f9f5de", INF3.other_MAGs = "#f0de8e",
                               SLU1.HQ_MAGs = "#e38236", SLU2.HQ_MAGs = "#fec191", SLU3.HQ_MAGs = "#daa187",
                               SLU1.other_MAGs = "#e5aa7c", SLU2.other_MAGs = "#ffdcc0", SLU3.other_MAGs = "#dbbeb1")) +
  ggtitle("Number of MAGs by HiFi MAG Pipeline") +  theme_minimal() +
  scale_y_continuous(expand = c(0,0)) +
  theme(
        text = element_text(family = "Palatino"),
        legend.position = "none",
        axis.title = element_blank(),
        axis.text.y = element_text(size = 30),
        axis.text.x = element_text(size = 24, face = "bold"),
        plot.title = element_text(size = 30))
p

ggsave(filename = "plots/Pipeline_MAGs.png",
       width = 16, height = 13, dpi = 300, units = "in", device="png", scale = 1)
```
# Summary of HiFi MAG Pipeline results
###### muokkaa tätä kokonaisuudessaan, tsekkaa luvut  #########
```{r}
colnames(MAG_count_sample) <- c("Sample", "other_MAGs")
colnames(HQ_MAG_count_sample) <- c("Sample", "HQ_MAGs")

MAG_counts <- cbind(MAG_count_sample, "HQ_MAGs" = HQ_MAG_count_sample$HQ_MAGs)

long_MAG_counts <- melt(setDT(MAG_counts), id.vars = c("Sample"), variable.name = "MAGs")
colnames(long_MAG_counts) <- c("sample", "MAGs", "count")

# Colors
#cols <- c("#87bbda", "#0540a0")

cols <- c("#17b2c2", "#47ecfd", "#29b3fc", "#f9e30a", "#fcf0a0", "#eac51e","#e38236","#fec191", "#daa187")

#cols <- c("EFF1" = "#17b2c2", "EFF2" = "#47ecfd", "EFF3" = "#29b3fc","INF1" = "#f9e30a", "INF2" = "#fcf0a0", "INF3" = "#eac51e","SLU1" = "#e38236", "SLU2" = "#fec191", "SLU3" = "#daa187")

# Reorder
long_MAG_counts$sample <- factor(long_MAG_counts$sample,
                                 levels=c("INF1", "INF2", "INF3",
                                          "EFF1", "EFF2", "EFF3",
                                          "SLU1", "SLU2", "SLU3"))
# Add ARGs
counts_MAGs_with_ARGs <- MAGs_with_ARGs %>% count(Sample)
View(counts_MAGs_with_ARGs)

counts_HQ_MAGs_with_ARGs <- HQ_MAGs_with_ARGs %>% count(Sample)
View(counts_HQ_MAGs_with_ARGs)

ARGs <- data.frame(sample = c("EFF1","EFF2","EFF3", "INF1","INF2","INF3","SLU1","SLU2","SLU3",
                              "EFF1","EFF2","EFF3","INF1","INF2","INF3","SLU1","SLU2","SLU3"),
                   MAGs = c("other_MAGs", "other_MAGs", "other_MAGs", "other_MAGs", "other_MAGs",
                            "other_MAGs", "other_MAGs", "other_MAGs", "other_MAGs",
                            "HQ_MAGs", "HQ_MAGs", "HQ_MAGs", "HQ_MAGs", "HQ_MAGs", "HQ_MAGs", "HQ_MAGs", "HQ_MAGs", "HQ_MAGs"),
                   ARGs = c(0,1,0,0,0,0,3,2,1,
                            0,1,0,0,0,0,0,2,2))

long_MAG_counts$ARGs <- as.numeric(ARGs$ARGs)

long_MAG_counts$ARGs[long_MAG_counts$ARGs == 0] <- " "

# Plot
p <- ggplot(long_MAG_counts, aes(x = sample, y = count, fill = MAGs, label = ARGs)) + 
  geom_bar(stat = "identity", alpha = 0.5) + scale_fill_manual(values = c(cols)) + theme_minimal() +
  geom_text(size = 10, position = position_stack(vjust = .5), color = "red", family = "Courier",) +
  ggtitle("Number of MAGs by HiFi MAG Pipeline") +  theme_minimal() +
  scale_y_continuous(expand = c(0,0)) +
  theme(
        text = element_text(family = "Palatino"),
        legend.text = element_text(size = 30),
        legend.title = element_blank(),
        axis.title = element_blank(),
        axis.text.y = element_text(size = 30),
        axis.text.x = element_text(size = 24, face = "bold"),
        plot.title = element_text(size = 30))
p


p <- ggplot(long_MAG_counts, aes(x = sample, y = count, fill = interaction(sample, MAGs), label = ARGs)) + 
  geom_bar(stat = "identity") +
  geom_text(size = 10, position = position_stack(vjust = .5), color = "red", family = "Palatino") +
  theme_minimal() +
  scale_fill_manual(values = c(EFF1.HQ_MAGs = "#17b2c2", EFF2.HQ_MAGs = "#47ecfd", EFF3.HQ_MAGs = "#29b3fc",
                               EFF1.other_MAGs = "#98e6ee", EFF2.other_MAGs = "#bff8fe", EFF3.other_MAGs = "#b7e0f6",
                               INF1.HQ_MAGs = "#f9e30a", INF2.HQ_MAGs = "#fcf0a0", INF3.HQ_MAGs = "#eac51e",
                               INF1.other_MAGs = "#fef174", INF2.other_MAGs = "#f9f5de", INF3.other_MAGs = "#f0de8e",
                               SLU1.HQ_MAGs = "#e38236", SLU2.HQ_MAGs = "#fec191", SLU3.HQ_MAGs = "#daa187",
                               SLU1.other_MAGs = "#e5aa7c", SLU2.other_MAGs = "#ffdcc0", SLU3.other_MAGs = "#dbbeb1")) +
  ggtitle("Number of MAGs by HiFi MAG Pipeline") +  theme_minimal() +
  scale_y_continuous(expand = c(0,0)) +
  theme(
        text = element_text(family = "Palatino"),
        legend.position = "none",
        axis.title = element_blank(),
        axis.text.y = element_text(size = 30),
        axis.text.x = element_text(size = 24, face = "bold"),
        plot.title = element_text(size = 30))
p

ggsave(filename = "plots/Pipeline_MAGs.png",
       width = 16, height = 13, dpi = 300, units = "in", device="png", scale = 1)
```
