{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "70b227aa-3598-459b-9c4c-6bb1c755791e",
   "metadata": {},
   "source": [
    "# UMAP clustering for the methylation signature Position Weight Matrices\n",
    "## HAMBI data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e929d811-8da6-47d5-aab2-aeea3493be08",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "import os \n",
    "import plotly.express as px\n",
    "import plotly.graph_objects as go\n",
    "from plotly.subplots import make_subplots\n",
    "import umap\n",
    "from sklearn.metrics import silhouette_score\n",
    "import seaborn as sns\n",
    "import os\n",
    "from PIL import Image, ImageFont\n",
    "from sklearn.cluster import KMeans\n",
    "\n",
    "seed = 98\n",
    "import matplotlib.pyplot as plt\n",
    "import seaborn as sns\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3e727344-51ac-43b6-ba2a-6fd8084b80d3",
   "metadata": {},
   "outputs": [],
   "source": [
    "path_to_images = '/scratch/project_2006608/Methylation/notebooks/UMAP_HAMBI/'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e8f02c19-29cf-42fb-a5c3-04073543ebbd",
   "metadata": {},
   "outputs": [],
   "source": [
    "file_path = '/scratch/project_2006608/Methylation/HAMBI_data/merged_data.tsv'\n",
    "df = pd.read_csv(file_path, sep='\\t', index_col=0, low_memory=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b82e9f92-98b6-4560-8532-ea58a8b26c3a",
   "metadata": {},
   "outputs": [],
   "source": [
    "print(df.shape)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b111d45d-7000-4b47-94f3-c6770c9b1ca6",
   "metadata": {},
   "source": [
    "## FILTERING\n",
    "### Removing the contigs that have no methyaltions (or under 20) detected for all of the three types. These are the rows where all values in the columns are 0s\n",
    "#### All data ðŸ’ "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "62810be5-9bed-4c7c-ad52-0711a3351cf1",
   "metadata": {},
   "outputs": [],
   "source": [
    "# remove rows where df.iloc[:, :-17] meaning no methylation detected or less than 20 detected since those are converted to 0 filled matrices\n",
    "df = df.loc[(df.iloc[:, :-17] != 0).any(axis=1)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e5cd2ad6-e9ee-465e-a8f1-930127b3ab82",
   "metadata": {},
   "outputs": [],
   "source": [
    "print(df.shape[0])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "08ec1a3c-55e2-42e7-b527-cd540a3da6c5",
   "metadata": {},
   "outputs": [],
   "source": [
    "df['Domain to species'].value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "16725f1f-ca6c-47db-8ebc-cfef787c7443",
   "metadata": {},
   "outputs": [],
   "source": [
    "df['Domain to species'].nunique(dropna = False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3a089cdd-f203-4a21-98cb-05a295e515c2",
   "metadata": {},
   "outputs": [],
   "source": [
    "df['Domain to genus'].nunique(dropna = False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1ee51b71-7823-4b25-8b55-fefdea5e060b",
   "metadata": {},
   "outputs": [],
   "source": [
    "df['Domain to family'].nunique(dropna = False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "38699591-4481-45a9-98cc-f38bbe863807",
   "metadata": {},
   "outputs": [],
   "source": [
    "df['Domain to order'].nunique(dropna = False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "58e19167-47f5-4298-9c5c-b37868ad893a",
   "metadata": {},
   "outputs": [],
   "source": [
    "df['Domain to class'].nunique(dropna = False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "85aef2be-0942-4946-bfe5-cdedb4647057",
   "metadata": {},
   "outputs": [],
   "source": [
    "df['Domain to phylum'].nunique(dropna = False)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "89c92161-db56-4689-9aed-b62b7b51b26e",
   "metadata": {},
   "source": [
    "### Number of contigs where missing the species information\n",
    "#### All data ðŸ’ "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5610762d-1520-4264-a73c-a04bafa2157e",
   "metadata": {},
   "outputs": [],
   "source": [
    "len(df[df['Domain to species'].isna()])"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d5f8355c-a5eb-4d7c-b574-84288458901d",
   "metadata": {},
   "source": [
    "### Add contig lengths"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b7e51efe-33f6-47f5-aba0-407075d903fa",
   "metadata": {},
   "outputs": [],
   "source": [
    "## Bring the contig lengths from Puhti:\n",
    "file_path = '/scratch/project_2006608/Methylation/HAMBI_data/metagenomic_assembly/contigs_lengths.tsv'\n",
    "\n",
    "df_lengths = pd.read_csv(file_path, sep='\\t', index_col=0, header=0, low_memory=False)\n",
    "print(df_lengths.shape[0])\n",
    "df_lengths.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "57a9591a-6b4d-4cc1-9007-cad1e504868a",
   "metadata": {},
   "outputs": [],
   "source": [
    "## Append to merged_data.tsv\n",
    "df_ext = df.copy()\n",
    "\n",
    "# Reorder to match\n",
    "df_ordered = df_lengths.loc[df_ext.index]\n",
    "\n",
    "df_lengths = pd.concat([df_ext, df_ordered], axis=1)\n",
    "df_lengths.tail()\n",
    "df = df_lengths.copy()\n",
    "df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "207f21f4-82bc-40c5-a464-8526d41ea9c2",
   "metadata": {},
   "outputs": [],
   "source": [
    "df.head()\n",
    "print(df.iloc[:, :-18])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ba820e9a-5889-4c78-a650-81fb1d65ef34",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "n_neighbors = [20, 30 ]\n",
    "min_dist = [0.01, 0.1, 0.2]\n",
    "colors = sns.color_palette(\"hls\", 32).as_hex()\n",
    "\n",
    "unique_species = df['Domain to species'].unique()\n",
    "num_species = len(unique_species)\n",
    "color_dict = [\n",
    "  \"#FF0000\", \"#00FF00\", \"#0000FF\", \"#FFFF00\", \"#FF00FF\", \"#00FFFF\", \"#800000\", \"#008000\",\n",
    "  \"#000080\", \"#808000\", \"#800080\", \"#008080\", \"#FF4500\", \"#3bccf7\", \"#4682B4\", \"#DAA520\",\n",
    "  \"#FF1493\", \"#9ACD32\", \"#8A2BE2\", \"#5F9EA0\", \"#D2691E\", \"#DC143C\", \"#B8860B\", \"#32CD32\",\n",
    "  \"#FFD700\", \"#20B2AA\", \"#FF6347\", \"#4169E1\", \"#8B4513\", \"#6A5ACD\", \"#FF7F50\", \"#2F4F4F\"\n",
    "]\n",
    "custom_colors = color_dict\n",
    "\n",
    "for n in n_neighbors:\n",
    "    for m in min_dist:\n",
    "        \n",
    "        reducer = umap.UMAP(n_neighbors=n, min_dist=m, random_state=seed)\n",
    "        embedding = reducer.fit_transform(df.iloc[:, :-18])\n",
    "        UMAP_df  = pd.DataFrame({\n",
    "            'UMAP1': embedding[:, 0],\n",
    "            'UMAP2': embedding[:, 1],\n",
    "            'Domain to species': df['Domain to species'],\n",
    "            'Domain to genus': df['Domain to genus'],\n",
    "            'element': df['element'],\n",
    "            'contig_length': df['contig_length']\n",
    "        })\n",
    "        # clustering\n",
    "        kmeans = KMeans(n_clusters=11, random_state=seed)\n",
    "        \n",
    "        k_labels = kmeans.fit_predict(embedding)\n",
    "        \n",
    "        UMAP_df['cluster'] = k_labels\n",
    "        UMAP_df['Domain to species'] = UMAP_df['Domain to species'].fillna('NA')\n",
    "        column = 'Domain to species'\n",
    "        value_counts = UMAP_df[column].value_counts().to_dict()\n",
    "\n",
    "        UMAP_df[f'{column}_with_counts'] = UMAP_df[column].apply(lambda x: f'{x} ({value_counts[x]})')\n",
    "\n",
    "        UMAP_df['count'] = UMAP_df[column].map(value_counts)\n",
    "        UMAP_df = UMAP_df.sort_values(by=['count'], ascending=False)\n",
    "\n",
    "        UMAP_df[f'{column}_with_counts'] = pd.Categorical(\n",
    "            UMAP_df[f'{column}_with_counts'],\n",
    "            categories=UMAP_df[f'{column}_with_counts'].unique(),\n",
    "            ordered=True\n",
    "        )\n",
    "        \n",
    "        UMAP_df['sqrt_contig_length'] = np.sqrt(UMAP_df['contig_length'])\n",
    "        \n",
    "        fig = px.scatter(UMAP_df, \n",
    "                            x='UMAP1', \n",
    "                            y='UMAP2', \n",
    "                            labels={f'{column}_with_counts': column},\n",
    "                            color=f'{column}_with_counts', \n",
    "                            title=f' HAMBI - UMAP with n_neighbors={n}, min_dist={m}', \n",
    "                            color_discrete_sequence=custom_colors,\n",
    "                            hover_data={'cluster': True, 'Domain to species': True, 'Domain to genus': True,\n",
    "                                        'element': True, 'contig_length': True},\n",
    "                        size='sqrt_contig_length')\n",
    "        title = f' HAMBI - UMAP with n_neighbors={n}, min_dist={m}'\n",
    "        fig.update_layout(\n",
    "            height=1500,\n",
    "            width=1200,\n",
    "            title_text=title,\n",
    "            showlegend=True,\n",
    "            legend=dict(\n",
    "                x=0.5,\n",
    "                y=-0.1,\n",
    "                traceorder=\"normal\",\n",
    "                xanchor='center',\n",
    "                yanchor='top',\n",
    "                orientation='h'\n",
    "            ),\n",
    "            template='simple_white',\n",
    "            xaxis=dict(\n",
    "                showgrid=True,\n",
    "                gridcolor='lightgray',\n",
    "                zeroline=False,\n",
    "                showline=True,  \n",
    "                linecolor='black', \n",
    "                linewidth=1,\n",
    "                mirror=True\n",
    "            ),\n",
    "            yaxis=dict(\n",
    "                showgrid=True,\n",
    "                gridcolor='lightgray',\n",
    "                zeroline=False,\n",
    "                showline=True,\n",
    "                linecolor='black',\n",
    "                linewidth=1,\n",
    "                mirror=True\n",
    "            )\n",
    "        )\n",
    "        fig.show()\n",
    "        fig.write_image(f'UMAP_HAMBI/UMAP_{n}_{m}.png')\n",
    "        fig.write_html(f'UMAP_HAMBI/UMAP_{n}_{m}.html')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9e8f2470-d522-4888-9544-05a309383af7",
   "metadata": {},
   "source": [
    "### Using the best n_neighbours and min_dist to get clear clusters\n",
    "#### All data ðŸ’ "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "97ab916d-9527-4d8d-9de4-c36910abd44e",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "n_neighbors = [20]\n",
    "min_dist = [0.5]\n",
    "color_dict = [\n",
    "  \"#FF0000\", \"#00FF00\", \"#0000FF\", \"#FFFF00\", \"#FF00FF\", \"#00FFFF\", \"#800000\", \"#008000\",\n",
    "  \"#000080\", \"#808000\", \"#800080\", \"#008080\", \"#FF4500\", \"#3bccf7\", \"#4682B4\", \"#DAA520\",\n",
    "  \"#FF1493\", \"#9ACD32\", \"#8A2BE2\", \"#5F9EA0\", \"#D2691E\", \"#DC143C\", \"#B8860B\", \"#32CD32\",\n",
    "  \"#FFD700\", \"#20B2AA\", \"#FF6347\", \"#4169E1\", \"#8B4513\", \"#6A5ACD\", \"#FF7F50\", \"#2F4F4F\"\n",
    "]\n",
    "\n",
    "custom_colors = color_dict\n",
    "for n in n_neighbors:\n",
    "    for m in min_dist:\n",
    "        reducer = umap.UMAP(n_neighbors=n, min_dist=m, random_state=seed)\n",
    "        embedding = reducer.fit_transform(df.iloc[:, :-18])\n",
    "        UMAP_df  = pd.DataFrame({\n",
    "            'UMAP1': embedding[:, 0],\n",
    "            'UMAP2': embedding[:, 1],\n",
    "            'Domain to species': df['Domain to species'],\n",
    "            'Domain to genus': df['Domain to genus'],\n",
    "            'contig': df.index,\n",
    "            'element': df['element'],\n",
    "            'contig_length': df['contig_length']\n",
    "        })\n",
    "        # clustering k defined choosing the k by looking at the UMAP \n",
    "        kmeans = KMeans(n_clusters=11, random_state=seed)\n",
    "        \n",
    "        k_labels = kmeans.fit_predict(embedding)\n",
    "        \n",
    "        UMAP_df['cluster'] = k_labels\n",
    "        \n",
    "        column = 'Domain to species'\n",
    "        UMAP_df['Domain to species'] = UMAP_df['Domain to species'].fillna('NA')\n",
    "\n",
    "        value_counts = UMAP_df[column].value_counts().to_dict()\n",
    "\n",
    "        UMAP_df[f'{column}_with_counts'] = UMAP_df[column].apply(lambda x: f'{x} ({value_counts[x]})')\n",
    "\n",
    "        UMAP_df['count'] = UMAP_df[column].map(value_counts)\n",
    "        UMAP_df = UMAP_df.sort_values(by=['count'], ascending=False)\n",
    "\n",
    "        UMAP_df[f'{column}_with_counts'] = pd.Categorical(\n",
    "            UMAP_df[f'{column}_with_counts'],\n",
    "            categories=UMAP_df[f'{column}_with_counts'].unique(),\n",
    "            ordered=True\n",
    "        )\n",
    "\n",
    "        UMAP_df['sqrt_contig_length'] = np.sqrt(UMAP_df['contig_length'])\n",
    "        \n",
    "        fig = px.scatter(UMAP_df, \n",
    "                            x='UMAP1', \n",
    "                            y='UMAP2', \n",
    "                            labels={f'{column}_with_counts': column},\n",
    "                            color=f'{column}_with_counts', \n",
    "                            title=f' HAMBI - UMAP with n_neighbors={n}, min_dist={m}', \n",
    "                            color_discrete_sequence=custom_colors,\n",
    "                            hover_data={'cluster': True, 'Domain to species': True, 'Domain to genus': True,\n",
    "                                        'element': True, 'contig_length': True, 'contig': True},\n",
    "                        size='sqrt_contig_length'\n",
    "                        )\n",
    "        title = f' HAMBI - UMAP with n_neighbors={n}, min_dist={m}'\n",
    "        fig.update_layout(\n",
    "            height=1700,\n",
    "            width=1200,\n",
    "            title_text=title,\n",
    "            showlegend=True,\n",
    "            legend=dict(\n",
    "                x=0.5,\n",
    "                y=-0.1,\n",
    "                traceorder=\"normal\",\n",
    "                xanchor='center',\n",
    "                yanchor='top',\n",
    "                orientation='h'\n",
    "            ),\n",
    "            template='simple_white',\n",
    "            xaxis=dict(\n",
    "                showgrid=True,\n",
    "                gridcolor='lightgray',\n",
    "                zeroline=False,\n",
    "                showline=True,  \n",
    "                linecolor='black', \n",
    "                linewidth=1,\n",
    "                mirror=True\n",
    "            ),\n",
    "            yaxis=dict(\n",
    "                showgrid=True,\n",
    "                gridcolor='lightgray',\n",
    "                zeroline=False,\n",
    "                showline=True,\n",
    "                linecolor='black',\n",
    "                linewidth=1,\n",
    "                mirror=True\n",
    "            )\n",
    "        )\n",
    "        fig.show()\n",
    "        #fig.write_image(f'UMAP_HAMBI/UMAP_{n}_{m}.png')\n",
    "        #fig.write_html(f'UMAP_HAMBI/UMAP_{n}_{m}.html')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "40107e25-30a5-46d0-bd70-c70b86d4f498",
   "metadata": {},
   "source": [
    "### Plot without NA\n",
    "#### All data ðŸ’ "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0fc8596b-6e7d-4ada-8f0a-18785c368ec1",
   "metadata": {},
   "outputs": [],
   "source": [
    "df_no_NA = df[df['s'].notna()]\n",
    "print(df_no_NA)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "031d9200-085d-4efa-ab73-494b8b6863ec",
   "metadata": {},
   "outputs": [],
   "source": [
    "n_neighbors = [20]\n",
    "min_dist = [0.5]\n",
    "color_dict = [\n",
    "  \"#00FF00\", \"#0000FF\", \"#FFFF00\", \"#FF00FF\", \"#00FFFF\", \"#800000\", \"#008000\",\n",
    "  \"#000080\", \"#808000\", \"#800080\", \"#008080\", \"#FF4500\", \"#3bccf7\", \"#4682B4\", \"#DAA520\",\n",
    "  \"#FF1493\", \"#9ACD32\", \"#8A2BE2\", \"#5F9EA0\", \"#D2691E\", \"#DC143C\", \"#B8860B\", \"#32CD32\",\n",
    "  \"#FFD700\", \"#20B2AA\", \"#FF6347\", \"#4169E1\", \"#8B4513\", \"#6A5ACD\", \"#FF7F50\", \"#2F4F4F\"\n",
    "]\n",
    "\n",
    "custom_colors = color_dict\n",
    "for n in n_neighbors:\n",
    "    for m in min_dist:\n",
    "        reducer = umap.UMAP(n_neighbors=n, min_dist=m, random_state=seed)\n",
    "        embedding = reducer.fit_transform(df_no_NA.iloc[:, :-18])\n",
    "        UMAP_df_no_NA  = pd.DataFrame({\n",
    "            'UMAP1': embedding[:, 0],\n",
    "            'UMAP2': embedding[:, 1],\n",
    "            'Domain to species': df_no_NA['Domain to species'],\n",
    "            'Domain to genus': df_no_NA['Domain to genus'],\n",
    "            'element': df_no_NA['element'],\n",
    "            'contig_length': df_no_NA['contig_length']\n",
    "        })\n",
    "        # clustering k defined choosing the k by looking at the UMAP \n",
    "        kmeans = KMeans(n_clusters=11, random_state=seed)\n",
    "        \n",
    "        k_labels = kmeans.fit_predict(embedding)\n",
    "        \n",
    "        UMAP_df_no_NA['cluster'] = k_labels\n",
    "        \n",
    "        column = 'Domain to species'\n",
    "        UMAP_df_no_NA['Domain to species'] = UMAP_df_no_NA['Domain to species'].fillna('NA')\n",
    "\n",
    "        value_counts = UMAP_df_no_NA[column].value_counts().to_dict()\n",
    "\n",
    "        UMAP_df_no_NA[f'{column}_with_counts'] = UMAP_df_no_NA[column].apply(lambda x: f'{x} ({value_counts[x]})')\n",
    "\n",
    "        UMAP_df_no_NA['count'] = UMAP_df_no_NA[column].map(value_counts)\n",
    "        UMAP_df_no_NA = UMAP_df_no_NA.sort_values(by=['count'], ascending=False)\n",
    "\n",
    "        UMAP_df_no_NA[f'{column}_with_counts'] = pd.Categorical(\n",
    "            UMAP_df_no_NA[f'{column}_with_counts'],\n",
    "            categories=UMAP_df_no_NA[f'{column}_with_counts'].unique(),\n",
    "            ordered=True\n",
    "        )\n",
    "        UMAP_df_no_NA['sqrt_contig_length'] = np.sqrt(UMAP_df_no_NA['contig_length'])\n",
    "        fig = px.scatter(UMAP_df_no_NA, \n",
    "                            x='UMAP1', \n",
    "                            y='UMAP2', \n",
    "                            labels={f'{column}_with_counts': column},\n",
    "                            color=f'{column}_with_counts', \n",
    "                            title=f' HAMBI - UMAP with n_neighbors={n}, min_dist={m}', \n",
    "                            color_discrete_sequence=custom_colors,\n",
    "                            hover_data={'cluster': True, 'Domain to species': True, 'Domain to genus': True,\n",
    "                                        'element': True, 'contig_length': True},\n",
    "                        size='sqrt_contig_length')\n",
    "        title = f' HAMBI - UMAP with n_neighbors={n}, min_dist={m}'\n",
    "        fig.update_layout(\n",
    "            height=1700,\n",
    "            width=1200,\n",
    "            title_text=title,\n",
    "            showlegend=True,\n",
    "            legend=dict(\n",
    "                x=0.5,\n",
    "                y=-0.1,\n",
    "                traceorder=\"normal\",\n",
    "                xanchor='center',\n",
    "                yanchor='top',\n",
    "                orientation='h'\n",
    "            ),\n",
    "            template='simple_white',\n",
    "            xaxis=dict(\n",
    "                showgrid=True,\n",
    "                gridcolor='lightgray',\n",
    "                zeroline=False,\n",
    "                showline=True,  \n",
    "                linecolor='black', \n",
    "                linewidth=1,\n",
    "                mirror=True\n",
    "            ),\n",
    "            yaxis=dict(\n",
    "                showgrid=True,\n",
    "                gridcolor='lightgray',\n",
    "                zeroline=False,\n",
    "                showline=True,\n",
    "                linecolor='black',\n",
    "                linewidth=1,\n",
    "                mirror=True\n",
    "            )\n",
    "        )\n",
    "        fig.show()\n",
    "        fig.write_image(f'UMAP_HAMBI/UMAP_{n}_{m}_no_NA.png')\n",
    "        fig.write_html(f'UMAP_HAMBI/UMAP_{n}_{m}_no_NA.html')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "80d0453b-86bd-42f2-9936-8a48613625fb",
   "metadata": {
    "jp-MarkdownHeadingCollapsed": true
   },
   "source": [
    "### Plot k-mean clusters\n",
    "#### All data ðŸ’ "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ffced132-8998-4b89-8c9b-0adc094cccff",
   "metadata": {},
   "outputs": [],
   "source": [
    "custom_colors = ['#1f77b4', '#ff7f0e', '#2ca02c', '#d62728', \n",
    "                 '#9467bd', '#8c564b', '#e377c2', '#7f7f7f', \n",
    "                 '#bcbd22', '#17becf', '#ff8da1']\n",
    "\n",
    "# Convert 'kmeans' to a categorical variable with explicit order from 0 to 10\n",
    "UMAP_df = UMAP_df.sort_values(by=['cluster'], ascending=True)\n",
    "UMAP_df['clusters'] = pd.Categorical(UMAP_df['cluster'].astype(int), categories=range(0, 11), ordered=True)\n",
    "\n",
    "# Plot UMAP with the kmeans clusters\n",
    "fig = px.scatter(UMAP_df, \n",
    "                    x='UMAP1', \n",
    "                    y='UMAP2', \n",
    "                    labels={f'{column}_with_counts': column},\n",
    "                    color='clusters', \n",
    "                    title=f' HAMBI - UMAP with n_neighbors={n}, min_dist={m}', \n",
    "                    color_discrete_sequence=custom_colors,\n",
    "                    hover_data={'cluster': True, 'Domain to species': True, 'Domain to genus': True, 'element': True} )\n",
    "title = f' HAMBI - UMAP with n_neighbors={n}, min_dist={m} with kmeans clusters'\n",
    "\n",
    "# Update the layout of the figure\n",
    "fig.update_layout(\n",
    "    height=1300,\n",
    "    width=1200,\n",
    "    title_text=title,\n",
    "    showlegend=True,\n",
    "    legend=dict(\n",
    "        x=0.5,\n",
    "        y=-0.1,\n",
    "        traceorder=\"normal\",\n",
    "        xanchor='center',\n",
    "        yanchor='top',\n",
    "        orientation='h'\n",
    "    ),\n",
    "    template='simple_white',\n",
    "    xaxis=dict(\n",
    "        showgrid=True,\n",
    "        gridcolor='lightgray',\n",
    "        zeroline=False,\n",
    "        showline=True,  \n",
    "        linecolor='black', \n",
    "        linewidth=1,\n",
    "        mirror=True\n",
    "    ),\n",
    "    yaxis=dict(\n",
    "        showgrid=True,\n",
    "        gridcolor='lightgray',\n",
    "        zeroline=False,\n",
    "        showline=True,\n",
    "        linecolor='black',\n",
    "        linewidth=1,\n",
    "        mirror=True\n",
    "    )\n",
    ")\n",
    "\n",
    "# Show the figure\n",
    "fig.show()\n",
    "fig.write_image(f'UMAP_HAMBI/UMAP_{n}_{m}_kmeans.png')\n",
    "fig.write_html(f'UMAP_HAMBI/UMAP_{n}_{m}_kmeans.html')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "36dbafab-61bb-46d5-89e5-7540a4b3c197",
   "metadata": {},
   "outputs": [],
   "source": [
    "len(UMAP_df['Domain to species'].value_counts())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e2f416e2-1ec2-4da8-8014-4ac540367dbc",
   "metadata": {},
   "outputs": [],
   "source": [
    "species_cluster_count = UMAP_df.groupby(['Domain to species', 'cluster']).size().reset_index(name='Count')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b34b164a-fcdf-423f-a161-22edc7dbad82",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Pivot the table to make species as rows and clusters as columns\n",
    "species_cluster_counts = UMAP_df.groupby(['Domain to species', 'cluster']).size().reset_index(name='Count')\n",
    "\n",
    "# Pivot the table to make species as rows and clusters as columns\n",
    "species_cluster_pivot = species_cluster_counts.pivot(index='Domain to species', columns='cluster', values='Count').fillna(0)\n",
    "\n",
    "# Convert all columns to numeric to ensure proper summation\n",
    "species_cluster_pivot = species_cluster_pivot.apply(pd.to_numeric)\n",
    "\n",
    "# Calculate the row sum (total contigs per species) and add it as a new column\n",
    "species_cluster_pivot['SUM'] = species_cluster_pivot.sum(axis=1)\n",
    "species_cluster_pivot.sort_values(by='SUM', ascending=False, inplace =True)\n",
    "species_cluster_pivot"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a54d161b-0e1f-420c-a0f1-d52af0c817b1",
   "metadata": {
    "jp-MarkdownHeadingCollapsed": true
   },
   "source": [
    "### Save to file\n",
    "#### All data ðŸ’ "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8535156d-4658-4b11-b00e-09a2ad452bff",
   "metadata": {},
   "outputs": [],
   "source": [
    "species_cluster_pivot.to_csv('UMAP_HAMBI/HAMBI_species_cluster_pivot.csv', index=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c225668e-89e5-4807-81ed-9a68ba45bc73",
   "metadata": {},
   "outputs": [],
   "source": [
    "def calculate_shannon_entropy(pivot_table):\n",
    "    cluster_columns = pivot_table.columns[:-1]  \n",
    "    total_samples = pivot_table['SUM'].sum() \n",
    "\n",
    "    total_weighted_entropy = 0\n",
    "\n",
    "    for cluster in cluster_columns:\n",
    "        total_in_cluster = pivot_table[cluster].sum()\n",
    "        \n",
    "        if total_in_cluster == 0:\n",
    "            continue  \n",
    "        \n",
    "        # the proportion of each species in this cluster\n",
    "        proportions = pivot_table[cluster] / total_in_cluster\n",
    "        \n",
    "        # the entropy for this cluster\n",
    "        entropy = -np.sum(proportions * np.log2(proportions + 1e-10))  # Small value to avoid log(0)\n",
    "        \n",
    "        # Weight the entropy by the proportion of the cluster size relative to the total\n",
    "        weighted_entropy = (total_in_cluster / total_samples) * entropy\n",
    "        total_weighted_entropy += weighted_entropy\n",
    "\n",
    "        print(f\"Entropy for cluster {cluster}: {entropy:.4f}\")\n",
    "    \n",
    "    print(f\"Total Weighted Entropy: {total_weighted_entropy:.4f}\")\n",
    "    return total_weighted_entropy\n",
    "\n",
    "calculate_shannon_entropy(species_cluster_pivot)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "00605bc0-6626-459e-801f-b4b2e73215f8",
   "metadata": {
    "jp-MarkdownHeadingCollapsed": true
   },
   "source": [
    "### Connecting clusters to taxa\n",
    "#### All data ðŸ’ "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7af5aa68-358c-4308-b458-3a15c77300d6",
   "metadata": {},
   "outputs": [],
   "source": [
    "def return_index(cluster:int, df:pd.DataFrame) -> tuple:\n",
    "    \"\"\"Returns the index of the cluster in the dataframe\n",
    "    \"\"\"\n",
    "    return df[df['cluster'] == cluster]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0b8b74b2-81c3-47eb-aafd-f39dc87233e4",
   "metadata": {},
   "outputs": [],
   "source": [
    "UMAP_df['cluster'].value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "efa9e1ac-9584-4ed8-ad53-cc5813619f0b",
   "metadata": {},
   "outputs": [],
   "source": [
    "cluster_data = {}\n",
    "for cluster in UMAP_df['cluster'].unique():\n",
    "    cluster_data[cluster] = UMAP_df[UMAP_df['cluster'] == cluster].index.tolist()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "056ed069-d284-4493-8d79-a2daf93079c1",
   "metadata": {},
   "outputs": [],
   "source": [
    "cluster_data.keys()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4fdab501-4ffd-476a-8951-528f5c043a78",
   "metadata": {
    "jp-MarkdownHeadingCollapsed": true
   },
   "source": [
    "### Plot the number of each species in each cluster\n",
    "#### All data ðŸ’ "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0f8ad5fe-03d4-485f-bbe2-598dcbd46ca5",
   "metadata": {},
   "outputs": [],
   "source": [
    "for cluster in cluster_data.keys():\n",
    "    # Select the subset of UMAP_df for the current cluster\n",
    "    cluster_df = UMAP_df.loc[cluster_data[cluster]]\n",
    "    \n",
    "    plt.figure(figsize=(20, 10))\n",
    "    sns.countplot(x='Domain to species', data=cluster_df, order=cluster_df['Domain to species'].value_counts().index)\n",
    "    plt.xticks(rotation=90)\n",
    "    plt.title(f'Cluster {cluster}')\n",
    "    plt.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "47f7113c-b9c8-44d7-8ea4-2df17214c2ff",
   "metadata": {
    "jp-MarkdownHeadingCollapsed": true
   },
   "source": [
    "### Plot to how many clusters one species has been assigned by the clustering algoritms\n",
    "#### All data ðŸ’ "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "75601958-659e-4a72-9a72-de62476e39e7",
   "metadata": {},
   "outputs": [],
   "source": [
    "species_kmeans_counts = UMAP_df.groupby('Domain to species')['cluster'].nunique()\n",
    "\n",
    "species_kmeans_counts.plot(kind='bar', color='lightgreen')\n",
    "plt.title('KMEANS')\n",
    "plt.xlabel('Species')\n",
    "plt.ylabel('Number of Unique KMeans Clusters')\n",
    "\n",
    "plt.xticks(rotation=90)\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0eb3a7d2-7b62-413e-a913-a93ea1645d94",
   "metadata": {
    "jp-MarkdownHeadingCollapsed": true
   },
   "source": [
    "### Create concatenated images of the sequencing logos for each cluster and species\n",
    "#### All data ðŸ’ "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a418e8f3-98d4-43cf-bac0-8fe21c07e3c4",
   "metadata": {},
   "outputs": [],
   "source": [
    "def search_images_in_subfolders(path_to_folders, index, image_types):\n",
    "    \"\"\"Search for images by index and type from subfolders, adding placeholders if necessary.\"\"\"\n",
    "    images = {}\n",
    "    \n",
    "    subfolders = [f for f in os.listdir(path_to_folders) if os.path.isdir(os.path.join(path_to_folders, f))]\n",
    "    for img_type in image_types:\n",
    "        image_found = False\n",
    "        for subfolder in subfolders:\n",
    "            image_path = os.path.join(path_to_folders, subfolder, 'logos', f\"{index}_{img_type}.png\")            \n",
    "            if os.path.exists(image_path):\n",
    "                try:\n",
    "                    images[img_type] = Image.open(image_path)\n",
    "                    image_found = True\n",
    "                    break\n",
    "                except Exception as e:\n",
    "                    print(f\"Failed to open image {image_path}: {e}\")\n",
    "                    images[img_type] = Image.new('RGB', (1200, 300), 'white')\n",
    "                    image_found = True\n",
    "                    break\n",
    "        if not image_found:\n",
    "            images[img_type] = Image.new('RGB', (1200, 300), 'white')\n",
    "            print(f\"No image found for {img_type}, using placeholder.\")\n",
    "    \n",
    "    return images\n",
    "\n",
    "def standardize_image_sizes(images, size=(1200, 300)):\n",
    "    \"\"\"Resize all images to a standard size.\"\"\"\n",
    "    resized_images = {}\n",
    "    for key, img in images.items():\n",
    "        resized_images[key] = img.resize(size)\n",
    "    return resized_images\n",
    "\n",
    "def combine_images_horizontally(images):\n",
    "    \"\"\"Combine images horizontally.\"\"\"\n",
    "    if not images:\n",
    "        return None\n",
    "    images = standardize_image_sizes(images)\n",
    "    \n",
    "    total_width = sum(img.width for img in images.values())\n",
    "    max_height = max(img.height for img in images.values())\n",
    "\n",
    "    new_im = Image.new('RGB', (total_width, max_height), 'white')\n",
    "    x_offset = 0\n",
    "    for img_type, img in images.items():\n",
    "        new_im.paste(img, (x_offset, 0))\n",
    "        x_offset += img.width\n",
    "    return new_im\n",
    "\n",
    "def combine_images_vertically(image_list):\n",
    "    \"\"\"Combine images vertically to form a single cluster image.\"\"\"\n",
    "    if not image_list or all(img is None for img in image_list):\n",
    "        return None\n",
    "    max_width = max(img.width for img in image_list)\n",
    "    total_height = sum(img.height for img in image_list)\n",
    "    \n",
    "    new_im = Image.new('RGB', (max_width, total_height), 'white')\n",
    "    y_offset = 0\n",
    "    for img in image_list:\n",
    "        new_im.paste(img, (0, y_offset))\n",
    "        y_offset += img.height\n",
    "    return new_im\n",
    "\n",
    "\n",
    "def display_and_save_image(image, title, save_path=None):\n",
    "    \"\"\"Display an image using matplotlib and optionally save it to a file.\"\"\"\n",
    "    if image is None:\n",
    "        print(\"No image to display or save.\")\n",
    "        return\n",
    "    \n",
    "    plt.figure(figsize=(10, 5))\n",
    "    plt.imshow(image)\n",
    "    plt.title(title)\n",
    "    plt.axis('off')\n",
    "\n",
    "    if save_path:\n",
    "        try:\n",
    "            if image.width == 0 or image.height == 0:\n",
    "                print(f\"Image {title} has invalid dimensions and cannot be saved.\")\n",
    "            else:\n",
    "                image.save(save_path)\n",
    "                print(f\"Saved image to {save_path}\")\n",
    "        except Exception as e:\n",
    "            print(f\"Failed to save the image {title} at {save_path}: {e}\")\n",
    "    plt.close()\n",
    "\n",
    "def process_clusters_and_species_images(UMAP_df, path_to_folders, image_types, outputpath):\n",
    "    \"\"\"Process images based on K-means clusters and species in the DataFrame.\"\"\"\n",
    "    cluster_col = 'cluster'  # Use only the K-means clustering column\n",
    "\n",
    "    for cluster in UMAP_df[cluster_col].unique():\n",
    "        print(f\"Processing K-means cluster {cluster}\")\n",
    "        cluster_df = UMAP_df[UMAP_df[cluster_col] == cluster]\n",
    "\n",
    "        for species in cluster_df['Domain to species'].unique():\n",
    "            print(f\"Processing species {species} in K-means cluster {cluster}\")\n",
    "            species_images = []\n",
    "            indexes = cluster_df[cluster_df['Domain to species'] == species].index\n",
    "\n",
    "            for index in indexes:\n",
    "                images = search_images_in_subfolders(path_to_folders, index, image_types)\n",
    "                combined_horizontal_image = combine_images_horizontally(images)\n",
    "                if combined_horizontal_image:\n",
    "                    species_images.append(combined_horizontal_image)\n",
    "\n",
    "            final_image = combine_images_vertically(species_images)\n",
    "\n",
    "            if final_image:\n",
    "                filename = f\"kmeans_cluster_{cluster}_{species}.png\"\n",
    "                display_path = os.path.join(outputpath, filename)\n",
    "                display_and_save_image(final_image, f\"K-means Cluster {cluster} - {species}\", display_path)\n",
    "    print(\"Done\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3be9fbb5-c2b0-479e-a1ae-b3becd8e849c",
   "metadata": {},
   "outputs": [],
   "source": [
    "path_to_folders = '/scratch/project_2006608/Methylation/HAMBI_data/'\n",
    "image_types = ['m4C_logo', 'm6A_logo', 'modified_base_logo']\n",
    "outputpath = '/scratch/project_2006608/Methylation/notebooks/logos_HAMBI/'\n",
    "\n",
    "process_clusters_and_species_images(UMAP_df, path_to_folders, image_types, outputpath)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c649add3-5c92-47a0-b3e2-1dbecbc5ac11",
   "metadata": {},
   "outputs": [],
   "source": [
    "species_kmeans_counts = filt_UMAP_df.groupby('Domain to species')['cluster'].nunique()\n",
    "\n",
    "species_kmeans_counts.plot(kind='bar', color='lightgreen')\n",
    "plt.title('KMEANS')\n",
    "plt.xlabel('Species')\n",
    "plt.ylabel('Number of Unique KMeans Clusters')\n",
    "\n",
    "plt.xticks(rotation=90)\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "983a80c8-77f7-4e77-a2e8-138333852b68",
   "metadata": {
    "jp-MarkdownHeadingCollapsed": true
   },
   "source": [
    "## Trying shorter range of the sequence\n",
    "### Less noise ðŸ”·\n",
    "#### -10...0...10\n",
    "#### -5...0...5"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d089110d-05a5-466c-8dfc-59f7e943ea0b",
   "metadata": {},
   "outputs": [],
   "source": [
    "path_to_images = '/scratch/project_2006608/Methylation/notebooks/UMAP_HAMBI_short/'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "63ed6a9c-5228-4676-a20f-426c29026c58",
   "metadata": {},
   "outputs": [],
   "source": [
    "file_path_10 = '/scratch/project_2006608/Methylation/HAMBI_data/merged_data_short_10.tsv'\n",
    "df_10 = pd.read_csv(file_path_10, sep='\\t', index_col=0, low_memory=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d471ddce-e2f6-42c1-9b37-2fe23c80924b",
   "metadata": {},
   "outputs": [],
   "source": [
    "file_path_5 = '/scratch/project_2006608/Methylation/HAMBI_data/merged_data_short_5.tsv'\n",
    "df_5 = pd.read_csv(file_path_5, sep='\\t', index_col=0, low_memory=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0f87d10f-f1de-4982-b53b-1d1807bb85cd",
   "metadata": {},
   "outputs": [],
   "source": [
    "print(df_10.shape)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7e47ca54-0eba-4543-a0d0-5809866ea3cc",
   "metadata": {},
   "outputs": [],
   "source": [
    "print(df_5.shape)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "664c9f83-c645-453b-87ac-8878394fea19",
   "metadata": {},
   "source": [
    "### FILTERING\n",
    "#### Removing the contigs that have no methyaltions (or under 20) detected for all of the three types. These are the rows where all values in the columns are 0s\n",
    "#### Less noise ðŸ”·"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f605be8a-aaa3-49d5-b1fd-57df3621151a",
   "metadata": {},
   "outputs": [],
   "source": [
    "# remove rows where df.iloc[:, :-17] meaning no methylation detected or less than 20 detected since those are converted to 0 filled matrices\n",
    "df_10 = df_10.loc[(df_10.iloc[:, :-17] != 0).any(axis=1)]\n",
    "df_5 = df_5.loc[(df_5.iloc[:, :-17] != 0).any(axis=1)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c1d860f1-4423-41e1-bfe9-9b7c01900686",
   "metadata": {},
   "outputs": [],
   "source": [
    "print(df_10.shape[0])\n",
    "print(df_5.shape[0])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3ba265eb-7fba-4838-b505-0c30512b9c5a",
   "metadata": {},
   "outputs": [],
   "source": [
    "df_10['Domain to species'].value_counts()\n",
    "df_5['Domain to species'].value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "83083b9b-5817-424b-82aa-4df941d1e7bf",
   "metadata": {},
   "outputs": [],
   "source": [
    "df_10['Domain to species'].nunique(dropna = False)\n",
    "df_5['Domain to species'].nunique(dropna = False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e2d3f478-f615-4b0d-b92b-8a30e1e618f9",
   "metadata": {},
   "outputs": [],
   "source": [
    "len(df_10[df_10['Domain to species'].isna()])\n",
    "len(df_5[df_5['Domain to species'].isna()])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "570c4c63-d6f3-428c-a5ef-55554d12688a",
   "metadata": {},
   "outputs": [],
   "source": [
    "df_10.head()\n",
    "df_5.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a35f8fb1-0880-4b8d-8b1a-94de456a239f",
   "metadata": {},
   "source": [
    "### Draw UMAP\n",
    "#### Less noise ðŸ”·\n",
    "#### -10...0...10"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6dc4b645-8b19-4b90-a875-723a3e0ec0e2",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "n_neighbors = [20, 30 ]\n",
    "min_dist = [0.01, 0.1, 0.2]\n",
    "colors = sns.color_palette(\"hls\", 32).as_hex()\n",
    "\n",
    "unique_species = df_10['Domain to species'].unique()\n",
    "num_species = len(unique_species)\n",
    "color_dict = [\n",
    "    \"#FF0000\", \"#00FF00\", \"#0000FF\", \"#FFFF00\", \"#FF00FF\", \"#00FFFF\", \"#800000\", \"#008000\",\n",
    "    \"#000080\", \"#808000\", \"#800080\", \"#008080\", \"#FF4500\", \"#2E8B57\", \"#4682B4\", \"#DAA520\",\n",
    "    \"#FF1493\", \"#7FFF00\", \"#8A2BE2\", \"#5F9EA0\", \"#D2691E\", \"#DC143C\", \"#B8860B\", \"#32CD32\",\n",
    "    \"#FFD700\", \"#20B2AA\", \"#FF6347\", \"#4169E1\", \"#8B4513\", \"#6A5ACD\", \"#FF7F50\", \"#2F4F4F\"\n",
    "]\n",
    "custom_colors = color_dict\n",
    "\n",
    "\n",
    "for n in n_neighbors:\n",
    "    for m in min_dist:\n",
    "        \n",
    "        reducer = umap.UMAP(n_neighbors=n, min_dist=m, random_state=seed)\n",
    "        embedding = reducer.fit_transform(df_10.iloc[:, :-17])\n",
    "        UMAP_df_10  = pd.DataFrame({\n",
    "            'UMAP1': embedding[:, 0],\n",
    "            'UMAP2': embedding[:, 1],\n",
    "            'Domain to species': df_10['Domain to species'],\n",
    "            'Domain to genus': df_10['Domain to genus'],\n",
    "            'element': df_10['element']\n",
    "        })\n",
    "        # clustering\n",
    "        kmeans = KMeans(n_clusters=11, random_state=seed)\n",
    "        \n",
    "        k_labels = kmeans.fit_predict(embedding)\n",
    "        \n",
    "        UMAP_df_10['cluster'] = k_labels\n",
    "        UMAP_df_10['Domain to species'] = UMAP_df_10['Domain to species'].fillna('NA')\n",
    "        column = 'Domain to species'\n",
    "        value_counts = UMAP_df_10[column].value_counts().to_dict()\n",
    "\n",
    "        UMAP_df_10[f'{column}_with_counts'] = UMAP_df_10[column].apply(lambda x: f'{x} ({value_counts[x]})')\n",
    "\n",
    "        UMAP_df_10['count'] = UMAP_df_10[column].map(value_counts)\n",
    "        UMAP_df_10 = UMAP_df_10.sort_values(by=['count'], ascending=False)\n",
    "\n",
    "        UMAP_df_10[f'{column}_with_counts'] = pd.Categorical(\n",
    "            UMAP_df_10[f'{column}_with_counts'],\n",
    "            categories=UMAP_df_10[f'{column}_with_counts'].unique(),\n",
    "            ordered=True\n",
    "        )\n",
    "        \n",
    "        fig = px.scatter(UMAP_df_10, \n",
    "                            x='UMAP1', \n",
    "                            y='UMAP2', \n",
    "                            labels={f'{column}_with_counts': column},\n",
    "                            color=f'{column}_with_counts', \n",
    "                            title=f' HAMBI - UMAP with n_neighbors={n}, min_dist={m}', \n",
    "                            color_discrete_sequence=custom_colors,\n",
    "                            hover_data={'cluster': True, 'Domain to species': True, 'Domain to genus': True, 'element': True} )\n",
    "        title = f' HAMBI - UMAP with n_neighbors={n}, min_dist={m}'\n",
    "        fig.update_layout(\n",
    "            height=1500,\n",
    "            width=1200,\n",
    "            title_text=title,\n",
    "            showlegend=True,\n",
    "            legend=dict(\n",
    "                x=0.5,\n",
    "                y=-0.1,\n",
    "                traceorder=\"normal\",\n",
    "                xanchor='center',\n",
    "                yanchor='top',\n",
    "                orientation='h'\n",
    "            ),\n",
    "            template='simple_white',\n",
    "            xaxis=dict(\n",
    "                showgrid=True,\n",
    "                gridcolor='lightgray',\n",
    "                zeroline=False,\n",
    "                showline=True,  \n",
    "                linecolor='black', \n",
    "                linewidth=1,\n",
    "                mirror=True\n",
    "            ),\n",
    "            yaxis=dict(\n",
    "                showgrid=True,\n",
    "                gridcolor='lightgray',\n",
    "                zeroline=False,\n",
    "                showline=True,\n",
    "                linecolor='black',\n",
    "                linewidth=1,\n",
    "                mirror=True\n",
    "            )\n",
    "        )\n",
    "        fig.show()\n",
    "        #fig.write_image(f'UMAP_HAMBI_short/UMAP_{n}_{m}.png')\n",
    "        #fig.write_html(f'UMAP_HAMBI_short/UMAP_{n}_{m}.html')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "478ebffa-3285-4644-ac2e-6f05e7040e6f",
   "metadata": {},
   "source": [
    "### Using the best n_neighbours and min_dist to get clear clusters\n",
    "#### Less noise ðŸ”·\n",
    "#### -10...0...10"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4bb6075a-df2c-41d0-b0ea-4e3b05a190a5",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "n_neighbors = [30]\n",
    "min_dist = [0.01]\n",
    "color_dict = [\n",
    "    \"#FF0000\", \"#00FF00\", \"#0000FF\", \"#FFFF00\", \"#FF00FF\", \"#00FFFF\", \"#800000\", \"#008000\",\n",
    "    \"#000080\", \"#808000\", \"#800080\", \"#008080\", \"#FF4500\", \"#2E8B57\", \"#4682B4\", \"#DAA520\",\n",
    "    \"#FF1493\", \"#9ACD32\", \"#8A2BE2\", \"#5F9EA0\", \"#D2691E\", \"#DC143C\", \"#B8860B\", \"#32CD32\",\n",
    "    \"#FFD700\", \"#20B2AA\", \"#FF6347\", \"#4169E1\", \"#8B4513\", \"#6A5ACD\", \"#FF7F50\", \"#2F4F4F\"\n",
    "]\n",
    "\n",
    "\n",
    "custom_colors = color_dict\n",
    "for n in n_neighbors:\n",
    "    for m in min_dist:\n",
    "        reducer = umap.UMAP(n_neighbors=n, min_dist=m, random_state=seed)\n",
    "        embedding = reducer.fit_transform(df_10.iloc[:, :-17])\n",
    "        UMAP_df_10  = pd.DataFrame({\n",
    "            'UMAP1': embedding[:, 0],\n",
    "            'UMAP2': embedding[:, 1],\n",
    "            'Domain to species': UMAP_df_10['Domain to species'],\n",
    "            'Domain to genus': UMAP_df_10['Domain to genus'],\n",
    "            'element': UMAP_df_10['element']\n",
    "        })\n",
    "        # clustering k defined choosing the k by looking at the UMAP \n",
    "        kmeans = KMeans(n_clusters=11, random_state=seed)\n",
    "        \n",
    "        k_labels = kmeans.fit_predict(embedding)\n",
    "        \n",
    "        UMAP_df_10['cluster'] = k_labels\n",
    "        \n",
    "        column = 'Domain to species'\n",
    "        UMAP_df_10['Domain to species'] = UMAP_df_10['Domain to species'].fillna('NA')\n",
    "\n",
    "        value_counts = UMAP_df_10[column].value_counts().to_dict()\n",
    "\n",
    "        UMAP_df_10[f'{column}_with_counts'] = UMAP_df_10[column].apply(lambda x: f'{x} ({value_counts[x]})')\n",
    "\n",
    "        UMAP_df_10['count'] = UMAP_df_10[column].map(value_counts)\n",
    "        UMAP_df_10 = UMAP_df_10.sort_values(by=['count'], ascending=False)\n",
    "\n",
    "        UMAP_df_10[f'{column}_with_counts'] = pd.Categorical(\n",
    "            UMAP_df_10[f'{column}_with_counts'],\n",
    "            categories=UMAP_df_10[f'{column}_with_counts'].unique(),\n",
    "            ordered=True\n",
    "        )\n",
    "        \n",
    "        fig = px.scatter(UMAP_df_10, \n",
    "                            x='UMAP1', \n",
    "                            y='UMAP2', \n",
    "                            labels={f'{column}_with_counts': column},\n",
    "                            color=f'{column}_with_counts', \n",
    "                            title=f' HAMBI - UMAP with n_neighbors={n}, min_dist={m}', \n",
    "                            color_discrete_sequence=custom_colors,\n",
    "                            hover_data={'cluster': True, 'Domain to species': True, 'Domain to genus': True, 'element': True} )\n",
    "        title = f' HAMBI - UMAP with n_neighbors={n}, min_dist={m}'\n",
    "        fig.update_layout(\n",
    "            height=1700,\n",
    "            width=1200,\n",
    "            title_text=title,\n",
    "            showlegend=True,\n",
    "            legend=dict(\n",
    "                x=0.5,\n",
    "                y=-0.1,\n",
    "                traceorder=\"normal\",\n",
    "                xanchor='center',\n",
    "                yanchor='top',\n",
    "                orientation='h'\n",
    "            ),\n",
    "            template='simple_white',\n",
    "            xaxis=dict(\n",
    "                showgrid=True,\n",
    "                gridcolor='lightgray',\n",
    "                zeroline=False,\n",
    "                showline=True,  \n",
    "                linecolor='black', \n",
    "                linewidth=1,\n",
    "                mirror=True\n",
    "            ),\n",
    "            yaxis=dict(\n",
    "                showgrid=True,\n",
    "                gridcolor='lightgray',\n",
    "                zeroline=False,\n",
    "                showline=True,\n",
    "                linecolor='black',\n",
    "                linewidth=1,\n",
    "                mirror=True\n",
    "            )\n",
    "        )\n",
    "        fig.show()\n",
    "        fig.write_image(f'UMAP_HAMBI_short/UMAP_{n}_{m}_short_10.png')\n",
    "        fig.write_html(f'UMAP_HAMBI_short/UMAP_{n}_{m}_short_10.html')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3f7e7ef4-587a-4245-b37d-bea3186fd7d4",
   "metadata": {},
   "source": [
    "### Draw UMAP\n",
    "#### Less noise ðŸ”·\n",
    "#### -5...0...5"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ca391858-777f-454b-9fd3-21724d77e1e9",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "n_neighbors = [20, 30 ]\n",
    "min_dist = [0.01, 0.1, 0.2]\n",
    "\n",
    "colors = sns.color_palette(\"hls\", 32).as_hex()\n",
    "\n",
    "unique_species = df_5['Domain to species'].unique()\n",
    "num_species = len(unique_species)\n",
    "color_dict = [\n",
    "    \"#FF0000\", \"#00FF00\", \"#0000FF\", \"#FFFF00\", \"#FF00FF\", \"#00FFFF\", \"#800000\", \"#008000\",\n",
    "    \"#000080\", \"#808000\", \"#800080\", \"#008080\", \"#FF4500\", \"#2E8B57\", \"#4682B4\", \"#DAA520\",\n",
    "    \"#FF1493\", \"#7FFF00\", \"#8A2BE2\", \"#5F9EA0\", \"#D2691E\", \"#DC143C\", \"#B8860B\", \"#32CD32\",\n",
    "    \"#FFD700\", \"#20B2AA\", \"#FF6347\", \"#4169E1\", \"#8B4513\", \"#6A5ACD\", \"#FF7F50\", \"#2F4F4F\"\n",
    "]\n",
    "custom_colors = color_dict\n",
    "\n",
    "\n",
    "for n in n_neighbors:\n",
    "    for m in min_dist:\n",
    "        \n",
    "        reducer = umap.UMAP(n_neighbors=n, min_dist=m, random_state=seed)\n",
    "        embedding = reducer.fit_transform(df_5.iloc[:, :-17])\n",
    "        UMAP_df_5  = pd.DataFrame({\n",
    "            'UMAP1': embedding[:, 0],\n",
    "            'UMAP2': embedding[:, 1],\n",
    "            'Domain to species': df_5['Domain to species'],\n",
    "            'Domain to genus': df_5['Domain to genus'],\n",
    "            'element': df_5['element']\n",
    "        })\n",
    "        # clustering\n",
    "        kmeans = KMeans(n_clusters=11, random_state=seed)\n",
    "        \n",
    "        k_labels = kmeans.fit_predict(embedding)\n",
    "        \n",
    "        UMAP_df_5['cluster'] = k_labels\n",
    "        UMAP_df_5['Domain to species'] = UMAP_df_5['Domain to species'].fillna('NA')\n",
    "        column = 'Domain to species'\n",
    "        value_counts = UMAP_df_5[column].value_counts().to_dict()\n",
    "\n",
    "        UMAP_df_5[f'{column}_with_counts'] = UMAP_df_5[column].apply(lambda x: f'{x} ({value_counts[x]})')\n",
    "\n",
    "        UMAP_df_5['count'] = UMAP_df_5[column].map(value_counts)\n",
    "        UMAP_df_5 = UMAP_df_5.sort_values(by=['count'], ascending=False)\n",
    "\n",
    "        UMAP_df_5[f'{column}_with_counts'] = pd.Categorical(\n",
    "            UMAP_df_5[f'{column}_with_counts'],\n",
    "            categories=UMAP_df_5[f'{column}_with_counts'].unique(),\n",
    "            ordered=True\n",
    "        )\n",
    "        \n",
    "        fig = px.scatter(UMAP_df_5, \n",
    "                            x='UMAP1', \n",
    "                            y='UMAP2', \n",
    "                            labels={f'{column}_with_counts': column},\n",
    "                            color=f'{column}_with_counts', \n",
    "                            title=f' HAMBI - UMAP with n_neighbors={n}, min_dist={m}', \n",
    "                            color_discrete_sequence=custom_colors,\n",
    "                            hover_data={'cluster': True, 'Domain to species': True, 'Domain to genus': True, 'element': True} )\n",
    "        title = f' HAMBI - UMAP with n_neighbors={n}, min_dist={m}'\n",
    "        fig.update_layout(\n",
    "            height=1500,\n",
    "            width=1200,\n",
    "            title_text=title,\n",
    "            showlegend=True,\n",
    "            legend=dict(\n",
    "                x=0.5,\n",
    "                y=-0.1,\n",
    "                traceorder=\"normal\",\n",
    "                xanchor='center',\n",
    "                yanchor='top',\n",
    "                orientation='h'\n",
    "            ),\n",
    "            template='simple_white',\n",
    "            xaxis=dict(\n",
    "                showgrid=True,\n",
    "                gridcolor='lightgray',\n",
    "                zeroline=False,\n",
    "                showline=True,  \n",
    "                linecolor='black', \n",
    "                linewidth=1,\n",
    "                mirror=True\n",
    "            ),\n",
    "            yaxis=dict(\n",
    "                showgrid=True,\n",
    "                gridcolor='lightgray',\n",
    "                zeroline=False,\n",
    "                showline=True,\n",
    "                linecolor='black',\n",
    "                linewidth=1,\n",
    "                mirror=True\n",
    "            )\n",
    "        )\n",
    "        fig.show()\n",
    "        #fig.write_image(f'UMAP_HAMBI_short/UMAP_{n}_{m}.png')\n",
    "        #fig.write_html(f'UMAP_HAMBI_short/UMAP_{n}_{m}.html')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "95e34b59-fc9d-46aa-bfda-d2a15045e1f7",
   "metadata": {},
   "source": [
    "### Using the best n_neighbours and min_dist to get clear clusters\n",
    "#### Less noise ðŸ”·\n",
    "#### -5...0...5"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3a65120b-dfc0-4a28-9134-af4907c4d318",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "n_neighbors = [30]\n",
    "min_dist = [0.01]\n",
    "color_dict = [\n",
    "    \"#FF0000\", \"#00FF00\", \"#0000FF\", \"#FFFF00\", \"#FF00FF\", \"#00FFFF\", \"#800000\", \"#008000\",\n",
    "    \"#000080\", \"#808000\", \"#800080\", \"#008080\", \"#FF4500\", \"#2E8B57\", \"#4682B4\", \"#DAA520\",\n",
    "    \"#FF1493\", \"#9ACD32\", \"#8A2BE2\", \"#5F9EA0\", \"#D2691E\", \"#DC143C\", \"#B8860B\", \"#32CD32\",\n",
    "    \"#FFD700\", \"#20B2AA\", \"#FF6347\", \"#4169E1\", \"#8B4513\", \"#6A5ACD\", \"#FF7F50\", \"#2F4F4F\"\n",
    "]\n",
    "\n",
    "\n",
    "custom_colors = color_dict\n",
    "for n in n_neighbors:\n",
    "    for m in min_dist:\n",
    "        reducer = umap.UMAP(n_neighbors=n, min_dist=m, random_state=seed)\n",
    "        embedding = reducer.fit_transform(df_5.iloc[:, :-17])\n",
    "        UMAP_df_5  = pd.DataFrame({\n",
    "            'UMAP1': embedding[:, 0],\n",
    "            'UMAP2': embedding[:, 1],\n",
    "            'Domain to species': UMAP_df_5['Domain to species'],\n",
    "            'Domain to genus': UMAP_df_5['Domain to genus'],\n",
    "            'element': UMAP_df_5['element']\n",
    "        })\n",
    "        # clustering k defined choosing the k by looking at the UMAP \n",
    "        kmeans = KMeans(n_clusters=11, random_state=seed)\n",
    "        \n",
    "        k_labels = kmeans.fit_predict(embedding)\n",
    "        \n",
    "        UMAP_df_5['cluster'] = k_labels\n",
    "        \n",
    "        column = 'Domain to species'\n",
    "        UMAP_df_5['Domain to species'] = UMAP_df_5['Domain to species'].fillna('NA')\n",
    "\n",
    "        value_counts = UMAP_df_5[column].value_counts().to_dict()\n",
    "\n",
    "        UMAP_df_5[f'{column}_with_counts'] = UMAP_df_5[column].apply(lambda x: f'{x} ({value_counts[x]})')\n",
    "\n",
    "        UMAP_df_5['count'] = UMAP_df_5[column].map(value_counts)\n",
    "        UMAP_df_5 = UMAP_df_5.sort_values(by=['count'], ascending=False)\n",
    "\n",
    "        UMAP_df_5[f'{column}_with_counts'] = pd.Categorical(\n",
    "            UMAP_df_5[f'{column}_with_counts'],\n",
    "            categories=UMAP_df_5[f'{column}_with_counts'].unique(),\n",
    "            ordered=True\n",
    "        )\n",
    "        \n",
    "        fig = px.scatter(UMAP_df_5, \n",
    "                            x='UMAP1', \n",
    "                            y='UMAP2', \n",
    "                            labels={f'{column}_with_counts': column},\n",
    "                            color=f'{column}_with_counts', \n",
    "                            title=f' HAMBI - UMAP with n_neighbors={n}, min_dist={m}', \n",
    "                            color_discrete_sequence=custom_colors,\n",
    "                            hover_data={'cluster': True, 'Domain to species': True, 'Domain to genus': True, 'element': True} )\n",
    "        title = f' HAMBI - UMAP with n_neighbors={n}, min_dist={m}'\n",
    "        fig.update_layout(\n",
    "            height=1700,\n",
    "            width=1200,\n",
    "            title_text=title,\n",
    "            showlegend=True,\n",
    "            legend=dict(\n",
    "                x=0.5,\n",
    "                y=-0.1,\n",
    "                traceorder=\"normal\",\n",
    "                xanchor='center',\n",
    "                yanchor='top',\n",
    "                orientation='h'\n",
    "            ),\n",
    "            template='simple_white',\n",
    "            xaxis=dict(\n",
    "                showgrid=True,\n",
    "                gridcolor='lightgray',\n",
    "                zeroline=False,\n",
    "                showline=True,  \n",
    "                linecolor='black', \n",
    "                linewidth=1,\n",
    "                mirror=True\n",
    "            ),\n",
    "            yaxis=dict(\n",
    "                showgrid=True,\n",
    "                gridcolor='lightgray',\n",
    "                zeroline=False,\n",
    "                showline=True,\n",
    "                linecolor='black',\n",
    "                linewidth=1,\n",
    "                mirror=True\n",
    "            )\n",
    "        )\n",
    "        fig.show()\n",
    "        fig.write_image(f'UMAP_HAMBI_short/UMAP_{n}_{m}_short_5.png')\n",
    "        fig.write_html(f'UMAP_HAMBI_short/UMAP_{n}_{m}_short_5.html')\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f93b8b9d-f699-4c23-8dfa-c2e74b7b2a10",
   "metadata": {
    "jp-MarkdownHeadingCollapsed": true
   },
   "source": [
    "**************************************************************************\n",
    "## UMAP clustering for the methylation signature Position Weight Matrices\n",
    "#### Top feature data ðŸ”\n",
    "##### Create data set with only top features (150)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "818f58a8-70f7-41eb-963d-f550d5954955",
   "metadata": {},
   "outputs": [],
   "source": [
    "path_to_images = '/scratch/project_2006608/Methylation/notebooks/UMAP_HAMBI_top150_features/'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f9b5931a-61db-464f-a73b-a45da5431caf",
   "metadata": {},
   "outputs": [],
   "source": [
    "file_path = '/scratch/project_2006608/Methylation/HAMBI_data/merged_data.tsv'\n",
    "df = pd.read_csv(file_path, sep='\\t', index_col=0, low_memory=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1f4568fb-ef9d-4ce3-949a-2fddfa4adae5",
   "metadata": {},
   "outputs": [],
   "source": [
    "print(df.shape)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f61a43fa-4c87-4245-9132-1feaf54c3b3c",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Filter based on the features in 'RF_HAMBI/Domain_to_species_top_features.csv'\n",
    "## Load file\n",
    "file_path = '/scratch/project_2006608/Methylation/notebooks/RF_HAMBI/Domain_to_species_top150_features.csv'\n",
    "feature_list = pd.read_csv(file_path, sep=',')\n",
    "feature_list.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e29225ef-d691-437c-bbc1-59ed2b4a3742",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Extract identifiers into list\n",
    "taxa_headers = df.columns[493:510]\n",
    "column_headers = pd.DataFrame(taxa_headers.tolist(), columns=['Feature'])\n",
    "print(column_headers.head())\n",
    "\n",
    "# Append the column_headers to it\n",
    "feature_list = pd.concat([feature_list, column_headers], ignore_index=True)\n",
    "\n",
    "print(feature_list)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8f987c9a-e25f-4c25-8a5c-c40d99604201",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Extract columns\n",
    "columns_to_keep = feature_list['Feature'].tolist()\n",
    "df_top150 = df[columns_to_keep]\n",
    "df_top150.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "176ace0c-15e3-4034-9a76-485634e56dc8",
   "metadata": {},
   "outputs": [],
   "source": [
    "df_top150['Domain to species'].value_counts(dropna = False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f3ad4400-89d5-40a9-86e5-392332787a56",
   "metadata": {},
   "outputs": [],
   "source": [
    "row_names = df_top150.index\n",
    "print(row_names)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "98c9ab88-143a-45fe-8c77-e8b064b54224",
   "metadata": {},
   "outputs": [],
   "source": [
    "len(df_top150['Domain to species'].value_counts(dropna = False))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "10727b87-df47-469f-8f5b-518f217f25d7",
   "metadata": {
    "jp-MarkdownHeadingCollapsed": true
   },
   "source": [
    "## FILTERING\n",
    "### Removing the contigs that have no methyaltions (or under 20) detected for all of the three types. These are the rows where all values in the columns are 0s\n",
    "#### Top feature data ðŸ”"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "77e816b7-7e11-46eb-ae0b-8bd608d4277a",
   "metadata": {},
   "outputs": [],
   "source": [
    "# remove rows where df.iloc[:, :-17] meaning no methylation detected or less than 20 detected since those are converted to 0 filled matrices\n",
    "df_top150 = df_top150.loc[(df_top150.iloc[:, :-17] != 0).any(axis=1)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3dfdcf54-52a1-4a09-92e9-efb9dbe44c76",
   "metadata": {},
   "outputs": [],
   "source": [
    "print(df_top150.shape[0])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "be3b2961-6c90-4ef2-9e3a-1d3f5f038951",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "df_top150['Domain to species'].value_counts(dropna = False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "01fb0c7f-cab1-4955-9e1d-a2b3c78945f5",
   "metadata": {},
   "outputs": [],
   "source": [
    "df_top150['Domain to species'].nunique(dropna = False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "99d53ffc-22b2-4023-9f67-2ed408fd0483",
   "metadata": {},
   "outputs": [],
   "source": [
    "df_top150['Domain to genus'].nunique(dropna = False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c35fd948-72c8-4019-beb8-c7917af2bd79",
   "metadata": {},
   "outputs": [],
   "source": [
    "df_top150['Domain to family'].nunique(dropna = False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b9ab47f5-3de5-4f5f-a3d0-944307a76285",
   "metadata": {},
   "outputs": [],
   "source": [
    "df_top150['Domain to order'].nunique(dropna = False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5a4dd626-9483-40ef-8d78-5a898384e686",
   "metadata": {},
   "outputs": [],
   "source": [
    "df_top150['Domain to class'].nunique(dropna = False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "06ecbd11-7749-4815-8b1a-4a385ea0e6b7",
   "metadata": {},
   "outputs": [],
   "source": [
    "df_top150['Domain to phylum'].nunique(dropna = False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f970d98a-94ec-4224-9f13-fb2789e3485f",
   "metadata": {},
   "outputs": [],
   "source": [
    "len(df_top150[df_top150['Domain to species'].isna()])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b6858697-a675-4dee-a9e4-50784b03c01e",
   "metadata": {},
   "outputs": [],
   "source": [
    "df_top150.head()\n",
    "print(df_top150.iloc[:, :-17])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5c06a52e-bcf9-4075-8d1f-f517d81b9150",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "n_neighbors = [20, 30 ]\n",
    "min_dist = [0.01, 0.1, 0.2]\n",
    "colors = sns.color_palette(\"hls\", 32).as_hex()\n",
    "\n",
    "unique_species = df_top150['Domain to species'].unique()\n",
    "num_species = len(unique_species)\n",
    "color_dict = [\n",
    "    \"#FF0000\", \"#00FF00\", \"#0000FF\", \"#FFFF00\", \"#FF00FF\", \"#00FFFF\", \"#800000\", \"#008000\",\n",
    "    \"#000080\", \"#808000\", \"#800080\", \"#008080\", \"#FF4500\", \"#2E8B57\", \"#4682B4\", \"#DAA520\",\n",
    "    \"#FF1493\", \"#7FFF00\", \"#8A2BE2\", \"#5F9EA0\", \"#D2691E\", \"#DC143C\", \"#B8860B\", \"#32CD32\",\n",
    "    \"#FFD700\", \"#20B2AA\", \"#FF6347\", \"#4169E1\", \"#8B4513\", \"#6A5ACD\", \"#FF7F50\", \"#2F4F4F\"\n",
    "]\n",
    "\n",
    "# If not distinct colors\n",
    "color_dict = [\"#FF0000\"]\n",
    "custom_colors = color_dict\n",
    "\n",
    "for n in n_neighbors:\n",
    "    for m in min_dist:\n",
    "        \n",
    "        reducer = umap.UMAP(n_neighbors=n, min_dist=m, random_state=seed)\n",
    "        embedding = reducer.fit_transform(df_top150.iloc[:, :-17])\n",
    "        top150_UMAP_df  = pd.DataFrame({\n",
    "            'UMAP1': embedding[:, 0],\n",
    "            'UMAP2': embedding[:, 1],\n",
    "            'Domain to species': df_top150['Domain to species'],\n",
    "            'Domain to genus': df_top150['Domain to genus'],\n",
    "            'element': df_top150['element'],\n",
    "            'contig': df_top150.index,\n",
    "        })\n",
    "        # clustering\n",
    "        kmeans = KMeans(n_clusters=11, random_state=seed)\n",
    "        \n",
    "        k_labels = kmeans.fit_predict(embedding)\n",
    "        \n",
    "        top150_UMAP_df['cluster'] = k_labels\n",
    "        top150_UMAP_df['Domain to species'] = top150_UMAP_df['Domain to species'].fillna('NA')\n",
    "        column = 'Domain to species'\n",
    "        value_counts = top150_UMAP_df[column].value_counts().to_dict()\n",
    "\n",
    "        top150_UMAP_df[f'{column}_with_counts'] = top150_UMAP_df[column].apply(lambda x: f'{x} ({value_counts[x]})')\n",
    "\n",
    "        top150_UMAP_df['count'] = top150_UMAP_df[column].map(value_counts)\n",
    "        top150_UMAP_df = top150_UMAP_df.sort_values(by=['count'], ascending=False)\n",
    "\n",
    "        top150_UMAP_df[f'{column}_with_counts'] = pd.Categorical(\n",
    "            top150_UMAP_df[f'{column}_with_counts'],\n",
    "            categories=top150_UMAP_df[f'{column}_with_counts'].unique(),\n",
    "            ordered=True\n",
    "        )\n",
    "        \n",
    "        fig = px.scatter(top150_UMAP_df, \n",
    "                            x='UMAP1', \n",
    "                            y='UMAP2', \n",
    "                            labels={f'{column}_with_counts': column},\n",
    "                            color=f'{column}_with_counts', \n",
    "                            title=f' HAMBI - UMAP with n_neighbors={n}, min_dist={m}', \n",
    "                            color_discrete_sequence=custom_colors,\n",
    "                            hover_data={'cluster': True, 'Domain to species': True, 'Domain to genus': True, 'element': True, 'contig': True} )\n",
    "        title = f' HAMBI - UMAP with n_neighbors={n}, min_dist={m}'\n",
    "        fig.update_layout(\n",
    "            height=1500,\n",
    "            width=1200,\n",
    "            title_text=title,\n",
    "            showlegend=True,\n",
    "            legend=dict(\n",
    "                x=0.5,\n",
    "                y=-0.1,\n",
    "                traceorder=\"normal\",\n",
    "                xanchor='center',\n",
    "                yanchor='top',\n",
    "                orientation='h'\n",
    "            ),\n",
    "            template='simple_white',\n",
    "            xaxis=dict(\n",
    "                showgrid=True,\n",
    "                gridcolor='lightgray',\n",
    "                zeroline=False,\n",
    "                showline=True,  \n",
    "                linecolor='black', \n",
    "                linewidth=1,\n",
    "                mirror=True\n",
    "            ),\n",
    "            yaxis=dict(\n",
    "                showgrid=True,\n",
    "                gridcolor='lightgray',\n",
    "                zeroline=False,\n",
    "                showline=True,\n",
    "                linecolor='black',\n",
    "                linewidth=1,\n",
    "                mirror=True\n",
    "            )\n",
    "        )\n",
    "        fig.show()\n",
    "        #fig.write_image(f'UMAP_HAMBI_top_features/UMAP_{n}_{m}.png')\n",
    "        #fig.write_html(f'UMAP_HAMBI_top_features/UMAP_{n}_{m}.html')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f44e9551-cdbc-44c4-b560-1959579535ea",
   "metadata": {},
   "source": [
    "### Using the best n_neighbours and min_dist to get clear clusters\n",
    "#### Top feature data ðŸ”"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "730b44d1-5fca-4e1f-bec7-dfafc867b8a6",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "n_neighbors = [30]\n",
    "min_dist = [0.1]\n",
    "color_dict = [\n",
    "    \"#FF0000\", \"#00FF00\", \"#0000FF\", \"#FFFF00\", \"#FF00FF\", \"#00FFFF\", \"#800000\", \"#008000\",\n",
    "    \"#000080\", \"#808000\", \"#800080\", \"#008080\", \"#FF4500\", \"#3bccf7\", \"#4682B4\", \"#DAA520\",\n",
    "    \"#FF1493\", \"#9ACD32\", \"#8A2BE2\", \"#5F9EA0\", \"#D2691E\", \"#DC143C\", \"#B8860B\", \"#32CD32\",\n",
    "    \"#FFD700\", \"#20B2AA\", \"#FF6347\", \"#4169E1\", \"#8B4513\", \"#6A5ACD\", \"#FF7F50\", \"#2F4F4F\"\n",
    "]\n",
    "\n",
    "# If not distinct colors\n",
    "#color_dict = [\"#FF0000\"]\n",
    "custom_colors = color_dict\n",
    "\n",
    "for n in n_neighbors:\n",
    "    for m in min_dist:\n",
    "        reducer = umap.UMAP(n_neighbors=n, min_dist=m, random_state=seed)\n",
    "        embedding = reducer.fit_transform(df_top150.iloc[:, :-17])\n",
    "        top150_UMAP_df  = pd.DataFrame({\n",
    "            'UMAP1': embedding[:, 0],\n",
    "            'UMAP2': embedding[:, 1],\n",
    "            'Domain to species': df_top150['Domain to species'],\n",
    "            'Domain to genus': df_top150['Domain to genus'],\n",
    "            'element': df_top150['element'],\n",
    "            'contig': df_top150.index,\n",
    "        })\n",
    "        # clustering k defined choosing the k by looking at the UMAP \n",
    "        kmeans = KMeans(n_clusters=11, random_state=seed)\n",
    "        \n",
    "        k_labels = kmeans.fit_predict(embedding)\n",
    "        \n",
    "        top150_UMAP_df['cluster'] = k_labels\n",
    "        \n",
    "        column = 'Domain to species'\n",
    "        top150_UMAP_df['Domain to species'] = top150_UMAP_df['Domain to species'].fillna('NA')\n",
    "\n",
    "        value_counts = top150_UMAP_df[column].value_counts().to_dict()\n",
    "\n",
    "        top150_UMAP_df[f'{column}_with_counts'] = top150_UMAP_df[column].apply(lambda x: f'{x} ({value_counts[x]})')\n",
    "\n",
    "        top150_UMAP_df['count'] = top150_UMAP_df[column].map(value_counts)\n",
    "        top150_UMAP_df = top150_UMAP_df.sort_values(by=['count'], ascending=False)\n",
    "\n",
    "        top150_UMAP_df[f'{column}_with_counts'] = pd.Categorical(\n",
    "            top150_UMAP_df[f'{column}_with_counts'],\n",
    "            categories=top150_UMAP_df[f'{column}_with_counts'].unique(),\n",
    "            ordered=True\n",
    "        )\n",
    "        \n",
    "        fig = px.scatter(top150_UMAP_df, \n",
    "                            x='UMAP1', \n",
    "                            y='UMAP2', \n",
    "                            labels={f'{column}_with_counts': column},\n",
    "                            color=f'{column}_with_counts', \n",
    "                            title=f' HAMBI - UMAP with n_neighbors={n}, min_dist={m}', \n",
    "                            color_discrete_sequence=custom_colors,\n",
    "                            hover_data={'cluster': True, 'Domain to species': True, 'Domain to genus': True,\n",
    "                                        'element': True, 'contig': True})\n",
    "        title = f' HAMBI - UMAP with n_neighbors={n}, min_dist={m}'\n",
    "        fig.update_layout(\n",
    "            height=1700,\n",
    "            width=1200,\n",
    "            title_text=title,\n",
    "            showlegend=True,\n",
    "            legend=dict(\n",
    "                x=0.5,\n",
    "                y=-0.1,\n",
    "                traceorder=\"normal\",\n",
    "                xanchor='center',\n",
    "                yanchor='top',\n",
    "                orientation='h'\n",
    "            ),\n",
    "            template='simple_white',\n",
    "            xaxis=dict(\n",
    "                showgrid=True,\n",
    "                gridcolor='lightgray',\n",
    "                zeroline=False,\n",
    "                showline=True,  \n",
    "                linecolor='black', \n",
    "                linewidth=1,\n",
    "                mirror=True\n",
    "            ),\n",
    "            yaxis=dict(\n",
    "                showgrid=True,\n",
    "                gridcolor='lightgray',\n",
    "                zeroline=False,\n",
    "                showline=True,\n",
    "                linecolor='black',\n",
    "                linewidth=1,\n",
    "                mirror=True\n",
    "            )\n",
    "        )\n",
    "        fig.show()\n",
    "        #fig.write_image(f'UMAP_HAMBI_top150_features/UMAP_{n}_{m}.png')\n",
    "        #fig.write_html(f'UMAP_HAMBI_top150_features/UMAP_{n}_{m}.html')\n",
    "        #fig.write_image(f'UMAP_HAMBI_top150_features/UMAP_{n}_{m}_no_colors.png')\n",
    "        #fig.write_html(f'UMAP_HAMBI_top150_features/UMAP_{n}_{m}_no_colors.html')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "32a73107-b30e-4b97-8ae2-5c8a4bf701c1",
   "metadata": {},
   "source": [
    "### Extract clusters and redraw UMAP with subset of data\n",
    "#### Top feature data ðŸ” (150)\n",
    "##### C1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f8f3efc5-84d9-4ee3-a9ca-0aa1ccb83ea8",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Extract based on UMAP1 (x axis) & UMAP2 (y axis) values:\n",
    "# Save contig IDs\n",
    "directory = 'UMAP_HAMBI_top150_features'\n",
    "\n",
    "## C1\n",
    "top150_UMAP_df_C1 = top150_UMAP_df_C1.loc[(top150_UMAP_df['UMAP1']>= 10.2) & (top150_UMAP_df['UMAP1']<= 10.6)\n",
    "    & (top150_UMAP_df['UMAP2']>= 9.5) & (top150_UMAP_df['UMAP2']<= 9.8)]\n",
    "\n",
    "df_top150_C1 = df_top150[df_top150.index.isin(top150_UMAP_df_C1.index)]\n",
    "\n",
    "top150_C1_contigs = df_top150_C1.index.to_list()\n",
    "file_path = os.path.join(directory, 'top150_C1_contigs.txt')\n",
    "\n",
    "with open(file_path, 'w') as file:\n",
    "    for item in top150_C1_contigs:\n",
    "        file.write(f\"{item}\\n\")\n",
    "\n",
    "## C2\n",
    "top150_UMAP_df_C2 = top150_UMAP_df.loc[(top150_UMAP_df['UMAP1']>= 10.2) & (top150_UMAP_df['UMAP1']<= 10.4)\n",
    "    & (top150_UMAP_df['UMAP2']>= 9.8) & (top150_UMAP_df['UMAP2']<= 10)]\n",
    "\n",
    "df_top150_C2 = df_top150[df_top150.index.isin(top150_UMAP_df_C2.index)]\n",
    "\n",
    "top150_C2_contigs = df_top150_C2.index.to_list()\n",
    "file_path = os.path.join(directory, 'top150_C2_contigs.txt')\n",
    "\n",
    "with open(file_path, 'w') as file:\n",
    "    for item in top150_C2_contigs:\n",
    "        file.write(f\"{item}\\n\")\n",
    "\n",
    "## C3\n",
    "top150_UMAP_df_C3 = top150_UMAP_df.loc[(top150_UMAP_df['UMAP1']>= 7) & (top150_UMAP_df['UMAP1']<= 8)\n",
    "    & (top150_UMAP_df['UMAP2']>= -5.15) & (top150_UMAP_df['UMAP2']<= -4.5)]\n",
    "\n",
    "df_top150_C3 = df_top150[df_top150.index.isin(top150_UMAP_df_C3.index)]\n",
    "\n",
    "top150_C3_contigs = df_top150_C3.index.to_list()\n",
    "file_path = os.path.join(directory, 'top150_C3_contigs.txt')\n",
    "\n",
    "with open(file_path, 'w') as file:\n",
    "    for item in top150_C3_contigs:\n",
    "        file.write(f\"{item}\\n\")\n",
    "\n",
    "## C4\n",
    "top150_UMAP_df_C4 = top150_UMAP_df.loc[(top150_UMAP_df['UMAP1']>= 8) & (top150_UMAP_df['UMAP1']<= 9)\n",
    "    & (top150_UMAP_df['UMAP2']>= -5.3) & (top150_UMAP_df['UMAP2']<= -4.6)]\n",
    "\n",
    "df_top150_C4 = df_top150[df_top150.index.isin(top150_UMAP_df_4.index)]\n",
    "\n",
    "top150_C4_contigs = df_top150_C4.index.to_list()\n",
    "file_path = os.path.join(directory, 'top150_C4_contigs.txt')\n",
    "\n",
    "with open(file_path, 'w') as file:\n",
    "    for item in top150_C4_contigs:\n",
    "        file.write(f\"{item}\\n\")\n",
    "\n",
    "## C5\n",
    "top150_UMAP_df_C5 = top150_UMAP_df.loc[(top150_UMAP_df['UMAP1']>= 7) & (top150_UMAP_df['UMAP1']<= 8.2)\n",
    "    & (top150_UMAP_df['UMAP2']>= -6.7) & (top150_UMAP_df['UMAP2']<= -5.5)]\n",
    "\n",
    "df_top150_C5 = df_top150[df_top150.index.isin(top150_UMAP_df_C5.index)]\n",
    "\n",
    "top150_C5_contigs = df_top150_C5.index.to_list()\n",
    "file_path = os.path.join(directory, 'top150_C5_contigs.txt')\n",
    "\n",
    "with open(file_path, 'w') as file:\n",
    "    for item in top150_C5_contigs:\n",
    "        file.write(f\"{item}\\n\")\n",
    "\n",
    "## C6\n",
    "top150_UMAP_df_C6 = top150_UMAP_df.loc[(top150_UMAP_df['UMAP1']>= 14.09) & (top150_UMAP_df['UMAP1']<= 15.5)\n",
    "    & (top150_UMAP_df['UMAP2']>= -3.3) & (top150_UMAP_df['UMAP2']<= -2.3)]\n",
    "\n",
    "df_top150_C6 = df_top150[df_top150.index.isin(top150_UMAP_df_C6.index)]\n",
    "\n",
    "top150_C6_contigs = df_top150_C6.index.to_list()\n",
    "file_path = os.path.join(directory, 'top150_C6_contigs.txt')\n",
    "\n",
    "with open(file_path, 'w') as file:\n",
    "    for item in top150_C6_contigs:\n",
    "        file.write(f\"{item}\\n\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4decfbd5-0087-4295-adc7-3d170183684c",
   "metadata": {},
   "source": [
    "### Extract clusters and redraw UMAP with subset of data\n",
    "#### Top feature data ðŸ” (150)\n",
    "##### C2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "139db67a-fe20-419d-87ef-ed3a71a0530c",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Extract based on UMAP1 (x axis) & UMAP2 (y axis) values:\n",
    "top150_UMAP_df_C2 = top150_UMAP_df.loc[(top150_UMAP_df['UMAP1']>= 5) & (top150_UMAP_df['UMAP1']<= 6.5)\n",
    "    & (top150_UMAP_df['UMAP2']>= 5) & (top150_UMAP_df['UMAP2']<= 7.5)]\n",
    "\n",
    "# Check\n",
    "top150_UMAP_df_C2.head()\n",
    "\n",
    "# Save those contig IDs to extract from the 'df_top150'\n",
    "top150_UMAP_df_C2.index.to_list()\n",
    "\n",
    "df_top150_C2 = df_top150[df_top150.index.isin(top150_UMAP_df_C2.index)]\n",
    "print(df_top150_C2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3c3b7a7f-edec-424e-8f85-5583defe4e7a",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Save contig IDs\n",
    "top150_C2_contigs = df_top150_C2.index.to_list()\n",
    "\n",
    "directory = 'UMAP_HAMBI_top150_features'\n",
    "file_path = os.path.join(directory, 'top150_C2_contigs.txt')\n",
    "\n",
    "with open(file_path, 'w') as file:\n",
    "    for item in top150_C2_contigs:\n",
    "        file.write(f\"{item}\\n\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "af04a844-cc52-438f-9f48-c6f75b86195f",
   "metadata": {},
   "outputs": [],
   "source": [
    "n_neighbors = [30]\n",
    "min_dist = [0.1]\n",
    "color_dict = [\"#000080\", \"#FF00FF\", \"#eadf27\", \"#00FFFF\", \"#0000FF\", \"#800000\"]\n",
    "\n",
    "custom_colors = color_dict\n",
    "for n in n_neighbors:\n",
    "    for m in min_dist:\n",
    "        reducer = umap.UMAP(n_neighbors=n, min_dist=m, random_state=seed)\n",
    "        embedding = reducer.fit_transform(df_top150_C2.iloc[:, :-17])\n",
    "        top150_UMAP_df_C2  = pd.DataFrame({\n",
    "            'UMAP1': embedding[:, 0],\n",
    "            'UMAP2': embedding[:, 1],\n",
    "            'Domain to species': df_top150_C2['Domain to species'],\n",
    "            'Domain to genus': df_top150_C2['Domain to genus'],\n",
    "            'element': df_top150_C2['element'],\n",
    "            'contig': df_top150_C2.index,\n",
    "        })\n",
    "        # clustering k defined choosing the k by looking at the UMAP \n",
    "        kmeans = KMeans(n_clusters=11, random_state=seed)\n",
    "        \n",
    "        k_labels = kmeans.fit_predict(embedding)\n",
    "        \n",
    "        top150_UMAP_df_C2['cluster'] = k_labels\n",
    "        \n",
    "        column = 'Domain to species'\n",
    "        top150_UMAP_df_C2['Domain to species'] = top150_UMAP_df_C2['Domain to species'].fillna('NA')\n",
    "\n",
    "        value_counts = top150_UMAP_df_C2[column].value_counts().to_dict()\n",
    "\n",
    "        top150_UMAP_df_C2[f'{column}_with_counts'] = top150_UMAP_df_C2[column].apply(lambda x: f'{x} ({value_counts[x]})')\n",
    "\n",
    "        top150_UMAP_df_C2['count'] = top150_UMAP_df_C2[column].map(value_counts)\n",
    "        top150_UMAP_df_C2 = top150_UMAP_df_C2.sort_values(by=['count'], ascending=False)\n",
    "\n",
    "        top150_UMAP_df_C2[f'{column}_with_counts'] = pd.Categorical(\n",
    "            top150_UMAP_df_C2[f'{column}_with_counts'],\n",
    "            categories=top150_UMAP_df_C2[f'{column}_with_counts'].unique(),\n",
    "            ordered=True\n",
    "        )\n",
    "        \n",
    "        fig = px.scatter(top150_UMAP_df_C2, \n",
    "                            x='UMAP1', \n",
    "                            y='UMAP2', \n",
    "                            labels={f'{column}_with_counts': column},\n",
    "                            color=f'{column}_with_counts', \n",
    "                            title=f' HAMBI - UMAP with n_neighbors={n}, min_dist={m}', \n",
    "                            color_discrete_sequence=custom_colors,\n",
    "                            hover_data={'cluster': True, 'Domain to species': True, 'Domain to genus': True, 'element': True, 'contig': True} )\n",
    "        title = f' HAMBI - UMAP with n_neighbors={n}, min_dist={m}'\n",
    "        fig.update_traces(marker=dict(size=10))\n",
    "        fig.update_layout(\n",
    "            height=1700,\n",
    "            width=1200,\n",
    "            title_text=title,\n",
    "            showlegend=True,\n",
    "            legend=dict(\n",
    "                x=0.5,\n",
    "                y=-0.1,\n",
    "                traceorder=\"normal\",\n",
    "                xanchor='center',\n",
    "                yanchor='top',\n",
    "                orientation='h'\n",
    "            ),\n",
    "            template='simple_white',\n",
    "            xaxis=dict(\n",
    "                showgrid=True,\n",
    "                gridcolor='lightgray',\n",
    "                zeroline=False,\n",
    "                showline=True,  \n",
    "                linecolor='black', \n",
    "                linewidth=1,\n",
    "                mirror=True\n",
    "            ),\n",
    "            yaxis=dict(\n",
    "                showgrid=True,\n",
    "                gridcolor='lightgray',\n",
    "                zeroline=False,\n",
    "                showline=True,\n",
    "                linecolor='black',\n",
    "                linewidth=1,\n",
    "                mirror=True\n",
    "            )\n",
    "        )\n",
    "        fig.show()\n",
    "        fig.write_image(f'UMAP_HAMBI_top150_features/UMAP_{n}_{m}_C2.png')\n",
    "        fig.write_html(f'UMAP_HAMBI_top150_features/UMAP_{n}_{m}_C2.html')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "bdf6fa04-e0b0-4df9-8d60-1c579af02567",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Extract based on UMAP1 (x axis) & UMAP2 (y axis) values:\n",
    "# Save contig IDs\n",
    "directory = 'UMAP_HAMBI_top150_features'\n",
    "\n",
    "## C1A\n",
    "top150_UMAP_df_C2A = top150_UMAP_df_C2.loc[(top150_UMAP_df_C2['UMAP1']>= -6) & (top150_UMAP_df_C2['UMAP1']<= -4.10)\n",
    "    & (top150_UMAP_df_C2['UMAP2']>= 12.5) & (top150_UMAP_df_C2['UMAP2']<= 14)]\n",
    "\n",
    "df_top150_C2A = df_top150[df_top150.index.isin(top150_UMAP_df_C2A.index)]\n",
    "\n",
    "top150_C2A_contigs = df_top150_C2A.index.to_list()\n",
    "file_path = os.path.join(directory, 'top150_C2A_contigs.txt')\n",
    "\n",
    "with open(file_path, 'w') as file:\n",
    "    for item in top150_C2A_contigs:\n",
    "        file.write(f\"{item}\\n\")\n",
    "\n",
    "## C1B\n",
    "top150_UMAP_df_C2B = top150_UMAP_df_C2.loc[(top150_UMAP_df_C2['UMAP1']>= -7) & (top150_UMAP_df_C2['UMAP1']<= -6.5)\n",
    "    & (top150_UMAP_df_C2['UMAP2']>= 10.5) & (top150_UMAP_df_C2['UMAP2']<= 11)]\n",
    "\n",
    "df_top150_C2B = df_top150[df_top150.index.isin(top150_UMAP_df_C2B.index)]\n",
    "\n",
    "top150_C2B_contigs = df_top150_C2B.index.to_list()\n",
    "file_path = os.path.join(directory, 'top150_C2B_contigs.txt')\n",
    "\n",
    "with open(file_path, 'w') as file:\n",
    "    for item in top150_C2B_contigs:\n",
    "        file.write(f\"{item}\\n\")\n",
    "\n",
    "## C1C\n",
    "top150_UMAP_df_C2C = top150_UMAP_df_C2.loc[(top150_UMAP_df_C2['UMAP1']>= -9) & (top150_UMAP_df_C2['UMAP1']<= -7.5)\n",
    "    & (top150_UMAP_df_C2['UMAP2']>= 9) & (top150_UMAP_df_C2['UMAP2']<= 10.5)]\n",
    "\n",
    "df_top150_C2C = df_top150[df_top150.index.isin(top150_UMAP_df_C2C.index)]\n",
    "\n",
    "top150_C2C_contigs = df_top150_C2C.index.to_list()\n",
    "file_path = os.path.join(directory, 'top150_C2C_contigs.txt')\n",
    "\n",
    "with open(file_path, 'w') as file:\n",
    "    for item in top150_C2C_contigs:\n",
    "        file.write(f\"{item}\\n\")\n",
    "\n",
    "## C1D\n",
    "top150_UMAP_df_C2D = top150_UMAP_df_C2.loc[(top150_UMAP_df_C2['UMAP1']>= -4) & (top150_UMAP_df_C2['UMAP1']<= -3)\n",
    "    & (top150_UMAP_df_C2['UMAP2']>= 10.5) & (top150_UMAP_df_C2['UMAP2']<= 11.7)]\n",
    "\n",
    "df_top150_C2D = df_top150[df_top150.index.isin(top150_UMAP_df_C2D.index)]\n",
    "\n",
    "top150_C2D_contigs = df_top150_C2D.index.to_list()\n",
    "file_path = os.path.join(directory, 'top150_C2D_contigs.txt')\n",
    "\n",
    "with open(file_path, 'w') as file:\n",
    "    for item in top150_C2D_contigs:\n",
    "        file.write(f\"{item}\\n\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "46f4ba1d-87ac-492b-a849-5a1412c3c26a",
   "metadata": {},
   "source": [
    "### Extract clusters and redraw UMAP with subset of data\n",
    "#### Top feature data ðŸ” (150)\n",
    "##### C3"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "796d506c-a6ef-4f0b-8e90-f9418db20432",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Extract based on UMAP1 (x axis) & UMAP2 (y axis) values:\n",
    "top150_UMAP_df_C3 = top150_UMAP_df.loc[(top150_UMAP_df['UMAP1']>= -2.4) & (top150_UMAP_df['UMAP1']<= 0)\n",
    "    & (top150_UMAP_df['UMAP2']>= 7) & (top150_UMAP_df['UMAP2']<= 8.2)]\n",
    "\n",
    "# Check\n",
    "top150_UMAP_df_C3.head()\n",
    "\n",
    "# Save those contig IDs to extract from the 'df_top150'\n",
    "top150_UMAP_df_C3.index.to_list()\n",
    "\n",
    "df_top150_C3 = df_top150[df_top150.index.isin(top150_UMAP_df_C3.index)]\n",
    "print(df_top150_C3)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "fa7bcbe0-e211-4bea-a0ea-92b0bd58a62c",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Save contig IDs\n",
    "top150_C3_contigs = df_top150_C3.index.to_list()\n",
    "\n",
    "directory = 'UMAP_HAMBI_top150_features'\n",
    "file_path = os.path.join(directory, 'top150_C3_contigs.txt')\n",
    "\n",
    "with open(file_path, 'w') as file:\n",
    "    for item in top150_C3_contigs:\n",
    "        file.write(f\"{item}\\n\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "088b0564-b478-4afb-84ac-0f9be67206c3",
   "metadata": {},
   "outputs": [],
   "source": [
    "n_neighbors = [30]\n",
    "min_dist = [0.1]\n",
    "color_dict = [\"#008000\", \"#800080\", \"#00FF00\", \"#FF0000\"]\n",
    "\n",
    "custom_colors = color_dict\n",
    "for n in n_neighbors:\n",
    "    for m in min_dist:\n",
    "        reducer = umap.UMAP(n_neighbors=n, min_dist=m, random_state=seed)\n",
    "        embedding = reducer.fit_transform(df_top150_C3.iloc[:, :-17])\n",
    "        top150_UMAP_df_C3  = pd.DataFrame({\n",
    "            'UMAP1': embedding[:, 0],\n",
    "            'UMAP2': embedding[:, 1],\n",
    "            'Domain to species': df_top150_C3['Domain to species'],\n",
    "            'Domain to genus': df_top150_C3['Domain to genus'],\n",
    "            'element': df_top150_C3['element'],\n",
    "            'contig': df_top150_C3.index,\n",
    "        })\n",
    "        # clustering k defined choosing the k by looking at the UMAP \n",
    "        kmeans = KMeans(n_clusters=11, random_state=seed)\n",
    "        \n",
    "        k_labels = kmeans.fit_predict(embedding)\n",
    "        \n",
    "        top150_UMAP_df_C3['cluster'] = k_labels\n",
    "        \n",
    "        column = 'Domain to species'\n",
    "        top150_UMAP_df_C3['Domain to species'] = top150_UMAP_df_C3['Domain to species'].fillna('NA')\n",
    "\n",
    "        value_counts = top150_UMAP_df_C3[column].value_counts().to_dict()\n",
    "\n",
    "        top150_UMAP_df_C3[f'{column}_with_counts'] = top150_UMAP_df_C3[column].apply(lambda x: f'{x} ({value_counts[x]})')\n",
    "\n",
    "        top150_UMAP_df_C3['count'] = top150_UMAP_df_C3[column].map(value_counts)\n",
    "        top150_UMAP_df_C3 = top150_UMAP_df_C3.sort_values(by=['count'], ascending=False)\n",
    "\n",
    "        top150_UMAP_df_C3[f'{column}_with_counts'] = pd.Categorical(\n",
    "            top150_UMAP_df_C3[f'{column}_with_counts'],\n",
    "            categories=top150_UMAP_df_C3[f'{column}_with_counts'].unique(),\n",
    "            ordered=True\n",
    "        )\n",
    "        \n",
    "        fig = px.scatter(top150_UMAP_df_C3, \n",
    "                            x='UMAP1', \n",
    "                            y='UMAP2', \n",
    "                            labels={f'{column}_with_counts': column},\n",
    "                            color=f'{column}_with_counts', \n",
    "                            title=f' HAMBI - UMAP with n_neighbors={n}, min_dist={m}', \n",
    "                            color_discrete_sequence=custom_colors,\n",
    "                            hover_data={'cluster': True, 'Domain to species': True, 'Domain to genus': True, 'element': True, 'contig': True} )\n",
    "        title = f' HAMBI - UMAP with n_neighbors={n}, min_dist={m}'\n",
    "        fig.update_traces(marker=dict(size=10))\n",
    "        fig.update_layout(\n",
    "            height=1700,\n",
    "            width=1200,\n",
    "            title_text=title,\n",
    "            showlegend=True,\n",
    "            legend=dict(\n",
    "                x=0.5,\n",
    "                y=-0.1,\n",
    "                traceorder=\"normal\",\n",
    "                xanchor='center',\n",
    "                yanchor='top',\n",
    "                orientation='h'\n",
    "            ),\n",
    "            template='simple_white',\n",
    "            xaxis=dict(\n",
    "                showgrid=True,\n",
    "                gridcolor='lightgray',\n",
    "                zeroline=False,\n",
    "                showline=True,  \n",
    "                linecolor='black', \n",
    "                linewidth=1,\n",
    "                mirror=True\n",
    "            ),\n",
    "            yaxis=dict(\n",
    "                showgrid=True,\n",
    "                gridcolor='lightgray',\n",
    "                zeroline=False,\n",
    "                showline=True,\n",
    "                linecolor='black',\n",
    "                linewidth=1,\n",
    "                mirror=True\n",
    "            )\n",
    "        )\n",
    "        fig.show()\n",
    "        fig.write_image(f'UMAP_HAMBI_top150_features/UMAP_{n}_{m}_C3.png')\n",
    "        fig.write_html(f'UMAP_HAMBI_top150_features/UMAP_{n}_{m}_C3.html')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "081ef9c5-78d4-45ed-a32e-0174e90b0b8e",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Extract based on UMAP1 (x axis) & UMAP2 (y axis) values:\n",
    "# Save contig IDs\n",
    "directory = 'UMAP_HAMBI_top150_features'\n",
    "\n",
    "## C3A\n",
    "top150_UMAP_df_C3A = top150_UMAP_df_C3.loc[(top150_UMAP_df_C3['UMAP1']>= 8.2) & (top150_UMAP_df_C3['UMAP1']<= 11)\n",
    "    & (top150_UMAP_df_C3['UMAP2']>= 7.5) & (top150_UMAP_df_C3['UMAP2']<= 9.2)]\n",
    "\n",
    "df_top150_C3A = df_top150[df_top150.index.isin(top150_UMAP_df_C3A.index)]\n",
    "\n",
    "top150_C3A_contigs = df_top150_C3A.index.to_list()\n",
    "file_path = os.path.join(directory, 'top150_C3A_contigs.txt')\n",
    "\n",
    "with open(file_path, 'w') as file:\n",
    "    for item in top150_C3A_contigs:\n",
    "        file.write(f\"{item}\\n\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "351be832-8477-4de6-a412-b6b5e6994f95",
   "metadata": {},
   "source": [
    "### Extract clusters and redraw UMAP with subset of data\n",
    "#### Top feature data ðŸ” (150)\n",
    "##### C4"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2883877b-989f-4528-93b7-75942e59fba7",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Extract based on UMAP1 (x axis) & UMAP2 (y axis) values:\n",
    "## C4A\n",
    "top150_UMAP_df_C4A = top150_UMAP_df.loc[(top150_UMAP_df['UMAP1']>= -3.7) & (top150_UMAP_df['UMAP1']<= -3.20)\n",
    "    & (top150_UMAP_df['UMAP2']>= 8.3) & (top150_UMAP_df['UMAP2']<= 10)]\n",
    "\n",
    "# Check\n",
    "top150_UMAP_df_C4A.head()\n",
    "\n",
    "# Save those contig IDs to extract from the 'df_top150'\n",
    "top150_UMAP_df_C4A.index.to_list()\n",
    "\n",
    "df_top150_C4A = df_top150[df_top150.index.isin(top150_UMAP_df_C4A.index)]\n",
    "print(df_top150_C4A)\n",
    "\n",
    "\n",
    "\n",
    "## C4\n",
    "top150_UMAP_df_C4 = top150_UMAP_df.loc[(top150_UMAP_df['UMAP1']>= -6.10) & (top150_UMAP_df['UMAP1']<= -2.8)\n",
    "    & (top150_UMAP_df['UMAP2']>= 6.6) & (top150_UMAP_df['UMAP2']<= 9.3)]\n",
    "\n",
    "# Check\n",
    "top150_UMAP_df_C4.head()\n",
    "\n",
    "# Save those contig IDs to extract from the 'df_top150'\n",
    "top150_UMAP_df_C4.index.to_list()\n",
    "\n",
    "df_top150_C4 = df_top150[df_top150.index.isin(top150_UMAP_df_C4.index)]\n",
    "print(df_top150_C4)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a4872b25-eaa3-40df-9dac-2733e6f9255f",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Save contig IDs\n",
    "top150_C4A_contigs = df_top150_C4A.index.to_list()\n",
    "\n",
    "directory = 'UMAP_HAMBI_top150_features'\n",
    "file_path = os.path.join(directory, 'top150_C4A_contigs.txt')\n",
    "\n",
    "with open(file_path, 'w') as file:\n",
    "    for item in top150_C4A_contigs:\n",
    "        file.write(f\"{item}\\n\")\n",
    "\n",
    "\n",
    "# Save contig IDs\n",
    "top150_C4_contigs = df_top150_C4.index.to_list()\n",
    "\n",
    "directory = 'UMAP_HAMBI_top150_features'\n",
    "file_path = os.path.join(directory, 'top150_C4_contigs.txt')\n",
    "\n",
    "with open(file_path, 'w') as file:\n",
    "    for item in top150_C4_contigs:\n",
    "        file.write(f\"{item}\\n\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6e2e3abf-d543-41ec-9f50-d817795c0108",
   "metadata": {},
   "outputs": [],
   "source": [
    "n_neighbors = [30]\n",
    "min_dist = [0.1]\n",
    "color_dict = [\"#00FF00\", \"#800080\", \"#FF0000\", \"#008000\"]\n",
    "\n",
    "custom_colors = color_dict\n",
    "for n in n_neighbors:\n",
    "    for m in min_dist:\n",
    "        reducer = umap.UMAP(n_neighbors=n, min_dist=m, random_state=seed)\n",
    "        embedding = reducer.fit_transform(df_top150_C4.iloc[:, :-17])\n",
    "        top150_UMAP_df_C4  = pd.DataFrame({\n",
    "            'UMAP1': embedding[:, 0],\n",
    "            'UMAP2': embedding[:, 1],\n",
    "            'Domain to species': df_top150_C4['Domain to species'],\n",
    "            'Domain to genus': df_top150_C4['Domain to genus'],\n",
    "            'element': df_top150_C4['element'],\n",
    "            'contig': df_top150_C4.index,\n",
    "        })\n",
    "        # clustering k defined choosing the k by looking at the UMAP \n",
    "        kmeans = KMeans(n_clusters=11, random_state=seed)\n",
    "        \n",
    "        k_labels = kmeans.fit_predict(embedding)\n",
    "        \n",
    "        top150_UMAP_df_C4['cluster'] = k_labels\n",
    "        \n",
    "        column = 'Domain to species'\n",
    "        top150_UMAP_df_C4['Domain to species'] = top150_UMAP_df_C4['Domain to species'].fillna('NA')\n",
    "\n",
    "        value_counts = top150_UMAP_df_C4[column].value_counts().to_dict()\n",
    "\n",
    "        top150_UMAP_df_C4[f'{column}_with_counts'] = top150_UMAP_df_C4[column].apply(lambda x: f'{x} ({value_counts[x]})')\n",
    "\n",
    "        top150_UMAP_df_C4['count'] = top150_UMAP_df_C4[column].map(value_counts)\n",
    "        top150_UMAP_df_C4 = top150_UMAP_df_C4.sort_values(by=['count'], ascending=False)\n",
    "\n",
    "        top150_UMAP_df_C4[f'{column}_with_counts'] = pd.Categorical(\n",
    "            top150_UMAP_df_C4[f'{column}_with_counts'],\n",
    "            categories=top150_UMAP_df_C4[f'{column}_with_counts'].unique(),\n",
    "            ordered=True\n",
    "        )\n",
    "        \n",
    "        fig = px.scatter(top150_UMAP_df_C4, \n",
    "                            x='UMAP1', \n",
    "                            y='UMAP2', \n",
    "                            labels={f'{column}_with_counts': column},\n",
    "                            color=f'{column}_with_counts', \n",
    "                            title=f' HAMBI - UMAP with n_neighbors={n}, min_dist={m}', \n",
    "                            color_discrete_sequence=custom_colors,\n",
    "                            hover_data={'cluster': True, 'Domain to species': True, 'Domain to genus': True, 'element': True, 'contig': True} )\n",
    "        title = f' HAMBI - UMAP with n_neighbors={n}, min_dist={m}'\n",
    "        fig.update_traces(marker=dict(size=10))\n",
    "        fig.update_layout(\n",
    "            height=1700,\n",
    "            width=1200,\n",
    "            title_text=title,\n",
    "            showlegend=True,\n",
    "            legend=dict(\n",
    "                x=0.5,\n",
    "                y=-0.1,\n",
    "                traceorder=\"normal\",\n",
    "                xanchor='center',\n",
    "                yanchor='top',\n",
    "                orientation='h'\n",
    "            ),\n",
    "            template='simple_white',\n",
    "            xaxis=dict(\n",
    "                showgrid=True,\n",
    "                gridcolor='lightgray',\n",
    "                zeroline=False,\n",
    "                showline=True,  \n",
    "                linecolor='black', \n",
    "                linewidth=1,\n",
    "                mirror=True\n",
    "            ),\n",
    "            yaxis=dict(\n",
    "                showgrid=True,\n",
    "                gridcolor='lightgray',\n",
    "                zeroline=False,\n",
    "                showline=True,\n",
    "                linecolor='black',\n",
    "                linewidth=1,\n",
    "                mirror=True\n",
    "            )\n",
    "        )\n",
    "        fig.show()\n",
    "        fig.write_image(f'UMAP_HAMBI_top150_features/UMAP_{n}_{m}_C4.png')\n",
    "        fig.write_html(f'UMAP_HAMBI_top150_features/UMAP_{n}_{m}_C4.html')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9f49c302-815e-4aec-a248-ee2e1a1a79ac",
   "metadata": {},
   "source": [
    "### Extract clusters and redraw UMAP with subset of data\n",
    "#### Top feature data ðŸ” (150)\n",
    "##### C7"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "74d32a3b-d179-43bf-8448-6341ac4b6afe",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Extract based on UMAP1 (x axis) & UMAP2 (y axis) values:\n",
    "## C7\n",
    "top150_UMAP_df_C7 = top150_UMAP_df.loc[(top150_UMAP_df['UMAP1']>= 5) & (top150_UMAP_df['UMAP1']<= 5.7)\n",
    "    & (top150_UMAP_df['UMAP2']>= 7.7) & (top150_UMAP_df['UMAP2']<= 8.3)]\n",
    "\n",
    "# Check\n",
    "top150_UMAP_df_C7.head()\n",
    "\n",
    "# Save those contig IDs to extract from the 'df_top150'\n",
    "top150_UMAP_df_C7.index.to_list()\n",
    "\n",
    "df_top150_C7 = df_top150[df_top150.index.isin(top150_UMAP_df_C7.index)]\n",
    "print(df_top150_C7)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "47a14071-293c-4c56-aed9-959b0b086c7d",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Save contig IDs\n",
    "top150_C7_contigs = df_top150_C7.index.to_list()\n",
    "\n",
    "directory = 'UMAP_HAMBI_top150_features'\n",
    "file_path = os.path.join(directory, 'top150_C7_contigs.txt')\n",
    "\n",
    "with open(file_path, 'w') as file:\n",
    "    for item in top150_C7_contigs:\n",
    "        file.write(f\"{item}\\n\")\n",
    "\n",
    "\n",
    "# Save contig IDs\n",
    "top150_C7_contigs = df_top150_C7.index.to_list()\n",
    "\n",
    "directory = 'UMAP_HAMBI_top150_features'\n",
    "file_path = os.path.join(directory, 'top150_C7_contigs.txt')\n",
    "\n",
    "with open(file_path, 'w') as file:\n",
    "    for item in top150_C7_contigs:\n",
    "        file.write(f\"{item}\\n\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "13aa39df-beda-40f8-82f8-11514ff28ab9",
   "metadata": {},
   "source": [
    "### Extract clusters and redraw UMAP with subset of data\n",
    "#### Top feature data ðŸ” (150)\n",
    "##### C5"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "48a3e8b1-7e23-4b54-a504-731fa3fdbad6",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Extract based on UMAP1 (x axis) & UMAP2 (y axis) values:\n",
    "## C5\n",
    "top150_UMAP_df_C5 = top150_UMAP_df.loc[(top150_UMAP_df['UMAP1']>= 14.5) & (top150_UMAP_df['UMAP1']<= 17)\n",
    "    & (top150_UMAP_df['UMAP2']>= 9) & (top150_UMAP_df['UMAP2']<= 11)]\n",
    "\n",
    "# Check\n",
    "top150_UMAP_df_C5.head()\n",
    "\n",
    "# Save those contig IDs to extract from the 'df_top150'\n",
    "top150_UMAP_df_C5.index.to_list()\n",
    "\n",
    "df_top150_C5 = df_top150[df_top150.index.isin(top150_UMAP_df_C5.index)]\n",
    "print(df_top150_C5)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "be2d9ae7-6b71-46fb-8c17-ea6486597f27",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Save contig IDs\n",
    "top150_C5_contigs = df_top150_C5.index.to_list()\n",
    "\n",
    "directory = 'UMAP_HAMBI_top150_features'\n",
    "file_path = os.path.join(directory, 'top150_C5_contigs.txt')\n",
    "\n",
    "with open(file_path, 'w') as file:\n",
    "    for item in top150_C5_contigs:\n",
    "        file.write(f\"{item}\\n\")\n",
    "\n",
    "\n",
    "# Save contig IDs\n",
    "top150_C5_contigs = df_top150_C5.index.to_list()\n",
    "\n",
    "directory = 'UMAP_HAMBI_top150_features'\n",
    "file_path = os.path.join(directory, 'top150_C5_contigs.txt')\n",
    "\n",
    "with open(file_path, 'w') as file:\n",
    "    for item in top150_C5_contigs:\n",
    "        file.write(f\"{item}\\n\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "46399a73-297a-4eee-958a-48f2de37ddb7",
   "metadata": {},
   "outputs": [],
   "source": [
    "n_neighbors = [30]\n",
    "min_dist = [0.1]\n",
    "color_dict = [\"#FF0000\"]\n",
    "\n",
    "custom_colors = color_dict\n",
    "for n in n_neighbors:\n",
    "    for m in min_dist:\n",
    "        reducer = umap.UMAP(n_neighbors=n, min_dist=m, random_state=seed)\n",
    "        embedding = reducer.fit_transform(df_top150_C5.iloc[:, :-17])\n",
    "        top150_UMAP_df_C5  = pd.DataFrame({\n",
    "            'UMAP1': embedding[:, 0],\n",
    "            'UMAP2': embedding[:, 1],\n",
    "            'Domain to species': df_top150_C5['Domain to species'],\n",
    "            'Domain to genus': df_top150_C5['Domain to genus'],\n",
    "            'element': df_top150_C5['element'],\n",
    "            'contig': df_top150_C5.index,\n",
    "        })\n",
    "        # clustering k defined choosing the k by looking at the UMAP \n",
    "        kmeans = KMeans(n_clusters=11, random_state=seed)\n",
    "        \n",
    "        k_labels = kmeans.fit_predict(embedding)\n",
    "        \n",
    "        top150_UMAP_df_C5['cluster'] = k_labels\n",
    "        \n",
    "        column = 'Domain to species'\n",
    "        top150_UMAP_df_C5['Domain to species'] = top150_UMAP_df_C5['Domain to species'].fillna('NA')\n",
    "\n",
    "        value_counts = top150_UMAP_df_C5[column].value_counts().to_dict()\n",
    "\n",
    "        top150_UMAP_df_C5[f'{column}_with_counts'] = top150_UMAP_df_C5[column].apply(lambda x: f'{x} ({value_counts[x]})')\n",
    "\n",
    "        top150_UMAP_df_C5['count'] = top150_UMAP_df_C5[column].map(value_counts)\n",
    "        top150_UMAP_df_C5 = top150_UMAP_df_C5.sort_values(by=['count'], ascending=False)\n",
    "\n",
    "        top150_UMAP_df_C5[f'{column}_with_counts'] = pd.Categorical(\n",
    "            top150_UMAP_df_C5[f'{column}_with_counts'],\n",
    "            categories=top150_UMAP_df_C5[f'{column}_with_counts'].unique(),\n",
    "            ordered=True\n",
    "        )\n",
    "        \n",
    "        fig = px.scatter(top150_UMAP_df_C5, \n",
    "                            x='UMAP1', \n",
    "                            y='UMAP2', \n",
    "                            labels={f'{column}_with_counts': column},\n",
    "                            color=f'{column}_with_counts', \n",
    "                            title=f' HAMBI - UMAP with n_neighbors={n}, min_dist={m}', \n",
    "                            color_discrete_sequence=custom_colors,\n",
    "                            hover_data={'cluster': True, 'Domain to species': True, 'Domain to genus': True, 'element': True, 'contig': True} )\n",
    "        title = f' HAMBI - UMAP with n_neighbors={n}, min_dist={m}'\n",
    "        fig.update_traces(marker=dict(size=10))\n",
    "        fig.update_layout(\n",
    "            height=1700,\n",
    "            width=1200,\n",
    "            title_text=title,\n",
    "            showlegend=True,\n",
    "            legend=dict(\n",
    "                x=0.5,\n",
    "                y=-0.1,\n",
    "                traceorder=\"normal\",\n",
    "                xanchor='center',\n",
    "                yanchor='top',\n",
    "                orientation='h'\n",
    "            ),\n",
    "            template='simple_white',\n",
    "            xaxis=dict(\n",
    "                showgrid=True,\n",
    "                gridcolor='lightgray',\n",
    "                zeroline=False,\n",
    "                showline=True,  \n",
    "                linecolor='black', \n",
    "                linewidth=1,\n",
    "                mirror=True\n",
    "            ),\n",
    "            yaxis=dict(\n",
    "                showgrid=True,\n",
    "                gridcolor='lightgray',\n",
    "                zeroline=False,\n",
    "                showline=True,\n",
    "                linecolor='black',\n",
    "                linewidth=1,\n",
    "                mirror=True\n",
    "            )\n",
    "        )\n",
    "        fig.show()\n",
    "        fig.write_image(f'UMAP_HAMBI_top150_features/UMAP_{n}_{m}_C5.png')\n",
    "        fig.write_html(f'UMAP_HAMBI_top150_features/UMAP_{n}_{m}_C5.html')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "fa5496e4-e665-41f3-9d1e-fbff99a18475",
   "metadata": {},
   "source": [
    "### Extract clusters and redraw UMAP with subset of data\n",
    "#### Top feature data ðŸ” (150)\n",
    "##### C6"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f6a4c9c5-41d6-4acd-bce7-b3df48a23779",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Extract based on UMAP1 (x axis) & UMAP2 (y axis) values:\n",
    "## C6\n",
    "top150_UMAP_df_C6 = top150_UMAP_df.loc[(top150_UMAP_df['UMAP1']>= 10) & (top150_UMAP_df['UMAP1']<= 11)\n",
    "    & (top150_UMAP_df['UMAP2']>= 9) & (top150_UMAP_df['UMAP2']<= 10)]\n",
    "\n",
    "# Check\n",
    "top150_UMAP_df_C6.head()\n",
    "\n",
    "# Save those contig IDs to extract from the 'df_top150'\n",
    "top150_UMAP_df_C6.index.to_list()\n",
    "\n",
    "df_top150_C6 = df_top150[df_top150.index.isin(top150_UMAP_df_C6.index)]\n",
    "print(df_top150_C6)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e0a35253-b89b-476f-a64f-7b0534595553",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Save contig IDs\n",
    "top150_C6_contigs = df_top150_C6.index.to_list()\n",
    "\n",
    "directory = 'UMAP_HAMBI_top150_features'\n",
    "file_path = os.path.join(directory, 'top150_C6_contigs.txt')\n",
    "\n",
    "with open(file_path, 'w') as file:\n",
    "    for item in top150_C6_contigs:\n",
    "        file.write(f\"{item}\\n\")\n",
    "\n",
    "\n",
    "# Save contig IDs\n",
    "top150_C6_contigs = df_top150_C6.index.to_list()\n",
    "\n",
    "directory = 'UMAP_HAMBI_top150_features'\n",
    "file_path = os.path.join(directory, 'top150_C6_contigs.txt')\n",
    "\n",
    "with open(file_path, 'w') as file:\n",
    "    for item in top150_C6_contigs:\n",
    "        file.write(f\"{item}\\n\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ec7c09c7-2efe-4d77-bdfe-08a7f22e384c",
   "metadata": {},
   "source": [
    "### Extract clusters and redraw UMAP with subset of data\n",
    "#### Top feature data ðŸ” (150)\n",
    "##### C8"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0657a93b-5927-42e9-823b-0ec0c0965124",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Extract based on UMAP1 (x axis) & UMAP2 (y axis) values:\n",
    "## C8\n",
    "top150_UMAP_df_C8 = top150_UMAP_df.loc[(top150_UMAP_df['UMAP1']>= 6.7) & (top150_UMAP_df['UMAP1']<= 6.9)\n",
    "    & (top150_UMAP_df['UMAP2']>= 8) & (top150_UMAP_df['UMAP2']<= 8.3)]\n",
    "\n",
    "# Check\n",
    "top150_UMAP_df_C8.head()\n",
    "\n",
    "# Save those contig IDs to extract from the 'df_top150'\n",
    "top150_UMAP_df_C8.index.to_list()\n",
    "\n",
    "df_top150_C8 = df_top150[df_top150.index.isin(top150_UMAP_df_C8.index)]\n",
    "print(df_top150_C8)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3218bdb4-1322-45cb-98f7-41fb9aded215",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Save contig IDs\n",
    "top150_C8_contigs = df_top150_C8.index.to_list()\n",
    "\n",
    "directory = 'UMAP_HAMBI_top150_features'\n",
    "file_path = os.path.join(directory, 'top150_C8_contigs.txt')\n",
    "\n",
    "with open(file_path, 'w') as file:\n",
    "    for item in top150_C8_contigs:\n",
    "        file.write(f\"{item}\\n\")\n",
    "\n",
    "\n",
    "# Save contig IDs\n",
    "top150_C8_contigs = df_top150_C8.index.to_list()\n",
    "\n",
    "directory = 'UMAP_HAMBI_top150_features'\n",
    "file_path = os.path.join(directory, 'top150_C8_contigs.txt')\n",
    "\n",
    "with open(file_path, 'w') as file:\n",
    "    for item in top150_C8_contigs:\n",
    "        file.write(f\"{item}\\n\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "cdfbb027-88e7-4d99-a1c6-1ff36cc5653c",
   "metadata": {},
   "source": [
    "### Extract clusters and redraw UMAP with subset of data\n",
    "#### Top feature data ðŸ” (150)\n",
    "##### C9"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8d2db378-aca1-4f89-8652-9f4d9ab9d0b0",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Extract based on UMAP1 (x axis) & UMAP2 (y axis) values:\n",
    "## C9\n",
    "top150_UMAP_df_C9 = top150_UMAP_df.loc[(top150_UMAP_df['UMAP1']>= 5.4) & (top150_UMAP_df['UMAP1']<= 5.6)\n",
    "    & (top150_UMAP_df['UMAP2']>= 5.2) & (top150_UMAP_df['UMAP2']<= 5.5)]\n",
    "\n",
    "# Check\n",
    "top150_UMAP_df_C9.head()\n",
    "\n",
    "# Save those contig IDs to extract from the 'df_top150'\n",
    "top150_UMAP_df_C9.index.to_list()\n",
    "\n",
    "df_top150_C9 = df_top150[df_top150.index.isin(top150_UMAP_df_C9.index)]\n",
    "print(df_top150_C9)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b00b714d-1a44-43ff-a43c-1e694b99fc84",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Save contig IDs\n",
    "top150_C9_contigs = df_top150_C9.index.to_list()\n",
    "\n",
    "directory = 'UMAP_HAMBI_top150_features'\n",
    "file_path = os.path.join(directory, 'top150_C9_contigs.txt')\n",
    "\n",
    "with open(file_path, 'w') as file:\n",
    "    for item in top150_C9_contigs:\n",
    "        file.write(f\"{item}\\n\")\n",
    "\n",
    "\n",
    "# Save contig IDs\n",
    "top150_C9_contigs = df_top150_C9.index.to_list()\n",
    "\n",
    "directory = 'UMAP_HAMBI_top150_features'\n",
    "file_path = os.path.join(directory, 'top150_C9_contigs.txt')\n",
    "\n",
    "with open(file_path, 'w') as file:\n",
    "    for item in top150_C9_contigs:\n",
    "        file.write(f\"{item}\\n\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "edc47f3c-9200-486d-9da6-6bd4224d309a",
   "metadata": {},
   "source": [
    "### Extract clusters and redraw UMAP with subset of data\n",
    "#### Top feature data ðŸ” (150)\n",
    "##### C10"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "017f2175-0c92-4672-a280-d1dfa2b6836b",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Extract based on UMAP1 (x axis) & UMAP2 (y axis) values:\n",
    "## C10\n",
    "top150_UMAP_df_C10 = top150_UMAP_df.loc[(top150_UMAP_df['UMAP1']>= 10) & (top150_UMAP_df['UMAP1']<= 15)\n",
    "    & (top150_UMAP_df['UMAP2']>= -6) & (top150_UMAP_df['UMAP2']<= -4)]\n",
    "\n",
    "# Check\n",
    "top150_UMAP_df_C10.head()\n",
    "\n",
    "# Save those contig IDs to extract from the 'df_top150'\n",
    "top150_UMAP_df_C10.index.to_list()\n",
    "\n",
    "df_top150_C10 = df_top150[df_top150.index.isin(top150_UMAP_df_C10.index)]\n",
    "print(df_top150_C10)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9bf144b1-e595-4e55-94ae-8fdec16bc952",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Save contig IDs\n",
    "top150_C10_contigs = df_top150_C10.index.to_list()\n",
    "\n",
    "directory = 'UMAP_HAMBI_top150_features'\n",
    "file_path = os.path.join(directory, 'top150_C10_contigs.txt')\n",
    "\n",
    "with open(file_path, 'w') as file:\n",
    "    for item in top150_C10_contigs:\n",
    "        file.write(f\"{item}\\n\")\n",
    "\n",
    "\n",
    "# Save contig IDs\n",
    "top150_C10_contigs = df_top150_C10.index.to_list()\n",
    "\n",
    "directory = 'UMAP_HAMBI_top150_features'\n",
    "file_path = os.path.join(directory, 'top150_C10_contigs.txt')\n",
    "\n",
    "with open(file_path, 'w') as file:\n",
    "    for item in top150_C10_contigs:\n",
    "        file.write(f\"{item}\\n\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3f90ade2-4ccc-467b-83c8-eac0de7c5e29",
   "metadata": {},
   "source": [
    "## Using contig lengths\n",
    "### All data ðŸ’ "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7b42cc18-0c7e-4d94-8c46-0478f765dbae",
   "metadata": {},
   "outputs": [],
   "source": [
    "df['contig_length_sqrt'] = np.sqrt(df['contig_length'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f8f19c05-0bf5-4c8f-93bb-ff0295ff432d",
   "metadata": {},
   "outputs": [],
   "source": [
    "df_lengths = df.copy()\n",
    "df_lengths.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b731d00d-e3a6-4653-b0ca-6bb3c55fafa5",
   "metadata": {},
   "outputs": [],
   "source": [
    "print(df_lengths.iloc[:, :-19])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5c59af12-1c4b-4414-8811-a90ce225c89a",
   "metadata": {},
   "outputs": [],
   "source": [
    "n_neighbors = [20]\n",
    "min_dist = [0.5]\n",
    "\n",
    "for n in n_neighbors:\n",
    "    for m in min_dist:\n",
    "        reducer = umap.UMAP(n_neighbors=n, min_dist=m, random_state=seed)\n",
    "        embedding = reducer.fit_transform(df_lengths.iloc[:, :-19])\n",
    "        lengths_UMAP_df  = pd.DataFrame({\n",
    "            'UMAP1': embedding[:, 0],\n",
    "            'UMAP2': embedding[:, 1],\n",
    "            'Domain to species': df_lengths['Domain to species'],\n",
    "            'Domain to genus': df_lengths['Domain to genus'],\n",
    "            'element': df_lengths['element'],\n",
    "            'contig': df_lengths.index,\n",
    "            'contig_length_sqrt':df_lengths['contig_length_sqrt']\n",
    "        })\n",
    "        # clustering k defined choosing the k by looking at the UMAP \n",
    "        kmeans = KMeans(n_clusters=11, random_state=seed)\n",
    "        \n",
    "        k_labels = kmeans.fit_predict(embedding)\n",
    "        \n",
    "        lengths_UMAP_df['cluster'] = k_labels\n",
    "        \n",
    "        column = 'Domain to species'\n",
    "        lengths_UMAP_df['Domain to species'] = lengths_UMAP_df['Domain to species'].fillna('NA')\n",
    "\n",
    "        value_counts = lengths_UMAP_df[column].value_counts().to_dict()\n",
    "\n",
    "        lengths_UMAP_df[f'{column}_with_counts'] = lengths_UMAP_df[column].apply(lambda x: f'{x} ({value_counts[x]})')\n",
    "\n",
    "        lengths_UMAP_df['count'] = lengths_UMAP_df[column].map(value_counts)\n",
    "        lengths_UMAP_df = lengths_UMAP_df.sort_values(by=['count'], ascending=False)\n",
    "\n",
    "        lengths_UMAP_df[f'{column}_with_counts'] = pd.Categorical(\n",
    "            lengths_UMAP_df[f'{column}_with_counts'],\n",
    "            categories=lengths_UMAP_df[f'{column}_with_counts'].unique(),\n",
    "            ordered=True\n",
    "        )\n",
    "        \n",
    "        fig = px.scatter(lengths_UMAP_df, \n",
    "                            x='UMAP1', \n",
    "                            y='UMAP2', \n",
    "                            labels={f'{column}_with_counts': column},\n",
    "                            color='contig_length_sqrt',\n",
    "                            title=f' HAMBI - UMAP with n_neighbors={n}, min_dist={m}', \n",
    "                            color_continuous_scale=px.colors.sequential.Rainbow,\n",
    "                            hover_data={'cluster': True, 'Domain to species': True, 'Domain to genus': True,\n",
    "                                        'element': True, 'contig': True},\n",
    "                         size='contig_length_sqrt'\n",
    "        )\n",
    "        title = f' HAMBI - UMAP with n_neighbors={n}, min_dist={m}'\n",
    "        fig.update_layout(\n",
    "            height=1700,\n",
    "            width=1200,\n",
    "            title_text=title,\n",
    "            showlegend=True,\n",
    "            legend=dict(\n",
    "                x=0.5,\n",
    "                y=-0.05,\n",
    "                traceorder=\"normal\",\n",
    "                xanchor='center',\n",
    "                yanchor='top',\n",
    "                orientation='h'\n",
    "            ),\n",
    "            template='simple_white',\n",
    "            xaxis=dict(\n",
    "                showgrid=True,\n",
    "                gridcolor='lightgray',\n",
    "                zeroline=False,\n",
    "                showline=True,  \n",
    "                linecolor='black', \n",
    "                linewidth=1,\n",
    "                mirror=True\n",
    "            ),\n",
    "            yaxis=dict(\n",
    "                showgrid=True,\n",
    "                gridcolor='lightgray',\n",
    "                zeroline=False,\n",
    "                showline=True,\n",
    "                linecolor='black',\n",
    "                linewidth=1,\n",
    "                mirror=True\n",
    "            )\n",
    "        )\n",
    "        fig.show()\n",
    "        #fig.write_image(f'UMAP_HAMBI/UMAP_{n}_{m}_contig_lengths_sqrt.png')\n",
    "        #fig.write_html(f'UMAP_HAMBI/UMAP_{n}_{m}_contig_lengths_sqrt.html')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "afe25a20-44f1-4daf-9d8e-c14775edc8b8",
   "metadata": {},
   "source": [
    "## Using mapped modifications\n",
    "### All data ðŸ’ "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "45d87b34-1097-4386-90b4-3e93fcc4616e",
   "metadata": {},
   "outputs": [],
   "source": [
    "file_path = '/scratch/project_2006608/Methylation/HAMBI_data/merged_data.tsv'\n",
    "df = pd.read_csv(file_path, sep='\\t', index_col=0, low_memory=False)\n",
    "\n",
    "# remove rows where df.iloc[:, :-17] meaning no methylation detected or less than 20 detected since those are converted to 0 filled matrices\n",
    "df = df.loc[(df.iloc[:, :-17] != 0).any(axis=1)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "aed217bf-227a-4a1b-937e-b0c1c7d3014d",
   "metadata": {},
   "outputs": [],
   "source": [
    "## Bring the contig lengths from Puhti:\n",
    "file_path = '/scratch/project_2006608/Methylation/HAMBI_data/metagenomic_assembly/contigs_mod_counts.tsv'\n",
    "\n",
    "df_mod_counts = pd.read_csv(file_path, sep='\\t', index_col=0, header=0, low_memory=False)\n",
    "print(df_mod_counts.shape[0])\n",
    "df_mod_counts.head()\n",
    "\n",
    "df_mod_counts = np.log(df_mod_counts)\n",
    "df_mod_counts.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "fe2ce4c4-67fe-4f4e-8819-0176fa2542ad",
   "metadata": {},
   "outputs": [],
   "source": [
    "## Append to merged_data.tsv\n",
    "df_ext = df.copy()\n",
    "\n",
    "# Reorder to match\n",
    "df_ordered = df_mod_counts.loc[df_ext.index]\n",
    "\n",
    "df_mod_counts = pd.concat([df_ext, df_ordered], axis=1)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f0123348-f090-4aa0-ba5d-45b2844af0f8",
   "metadata": {},
   "source": [
    "### Add contig lengths"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "17d02921-abbc-414f-b68f-82a74662da9b",
   "metadata": {},
   "outputs": [],
   "source": [
    "## Bring the contig lengths from Puhti:\n",
    "file_path = '/scratch/project_2006608/Methylation/HAMBI_data/metagenomic_assembly/contigs_lengths.tsv'\n",
    "\n",
    "df_lengths = pd.read_csv(file_path, sep='\\t', index_col=0, header=0, low_memory=False)\n",
    "print(df_lengths.shape[0])\n",
    "df_lengths.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6d7eb246-514f-4888-8719-e8b32563a2f0",
   "metadata": {},
   "outputs": [],
   "source": [
    "## Append to merged_data.tsv\n",
    "df_ext = df_mod_counts.copy()\n",
    "\n",
    "# Reorder to match\n",
    "df_ordered = df_lengths.loc[df_ext.index]\n",
    "\n",
    "df_mod_counts = pd.concat([df_ext, df_ordered], axis=1)\n",
    "# transform\n",
    "df_mod_counts['contig_length_sqrt'] = np.sqrt(df_mod_counts['contig_length'])\n",
    "df_mod_counts.tail()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d2db9fc5-379e-4bee-98b9-2eae56b25e88",
   "metadata": {},
   "source": [
    "#### Draw mod counts\n",
    "#### All data ðŸ’ "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1159d8cc-68d1-4005-a3b3-c2e9a5dda150",
   "metadata": {},
   "outputs": [],
   "source": [
    "n_neighbors = [20]\n",
    "min_dist = [0.5]\n",
    "\n",
    "for n in n_neighbors:\n",
    "    for m in min_dist:\n",
    "        reducer = umap.UMAP(n_neighbors=n, min_dist=m, random_state=seed)\n",
    "        embedding = reducer.fit_transform(df_mod_counts.iloc[:, :-20])\n",
    "        mod_counts_UMAP_df  = pd.DataFrame({\n",
    "            'UMAP1': embedding[:, 0],\n",
    "            'UMAP2': embedding[:, 1],\n",
    "            'Domain to species': df_mod_counts['Domain to species'],\n",
    "            'Domain to genus': df_mod_counts['Domain to genus'],\n",
    "            'element': df_mod_counts['element'],\n",
    "            'contig': df_mod_counts.index,\n",
    "            'mod_count':df_mod_counts['mod_count'],\n",
    "            'contig_length_sqrt':df_mod_counts['contig_length_sqrt']\n",
    "        })\n",
    "        # clustering k defined choosing the k by looking at the UMAP \n",
    "        kmeans = KMeans(n_clusters=11, random_state=seed)\n",
    "        \n",
    "        k_labels = kmeans.fit_predict(embedding)\n",
    "        \n",
    "        mod_counts_UMAP_df['cluster'] = k_labels\n",
    "        \n",
    "        column = 'Domain to species'\n",
    "        mod_counts_UMAP_df['Domain to species'] = mod_counts_UMAP_df['Domain to species'].fillna('NA')\n",
    "\n",
    "        value_counts = mod_counts_UMAP_df[column].value_counts().to_dict()\n",
    "\n",
    "        mod_counts_UMAP_df[f'{column}_with_counts'] = mod_counts_UMAP_df[column].apply(lambda x: f'{x} ({value_counts[x]})')\n",
    "\n",
    "        mod_counts_UMAP_df['count'] = mod_counts_UMAP_df[column].map(value_counts)\n",
    "        mod_counts_UMAP_df = mod_counts_UMAP_df.sort_values(by=['count'], ascending=False)\n",
    "\n",
    "        mod_counts_UMAP_df[f'{column}_with_counts'] = pd.Categorical(\n",
    "            mod_counts_UMAP_df[f'{column}_with_counts'],\n",
    "            categories=mod_counts_UMAP_df[f'{column}_with_counts'].unique(),\n",
    "            ordered=True\n",
    "        )\n",
    "        \n",
    "        fig = px.scatter(mod_counts_UMAP_df, \n",
    "                            x='UMAP1', \n",
    "                            y='UMAP2', \n",
    "                            labels={f'{column}_with_counts': column},\n",
    "                            color='mod_count',\n",
    "                            title=f' HAMBI - UMAP with n_neighbors={n}, min_dist={m}', \n",
    "                            color_continuous_scale=px.colors.sequential.Rainbow,\n",
    "                            hover_data={'cluster': True, 'Domain to species': True, 'Domain to genus': True,\n",
    "                                        'element': True, 'contig': True},\n",
    "                            size='contig_length_sqrt')\n",
    "        title = f' HAMBI - UMAP with n_neighbors={n}, min_dist={m}'\n",
    "        fig.update_layout(\n",
    "            height=1700,\n",
    "            width=1200,\n",
    "            title_text=title,\n",
    "            showlegend=True,\n",
    "            legend=dict(\n",
    "                x=0.5,\n",
    "                y=-0.05,\n",
    "                traceorder=\"normal\",\n",
    "                xanchor='center',\n",
    "                yanchor='top',\n",
    "                orientation='h'\n",
    "            ),\n",
    "            template='simple_white',\n",
    "            xaxis=dict(\n",
    "                showgrid=True,\n",
    "                gridcolor='lightgray',\n",
    "                zeroline=False,\n",
    "                showline=True,  \n",
    "                linecolor='black', \n",
    "                linewidth=1,\n",
    "                mirror=True\n",
    "            ),\n",
    "            yaxis=dict(\n",
    "                showgrid=True,\n",
    "                gridcolor='lightgray',\n",
    "                zeroline=False,\n",
    "                showline=True,\n",
    "                linecolor='black',\n",
    "                linewidth=1,\n",
    "                mirror=True\n",
    "            )\n",
    "        )\n",
    "        fig.show()\n",
    "        #fig.write_image(f'UMAP_HAMBI/UMAP_{n}_{m}_contig_mod_counts_log_contig_length_sqrt.png')\n",
    "        #fig.write_html(f'UMAP_HAMBI/UMAP_{n}_{m}_contig_mod_counts_log_contig_length_sqrt.html')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b82e0440-a661-4df7-8016-91ad5d15e14e",
   "metadata": {},
   "source": [
    "## Redraw filtered data based on mod counts\n",
    "### All data ðŸ’ \n",
    "#### >=50, 100, 200, 500"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e0026bb5-01ec-4e66-b31c-b08e14a395f3",
   "metadata": {},
   "outputs": [],
   "source": [
    "## Bring the contig lengths from Puhti:\n",
    "file_path = '/scratch/project_2006608/Methylation/HAMBI_data/metagenomic_assembly/contigs_mod_counts.tsv'\n",
    "\n",
    "df_mod_counts = pd.read_csv(file_path, sep='\\t', index_col=0, header=0, low_memory=False)\n",
    "print(df_mod_counts.shape[0])\n",
    "df_mod_counts.head()\n",
    "\n",
    "## Append to merged_data.tsv\n",
    "df_ext = df.copy()\n",
    "\n",
    "# Reorder to match\n",
    "df_ordered = df_mod_counts.loc[df_ext.index]\n",
    "\n",
    "df_mod_counts = pd.concat([df_ext, df_ordered], axis=1)\n",
    "df_mod_counts.tail()\n",
    "\n",
    "df_mod_counts = df_mod_counts.loc[df_mod_counts['mod_count'] >= 50]\n",
    "#df_mod_counts = df_mod_counts.loc[df_mod_counts['mod_count'] >= 100]\n",
    "#df_mod_counts = df_mod_counts.loc[df_mod_counts['mod_count'] >= 200]\n",
    "#df_mod_counts = df_mod_counts.loc[df_mod_counts['mod_count'] >= 500]\n",
    "\n",
    "print(df_mod_counts.shape[0])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d9cba0be-c02b-4ee4-a875-0f9ee949cfea",
   "metadata": {},
   "outputs": [],
   "source": [
    "## Bring the contig lengths from Puhti:\n",
    "file_path = '/scratch/project_2006608/Methylation/HAMBI_data/metagenomic_assembly/contigs_lengths.tsv'\n",
    "\n",
    "df_lengths = pd.read_csv(file_path, sep='\\t', index_col=0, header=0, low_memory=False)\n",
    "print(df_lengths.shape[0])\n",
    "df_lengths.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3aeff150-595a-4014-b04b-9a6d3e67be7a",
   "metadata": {},
   "outputs": [],
   "source": [
    "## Append to merged_data.tsv\n",
    "df_ext = df_mod_counts.copy()\n",
    "\n",
    "# Reorder to match\n",
    "df_ordered = df_lengths.loc[df_ext.index]\n",
    "\n",
    "df_mod_counts = pd.concat([df_ext, df_ordered], axis=1)\n",
    "# transform\n",
    "df_mod_counts['contig_length_sqrt'] = np.sqrt(df_mod_counts['contig_length'])\n",
    "df_mod_counts.tail()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d3d6ad15-5f18-4c01-affb-c253a0b876ba",
   "metadata": {},
   "outputs": [],
   "source": [
    "#n_neighbors = [20, 30 ]\n",
    "#min_dist = [0.01, 0.1, 0.2]\n",
    "\n",
    "n_neighbors = [20]\n",
    "min_dist = [0.5]\n",
    "\n",
    "#n_neighbors = [20]\n",
    "#min_dist = [0.4]\n",
    "\n",
    "# 50\n",
    "color_dict = [\"#FF0000\", \"#00FF00\", \"#00FFFF\", \"#FF00FF\", \"#0000FF\", \"#800000\", \"#FFFF00\", \"#000080\", \"#008000\", \"#808000\", \"#008080\", \"#800080\", \"#3bccf7\", \"#FF4500\", \"#4682B4\", \"#DAA520\"]\n",
    "\n",
    "# 100\n",
    "#color_dict = [\"#00FF00\", \"#FF0000\", \"#00FFFF\", \"#000080\", \"#FF00FF\", \"#800000\", \"#0000FF\", \"#008000\", \"#FFFF00\",\"#008080\", \"#800080\", \"#808000\", \"#3bccf7\", \"#4682B4\", \"#DAA520\"]\n",
    "\n",
    "    ## no color\n",
    "#color_dict = [\"#FF0000\"]\n",
    "\n",
    "# 200\n",
    "#color_dict = [\"#00FF00\", \"#FF0000\", \"#00FFFF\", \"#000080\", \"#FF00FF\", \"#008080\", \"#800000\", \"#800080\", \"#008000\",\"#0000FF\", \"#3bccf7\", \"#FFFF00\", \"#4682B4\", \"#DAA520\"]\n",
    "\n",
    "# 500\n",
    "#color_dict = [\"#00FF00\", \"#FF0000\", \"#00FFFF\", \"#000080\", \"#008080\", \"#800080\", \"#800000\", \"#FF00FF\", \"#3bccf7\",\"#4682B4\", \"#008000\", \"#DAA520\"]\n",
    "\n",
    "custom_colors = color_dict\n",
    "\n",
    "for n in n_neighbors:\n",
    "    for m in min_dist:\n",
    "        reducer = umap.UMAP(n_neighbors=n, min_dist=m, random_state=seed)\n",
    "        embedding = reducer.fit_transform(df_mod_counts.iloc[:, :-20])\n",
    "        mod_counts_UMAP_df  = pd.DataFrame({\n",
    "            'UMAP1': embedding[:, 0],\n",
    "            'UMAP2': embedding[:, 1],\n",
    "            'Domain to species': df_mod_counts['Domain to species'],\n",
    "            'Domain to genus': df_mod_counts['Domain to genus'],\n",
    "            'element': df_mod_counts['element'],\n",
    "            'contig': df_mod_counts.index,\n",
    "            'contig_length_sqrt': df_mod_counts['contig_length_sqrt']\n",
    "        })\n",
    "        \n",
    "        column = 'Domain to species'\n",
    "        mod_counts_UMAP_df['Domain to species'] = mod_counts_UMAP_df['Domain to species'].fillna('NA')\n",
    "\n",
    "        value_counts = mod_counts_UMAP_df[column].value_counts().to_dict()\n",
    "\n",
    "        mod_counts_UMAP_df[f'{column}_with_counts'] = mod_counts_UMAP_df[column].apply(lambda x: f'{x} ({value_counts[x]})')\n",
    "\n",
    "        mod_counts_UMAP_df['count'] = mod_counts_UMAP_df[column].map(value_counts)\n",
    "        mod_counts_UMAP_df = mod_counts_UMAP_df.sort_values(by=['count'], ascending=False)\n",
    "\n",
    "        mod_counts_UMAP_df[f'{column}_with_counts'] = pd.Categorical(\n",
    "            mod_counts_UMAP_df[f'{column}_with_counts'],\n",
    "            categories=mod_counts_UMAP_df[f'{column}_with_counts'].unique(),\n",
    "            ordered=True\n",
    "        )\n",
    "        \n",
    "        fig = px.scatter(mod_counts_UMAP_df, \n",
    "                            x='UMAP1', \n",
    "                            y='UMAP2', \n",
    "                            labels={f'{column}_with_counts': column},\n",
    "                            color=f'{column}_with_counts', \n",
    "                            title=f' HAMBI - UMAP with n_neighbors={n}, min_dist={m}', \n",
    "                            color_discrete_sequence=custom_colors,\n",
    "                            hover_data={'Domain to species': True, 'Domain to genus': True,\n",
    "                                        'element': True, 'contig': True},\n",
    "                        size = 'contig_length_sqrt'\n",
    "                        )\n",
    "        title = f' HAMBI - UMAP with n_neighbors={n}, min_dist={m}'\n",
    "        fig.update_layout(\n",
    "            height=1700,\n",
    "            width=1200,\n",
    "            title_text=title,\n",
    "            showlegend=True,\n",
    "            legend=dict(\n",
    "                x=0.5,\n",
    "                y=-0.1,\n",
    "                traceorder=\"normal\",\n",
    "                xanchor='center',\n",
    "                yanchor='top',\n",
    "                orientation='h'\n",
    "            ),\n",
    "            template='simple_white',\n",
    "            xaxis=dict(\n",
    "                showgrid=True,\n",
    "                gridcolor='lightgray',\n",
    "                zeroline=False,\n",
    "                showline=True,  \n",
    "                linecolor='black', \n",
    "                linewidth=1,\n",
    "                mirror=True\n",
    "            ),\n",
    "            yaxis=dict(\n",
    "                showgrid=True,\n",
    "                gridcolor='lightgray',\n",
    "                zeroline=False,\n",
    "                showline=True,\n",
    "                linecolor='black',\n",
    "                linewidth=1,\n",
    "                mirror=True\n",
    "            )\n",
    "        )\n",
    "        fig.show()\n",
    "        #fig.write_image(f'UMAP_HAMBI/UMAP_{n}_{m}_filtered_by_50_mod_counts.png')\n",
    "        #fig.write_html(f'UMAP_HAMBI/UMAP_{n}_{m}_filtered_by_50_mod_counts.html')\n",
    "        \n",
    "        #fig.write_image(f'UMAP_HAMBI/UMAP_{n}_{m}_filtered_by_100_mod_counts.png')\n",
    "        #fig.write_html(f'UMAP_HAMBI/UMAP_{n}_{m}_filtered_by_100_mod_counts.html')\n",
    "        #fig.write_image(f'UMAP_HAMBI/UMAP_{n}_{m}_filtered_by_100_mod_counts_no_color.png')\n",
    "        #fig.write_html(f'UMAP_HAMBI/UMAP_{n}_{m}_filtered_by_100_mod_counts_no_color.html')\n",
    "        #fig.write_image(f'UMAP_HAMBI/UMAP_{n}_{m}_filtered_by_200_mod_counts.png')\n",
    "        #fig.write_html(f'UMAP_HAMBI/UMAP_{n}_{m}_filtered_by_200_mod_counts.html')\n",
    "        #fig.write_image(f'UMAP_HAMBI/UMAP_{n}_{m}_filtered_by_500_mod_counts.png')\n",
    "        #fig.write_html(f'UMAP_HAMBI/UMAP_{n}_{m}_filtered_by_500_mod_counts.html')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "15435231-4e5a-4d42-9ab4-08b283c95335",
   "metadata": {},
   "source": [
    "### Save data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "id": "bcea13e8-71dd-4cef-b724-d1bb9c0b77e4",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>UMAP1</th>\n",
       "      <th>UMAP2</th>\n",
       "      <th>Domain to species</th>\n",
       "      <th>Domain to genus</th>\n",
       "      <th>element</th>\n",
       "      <th>contig</th>\n",
       "      <th>contig_length_sqrt</th>\n",
       "      <th>Domain to species_with_counts</th>\n",
       "      <th>count</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>bcAd1063T--bcAd1063T_ptg000223l</th>\n",
       "      <td>-1.760182</td>\n",
       "      <td>5.015911</td>\n",
       "      <td>NA</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>bcAd1063T--bcAd1063T_ptg000223l</td>\n",
       "      <td>363.877727</td>\n",
       "      <td>NA (266)</td>\n",
       "      <td>266</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>bcAd1023T--bcAd1023T_ptg000162l</th>\n",
       "      <td>0.189562</td>\n",
       "      <td>-7.648015</td>\n",
       "      <td>NA</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>bcAd1023T--bcAd1023T_ptg000162l</td>\n",
       "      <td>353.213816</td>\n",
       "      <td>NA (266)</td>\n",
       "      <td>266</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>bcAd1063T--bcAd1063T_ptg000318l</th>\n",
       "      <td>-2.079180</td>\n",
       "      <td>5.882244</td>\n",
       "      <td>NA</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>bcAd1063T--bcAd1063T_ptg000318l</td>\n",
       "      <td>179.938879</td>\n",
       "      <td>NA (266)</td>\n",
       "      <td>266</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>bcAd1063T--bcAd1063T_ptg000224l</th>\n",
       "      <td>-2.305873</td>\n",
       "      <td>6.072272</td>\n",
       "      <td>NA</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>bcAd1063T--bcAd1063T_ptg000224l</td>\n",
       "      <td>241.679540</td>\n",
       "      <td>NA (266)</td>\n",
       "      <td>266</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>bcAd1063T--bcAd1063T_ptg000364l</th>\n",
       "      <td>-1.512699</td>\n",
       "      <td>5.406713</td>\n",
       "      <td>NA</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>bcAd1063T--bcAd1063T_ptg000364l</td>\n",
       "      <td>199.376528</td>\n",
       "      <td>NA (266)</td>\n",
       "      <td>266</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                    UMAP1     UMAP2 Domain to species  \\\n",
       "0                                                                       \n",
       "bcAd1063T--bcAd1063T_ptg000223l -1.760182  5.015911                NA   \n",
       "bcAd1023T--bcAd1023T_ptg000162l  0.189562 -7.648015                NA   \n",
       "bcAd1063T--bcAd1063T_ptg000318l -2.079180  5.882244                NA   \n",
       "bcAd1063T--bcAd1063T_ptg000224l -2.305873  6.072272                NA   \n",
       "bcAd1063T--bcAd1063T_ptg000364l -1.512699  5.406713                NA   \n",
       "\n",
       "                                Domain to genus element  \\\n",
       "0                                                         \n",
       "bcAd1063T--bcAd1063T_ptg000223l             NaN     NaN   \n",
       "bcAd1023T--bcAd1023T_ptg000162l             NaN     NaN   \n",
       "bcAd1063T--bcAd1063T_ptg000318l             NaN     NaN   \n",
       "bcAd1063T--bcAd1063T_ptg000224l             NaN     NaN   \n",
       "bcAd1063T--bcAd1063T_ptg000364l             NaN     NaN   \n",
       "\n",
       "                                                          contig  \\\n",
       "0                                                                  \n",
       "bcAd1063T--bcAd1063T_ptg000223l  bcAd1063T--bcAd1063T_ptg000223l   \n",
       "bcAd1023T--bcAd1023T_ptg000162l  bcAd1023T--bcAd1023T_ptg000162l   \n",
       "bcAd1063T--bcAd1063T_ptg000318l  bcAd1063T--bcAd1063T_ptg000318l   \n",
       "bcAd1063T--bcAd1063T_ptg000224l  bcAd1063T--bcAd1063T_ptg000224l   \n",
       "bcAd1063T--bcAd1063T_ptg000364l  bcAd1063T--bcAd1063T_ptg000364l   \n",
       "\n",
       "                                 contig_length_sqrt  \\\n",
       "0                                                     \n",
       "bcAd1063T--bcAd1063T_ptg000223l          363.877727   \n",
       "bcAd1023T--bcAd1023T_ptg000162l          353.213816   \n",
       "bcAd1063T--bcAd1063T_ptg000318l          179.938879   \n",
       "bcAd1063T--bcAd1063T_ptg000224l          241.679540   \n",
       "bcAd1063T--bcAd1063T_ptg000364l          199.376528   \n",
       "\n",
       "                                Domain to species_with_counts  count  \n",
       "0                                                                     \n",
       "bcAd1063T--bcAd1063T_ptg000223l                      NA (266)    266  \n",
       "bcAd1023T--bcAd1023T_ptg000162l                      NA (266)    266  \n",
       "bcAd1063T--bcAd1063T_ptg000318l                      NA (266)    266  \n",
       "bcAd1063T--bcAd1063T_ptg000224l                      NA (266)    266  \n",
       "bcAd1063T--bcAd1063T_ptg000364l                      NA (266)    266  "
      ]
     },
     "execution_count": 26,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Check\n",
    "df_mod_counts.head()\n",
    "mod_counts_UMAP_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "id": "2e1e89b7-6a16-4b22-997e-a87078ece177",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Save data\n",
    "df_mod_counts.to_csv('UMAP_HAMBI_above50/df_mod_counts.csv', sep='\\t', index=True)\n",
    "mod_counts_UMAP_df.to_csv('UMAP_HAMBI_above50/mod_counts_UMAP_df.csv', sep='\\t', index=True)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ba6f959d-f137-4c73-9d67-c589ce67c21b",
   "metadata": {},
   "source": [
    "#### Draw mod counts\n",
    "#### > 50 mod counts"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a84d189b-faa7-4396-8b2f-956fe242b07e",
   "metadata": {},
   "outputs": [],
   "source": [
    "df_mod_counts['mod_count_log'] = np.log2(df_mod_counts['mod_count'])\n",
    "df_mod_counts.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ac954da0-465a-4118-bfdf-15fb9c74828e",
   "metadata": {},
   "outputs": [],
   "source": [
    "n_neighbors = [20]\n",
    "min_dist = [0.5]\n",
    "\n",
    "for n in n_neighbors:\n",
    "    for m in min_dist:\n",
    "        reducer = umap.UMAP(n_neighbors=n, min_dist=m, random_state=seed)\n",
    "        embedding = reducer.fit_transform(df_mod_counts.iloc[:, :-21])\n",
    "        mod_counts_UMAP_df  = pd.DataFrame({\n",
    "            'UMAP1': embedding[:, 0],\n",
    "            'UMAP2': embedding[:, 1],\n",
    "            'Domain to species': df_mod_counts['Domain to species'],\n",
    "            'Domain to genus': df_mod_counts['Domain to genus'],\n",
    "            'element': df_mod_counts['element'],\n",
    "            'contig': df_mod_counts.index,\n",
    "            'mod_count':df_mod_counts['mod_count'],\n",
    "            'mod_count_log':df_mod_counts['mod_count_log'],\n",
    "            'contig_length_sqrt':df_mod_counts['contig_length_sqrt']\n",
    "        })\n",
    "        \n",
    "        fig = px.scatter(mod_counts_UMAP_df, \n",
    "                            x='UMAP1', \n",
    "                            y='UMAP2', \n",
    "                            labels={f'{column}_with_counts': column},\n",
    "                            color='mod_count_log',\n",
    "                            title=f' HAMBI - UMAP with n_neighbors={n}, min_dist={m}', \n",
    "                            color_continuous_scale=px.colors.sequential.Rainbow,\n",
    "                            hover_data={'Domain to species': True, 'Domain to genus': True,\n",
    "                                        'element': True, 'contig': True, 'mod_count': True},\n",
    "                            size='contig_length_sqrt')\n",
    "        title = f' HAMBI - UMAP with n_neighbors={n}, min_dist={m}'\n",
    "        fig.update_layout(\n",
    "            height=1700,\n",
    "            width=1200,\n",
    "            title_text=title,\n",
    "            showlegend=True,\n",
    "            legend=dict(\n",
    "                x=0.5,\n",
    "                y=-0.05,\n",
    "                traceorder=\"normal\",\n",
    "                xanchor='center',\n",
    "                yanchor='top',\n",
    "                orientation='h'\n",
    "            ),\n",
    "            template='simple_white',\n",
    "            xaxis=dict(\n",
    "                showgrid=True,\n",
    "                gridcolor='lightgray',\n",
    "                zeroline=False,\n",
    "                showline=True,  \n",
    "                linecolor='black', \n",
    "                linewidth=1,\n",
    "                mirror=True\n",
    "            ),\n",
    "            yaxis=dict(\n",
    "                showgrid=True,\n",
    "                gridcolor='lightgray',\n",
    "                zeroline=False,\n",
    "                showline=True,\n",
    "                linecolor='black',\n",
    "                linewidth=1,\n",
    "                mirror=True\n",
    "            )\n",
    "        )\n",
    "        fig.show()\n",
    "        #fig.write_image(f'UMAP_HAMBI/UMAP_{n}_{m}_filtered_by_50_mod_counts_color_mod.png')\n",
    "        #fig.write_html(f'UMAP_HAMBI/UMAP_{n}_{m}_filtered_by_50_mod_counts_color_mod.html')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "af4a1cc1-436c-48e6-9996-cc820611488d",
   "metadata": {},
   "outputs": [],
   "source": [
    "## Read in and plot"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "id": "d5a298aa-cccd-4e1c-a0e1-5c0530726e49",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>UMAP1</th>\n",
       "      <th>UMAP2</th>\n",
       "      <th>Domain to species</th>\n",
       "      <th>Domain to genus</th>\n",
       "      <th>element</th>\n",
       "      <th>contig</th>\n",
       "      <th>contig_length_sqrt</th>\n",
       "      <th>Domain to species_with_counts</th>\n",
       "      <th>count</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>bcAd1063T--bcAd1063T_ptg000223l</th>\n",
       "      <td>-1.760182</td>\n",
       "      <td>5.015911</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>bcAd1063T--bcAd1063T_ptg000223l</td>\n",
       "      <td>363.877727</td>\n",
       "      <td>NA (266)</td>\n",
       "      <td>266</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>bcAd1023T--bcAd1023T_ptg000162l</th>\n",
       "      <td>0.189562</td>\n",
       "      <td>-7.648015</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>bcAd1023T--bcAd1023T_ptg000162l</td>\n",
       "      <td>353.213816</td>\n",
       "      <td>NA (266)</td>\n",
       "      <td>266</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>bcAd1063T--bcAd1063T_ptg000318l</th>\n",
       "      <td>-2.079180</td>\n",
       "      <td>5.882244</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>bcAd1063T--bcAd1063T_ptg000318l</td>\n",
       "      <td>179.938879</td>\n",
       "      <td>NA (266)</td>\n",
       "      <td>266</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>bcAd1063T--bcAd1063T_ptg000224l</th>\n",
       "      <td>-2.305873</td>\n",
       "      <td>6.072272</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>bcAd1063T--bcAd1063T_ptg000224l</td>\n",
       "      <td>241.679540</td>\n",
       "      <td>NA (266)</td>\n",
       "      <td>266</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>bcAd1063T--bcAd1063T_ptg000364l</th>\n",
       "      <td>-1.512699</td>\n",
       "      <td>5.406714</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>bcAd1063T--bcAd1063T_ptg000364l</td>\n",
       "      <td>199.376528</td>\n",
       "      <td>NA (266)</td>\n",
       "      <td>266</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                    UMAP1     UMAP2 Domain to species  \\\n",
       "0                                                                       \n",
       "bcAd1063T--bcAd1063T_ptg000223l -1.760182  5.015911               NaN   \n",
       "bcAd1023T--bcAd1023T_ptg000162l  0.189562 -7.648015               NaN   \n",
       "bcAd1063T--bcAd1063T_ptg000318l -2.079180  5.882244               NaN   \n",
       "bcAd1063T--bcAd1063T_ptg000224l -2.305873  6.072272               NaN   \n",
       "bcAd1063T--bcAd1063T_ptg000364l -1.512699  5.406714               NaN   \n",
       "\n",
       "                                Domain to genus element  \\\n",
       "0                                                         \n",
       "bcAd1063T--bcAd1063T_ptg000223l             NaN     NaN   \n",
       "bcAd1023T--bcAd1023T_ptg000162l             NaN     NaN   \n",
       "bcAd1063T--bcAd1063T_ptg000318l             NaN     NaN   \n",
       "bcAd1063T--bcAd1063T_ptg000224l             NaN     NaN   \n",
       "bcAd1063T--bcAd1063T_ptg000364l             NaN     NaN   \n",
       "\n",
       "                                                          contig  \\\n",
       "0                                                                  \n",
       "bcAd1063T--bcAd1063T_ptg000223l  bcAd1063T--bcAd1063T_ptg000223l   \n",
       "bcAd1023T--bcAd1023T_ptg000162l  bcAd1023T--bcAd1023T_ptg000162l   \n",
       "bcAd1063T--bcAd1063T_ptg000318l  bcAd1063T--bcAd1063T_ptg000318l   \n",
       "bcAd1063T--bcAd1063T_ptg000224l  bcAd1063T--bcAd1063T_ptg000224l   \n",
       "bcAd1063T--bcAd1063T_ptg000364l  bcAd1063T--bcAd1063T_ptg000364l   \n",
       "\n",
       "                                 contig_length_sqrt  \\\n",
       "0                                                     \n",
       "bcAd1063T--bcAd1063T_ptg000223l          363.877727   \n",
       "bcAd1023T--bcAd1023T_ptg000162l          353.213816   \n",
       "bcAd1063T--bcAd1063T_ptg000318l          179.938879   \n",
       "bcAd1063T--bcAd1063T_ptg000224l          241.679540   \n",
       "bcAd1063T--bcAd1063T_ptg000364l          199.376528   \n",
       "\n",
       "                                Domain to species_with_counts  count  \n",
       "0                                                                     \n",
       "bcAd1063T--bcAd1063T_ptg000223l                      NA (266)    266  \n",
       "bcAd1023T--bcAd1023T_ptg000162l                      NA (266)    266  \n",
       "bcAd1063T--bcAd1063T_ptg000318l                      NA (266)    266  \n",
       "bcAd1063T--bcAd1063T_ptg000224l                      NA (266)    266  \n",
       "bcAd1063T--bcAd1063T_ptg000364l                      NA (266)    266  "
      ]
     },
     "execution_count": 28,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Read data\n",
    "df_mod_counts = pd.read_csv('UMAP_HAMBI_above50/df_mod_counts.csv', sep='\\t', index_col=0, low_memory=False)\n",
    "mod_counts_UMAP_df = pd.read_csv('UMAP_HAMBI_above50/mod_counts_UMAP_df.csv', sep='\\t', index_col=0, low_memory=False)\n",
    "\n",
    "# Check\n",
    "df_mod_counts.head()\n",
    "mod_counts_UMAP_df.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c830e6fd-0e74-4d92-af71-e6f49b1a3833",
   "metadata": {},
   "source": [
    "## Extract clusters and redraw UMAP with subset of data\n",
    "### All data ðŸ’ \n",
    "#### > 50 mod counts\n",
    "##### C1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "id": "3ce4bc0f-dbcc-4d91-b23c-10a540c2d0cb",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Extract based on UMAP1 (x axis) & UMAP2 (y axis) values:\n",
    "mod_counts_UMAP_df_C1 = mod_counts_UMAP_df.loc[(mod_counts_UMAP_df['UMAP1']>= 7.4) & (mod_counts_UMAP_df['UMAP1']<= 7.7)\n",
    "    & (mod_counts_UMAP_df['UMAP2']>= 3.9) & (mod_counts_UMAP_df['UMAP2']<= 4.5)]\n",
    "\n",
    "# Check\n",
    "mod_counts_UMAP_df_C1.head()\n",
    "\n",
    "# Save those contig IDs\n",
    "mod_counts_UMAP_df_C1.index.to_list()\n",
    "\n",
    "df_mod_counts_C1 = df_mod_counts[df_mod_counts.index.isin(mod_counts_UMAP_df_C1.index)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "id": "c77c132d-76a1-4efc-9d0d-392548d14328",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Save contig IDs\n",
    "mod_counts_C1_contigs = df_mod_counts_C1.index.to_list()\n",
    "\n",
    "directory = 'UMAP_HAMBI_above50'\n",
    "file_path = os.path.join(directory, 'mod_counts_C1_contigs.txt')\n",
    "\n",
    "with open(file_path, 'w') as file:\n",
    "    for item in mod_counts_C1_contigs:\n",
    "        file.write(f\"{item}\\n\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e71e0fbc-78ee-485e-96d3-0f3bf612adc0",
   "metadata": {},
   "source": [
    "## Extract clusters and redraw UMAP with subset of data\n",
    "### All data ðŸ’ \n",
    "#### > 50 mod counts\n",
    "##### C2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "id": "a94018ce-4a55-461d-9337-ec0940179e82",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Extract based on UMAP1 (x axis) & UMAP2 (y axis) values:\n",
    "mod_counts_UMAP_df_C2 = mod_counts_UMAP_df.loc[(mod_counts_UMAP_df['UMAP1']>= -6) & (mod_counts_UMAP_df['UMAP1']<= -5.6)\n",
    "    & (mod_counts_UMAP_df['UMAP2']>= -0.6) & (mod_counts_UMAP_df['UMAP2']<= 0)]\n",
    "\n",
    "# Check\n",
    "mod_counts_UMAP_df_C2.head()\n",
    "\n",
    "# Save those contig IDs\n",
    "mod_counts_UMAP_df_C2.index.to_list()\n",
    "\n",
    "df_mod_counts_C2 = df_mod_counts[df_mod_counts.index.isin(mod_counts_UMAP_df_C2.index)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "id": "ec6dc556-71b3-47eb-8732-c29da34847a4",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Save contig IDs\n",
    "mod_counts_C2_contigs = df_mod_counts_C2.index.to_list()\n",
    "\n",
    "directory = 'UMAP_HAMBI_above50'\n",
    "file_path = os.path.join(directory, 'mod_counts_C2_contigs.txt')\n",
    "\n",
    "with open(file_path, 'w') as file:\n",
    "    for item in mod_counts_C2_contigs:\n",
    "        file.write(f\"{item}\\n\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1fa9b5cb-21be-4bbe-a68b-7d50ce397610",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "f29bdc75-c8cd-4838-9e54-2c586e1e43d2",
   "metadata": {},
   "source": [
    "## Extract clusters and redraw UMAP with subset of data\n",
    "### All data ðŸ’ \n",
    "#### > 50 mod counts\n",
    "##### C3"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "id": "747cf9d6-beb1-48a4-ad95-4e1d4e5fb06f",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Extract based on UMAP1 (x axis) & UMAP2 (y axis) values:\n",
    "mod_counts_UMAP_df_C3 = mod_counts_UMAP_df.loc[(mod_counts_UMAP_df['UMAP1']>= -7) & (mod_counts_UMAP_df['UMAP1']<= -5.6)\n",
    "    & (mod_counts_UMAP_df['UMAP2']>= -2) & (mod_counts_UMAP_df['UMAP2']<= -0.8)]\n",
    "\n",
    "# Check\n",
    "mod_counts_UMAP_df_C3.head()\n",
    "\n",
    "# Save those contig IDs\n",
    "mod_counts_UMAP_df_C3.index.to_list()\n",
    "\n",
    "df_mod_counts_C3 = df_mod_counts[df_mod_counts.index.isin(mod_counts_UMAP_df_C3.index)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "id": "0463d533-1113-42db-8fbe-02130c0f6cae",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Save contig IDs\n",
    "mod_counts_C3_contigs = df_mod_counts_C3.index.to_list()\n",
    "\n",
    "directory = 'UMAP_HAMBI_above50'\n",
    "file_path = os.path.join(directory, 'mod_counts_C3_contigs.txt')\n",
    "\n",
    "with open(file_path, 'w') as file:\n",
    "    for item in mod_counts_C3_contigs:\n",
    "        file.write(f\"{item}\\n\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ad59f070-a690-4108-bc7b-0d19990fb736",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "1d500c01-e930-4935-a6ba-bc45b9e3f686",
   "metadata": {},
   "source": [
    "## Extract clusters and redraw UMAP with subset of data\n",
    "### All data ðŸ’ \n",
    "#### > 50 mod counts\n",
    "##### C4"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "id": "c7a40281-77aa-4072-abb7-67c9171466a6",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Extract based on UMAP1 (x axis) & UMAP2 (y axis) values:\n",
    "mod_counts_UMAP_df_C4 = mod_counts_UMAP_df.loc[(mod_counts_UMAP_df['UMAP1']>= -8) & (mod_counts_UMAP_df['UMAP1']<= -7)\n",
    "    & (mod_counts_UMAP_df['UMAP2']>= 0.6) & (mod_counts_UMAP_df['UMAP2']<= 1.2)]\n",
    "\n",
    "# Check\n",
    "mod_counts_UMAP_df_C4.head()\n",
    "\n",
    "# Save those contig IDs\n",
    "mod_counts_UMAP_df_C4.index.to_list()\n",
    "\n",
    "df_mod_counts_C4 = df_mod_counts[df_mod_counts.index.isin(mod_counts_UMAP_df_C4.index)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "id": "c7d1d349-a2cf-42ea-8ee7-9df975d57889",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Save contig IDs\n",
    "mod_counts_C4_contigs = df_mod_counts_C4.index.to_list()\n",
    "\n",
    "directory = 'UMAP_HAMBI_above50'\n",
    "file_path = os.path.join(directory, 'mod_counts_C4_contigs.txt')\n",
    "\n",
    "with open(file_path, 'w') as file:\n",
    "    for item in mod_counts_C4_contigs:\n",
    "        file.write(f\"{item}\\n\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d0d0234d-334d-4043-982e-3faba06feb75",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "e4adb739-e1e1-4488-a541-74c72bde7e94",
   "metadata": {},
   "source": [
    "## Extract clusters and redraw UMAP with subset of data\n",
    "### All data ðŸ’ \n",
    "#### > 50 mod counts\n",
    "##### C11"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "id": "111f85c9-d4f9-4a18-982b-d5a491c0fe26",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Extract based on UMAP1 (x axis) & UMAP2 (y axis) values:\n",
    "mod_counts_UMAP_df_C11 = mod_counts_UMAP_df.loc[(mod_counts_UMAP_df['UMAP1']>= -2.5) & (mod_counts_UMAP_df['UMAP1']<= -1.5)\n",
    "    & (mod_counts_UMAP_df['UMAP2']>= 4.2) & (mod_counts_UMAP_df['UMAP2']<= 4.8)]\n",
    "\n",
    "# Check\n",
    "mod_counts_UMAP_df_C11.head()\n",
    "\n",
    "# Save those contig IDs\n",
    "mod_counts_UMAP_df_C11.index.to_list()\n",
    "\n",
    "df_mod_counts_C11 = df_mod_counts[df_mod_counts.index.isin(mod_counts_UMAP_df_C11.index)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "id": "b1879b72-b7e7-4401-84b0-2ea6cb0a66f8",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Save contig IDs\n",
    "mod_counts_C11_contigs = df_mod_counts_C11.index.to_list()\n",
    "\n",
    "directory = 'UMAP_HAMBI_above50'\n",
    "file_path = os.path.join(directory, 'mod_counts_C11_contigs.txt')\n",
    "\n",
    "with open(file_path, 'w') as file:\n",
    "    for item in mod_counts_C11_contigs:\n",
    "        file.write(f\"{item}\\n\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "bfa8d23d-c819-4d3c-a150-1240e0b3a24f",
   "metadata": {},
   "source": [
    "## Extract clusters and redraw UMAP with subset of data\n",
    "### All data ðŸ’ \n",
    "#### > 50 mod counts\n",
    "##### C6"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "id": "87a70d76-9456-4804-8a52-cd579986ea6a",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Extract based on UMAP1 (x axis) & UMAP2 (y axis) values:\n",
    "mod_counts_UMAP_df_C6 = mod_counts_UMAP_df.loc[(mod_counts_UMAP_df['UMAP1']>= -2.5) & (mod_counts_UMAP_df['UMAP1']<= -1.5)\n",
    "    & (mod_counts_UMAP_df['UMAP2']>= -4.1) & (mod_counts_UMAP_df['UMAP2']<= -3)]\n",
    "\n",
    "# Check\n",
    "mod_counts_UMAP_df_C6.head()\n",
    "\n",
    "# Save those contig IDs\n",
    "mod_counts_UMAP_df_C6.index.to_list()\n",
    "\n",
    "df_mod_counts_C6 = df_mod_counts[df_mod_counts.index.isin(mod_counts_UMAP_df_C6.index)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "id": "05ed76af-718d-48e0-80ad-35401f95a961",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Save contig IDs\n",
    "mod_counts_C6_contigs = df_mod_counts_C6.index.to_list()\n",
    "\n",
    "directory = 'UMAP_HAMBI_above50'\n",
    "file_path = os.path.join(directory, 'mod_counts_C6_contigs.txt')\n",
    "\n",
    "with open(file_path, 'w') as file:\n",
    "    for item in mod_counts_C6_contigs:\n",
    "        file.write(f\"{item}\\n\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2dcfbc87-7959-49c3-ab30-2812c3b167c7",
   "metadata": {},
   "source": [
    "## Extract clusters and redraw UMAP with subset of data\n",
    "### All data ðŸ’ \n",
    "#### > 50 mod counts\n",
    "##### C5"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "id": "9ea6524a-69d1-4b0b-886c-7ca36af419a7",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Extract based on UMAP1 (x axis) & UMAP2 (y axis) values:\n",
    "mod_counts_UMAP_df_C5 = mod_counts_UMAP_df.loc[(mod_counts_UMAP_df['UMAP1']>= -2.5) & (mod_counts_UMAP_df['UMAP1']<= -2)\n",
    "    & (mod_counts_UMAP_df['UMAP2']>= -2.1) & (mod_counts_UMAP_df['UMAP2']<= -1.5)]\n",
    "\n",
    "# Check\n",
    "mod_counts_UMAP_df_C5.head()\n",
    "\n",
    "# Save those contig IDs\n",
    "mod_counts_UMAP_df_C5.index.to_list()\n",
    "\n",
    "df_mod_counts_C5 = df_mod_counts[df_mod_counts.index.isin(mod_counts_UMAP_df_C5.index)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "id": "1219e802-7504-4f13-a664-d9d6e00e4b9a",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Save contig IDs\n",
    "mod_counts_C5_contigs = df_mod_counts_C5.index.to_list()\n",
    "\n",
    "directory = 'UMAP_HAMBI_above50'\n",
    "file_path = os.path.join(directory, 'mod_counts_C5_contigs.txt')\n",
    "\n",
    "with open(file_path, 'w') as file:\n",
    "    for item in mod_counts_C5_contigs:\n",
    "        file.write(f\"{item}\\n\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "eaf827c6-99ed-4361-817e-735fdc979032",
   "metadata": {},
   "source": [
    "## Extract clusters and redraw UMAP with subset of data\n",
    "### All data ðŸ’ \n",
    "#### > 50 mod counts\n",
    "##### C7"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 51,
   "id": "ec0583b2-dfd7-4e62-8e2d-52d8d3239936",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Extract based on UMAP1 (x axis) & UMAP2 (y axis) values:\n",
    "mod_counts_UMAP_df_C7 = mod_counts_UMAP_df.loc[(mod_counts_UMAP_df['UMAP1']>= -2) & (mod_counts_UMAP_df['UMAP1']<= -1.36)\n",
    "    & (mod_counts_UMAP_df['UMAP2']>= -2.271) & (mod_counts_UMAP_df['UMAP2']<= -1.79)]\n",
    "\n",
    "# Check\n",
    "mod_counts_UMAP_df_C7.head()\n",
    "\n",
    "# Save those contig IDs\n",
    "mod_counts_UMAP_df_C7.index.to_list()\n",
    "\n",
    "df_mod_counts_C7 = df_mod_counts[df_mod_counts.index.isin(mod_counts_UMAP_df_C7.index)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 52,
   "id": "f00df4b4-9435-43da-97c1-866eb7629ac5",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Save contig IDs\n",
    "mod_counts_C7_contigs = df_mod_counts_C7.index.to_list()\n",
    "\n",
    "directory = 'UMAP_HAMBI_above50'\n",
    "file_path = os.path.join(directory, 'mod_counts_C7_contigs.txt')\n",
    "\n",
    "with open(file_path, 'w') as file:\n",
    "    for item in mod_counts_C7_contigs:\n",
    "        file.write(f\"{item}\\n\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9f3f1609-a4cf-4d20-a484-cfc2215e2924",
   "metadata": {},
   "source": [
    "## Extract clusters and redraw UMAP with subset of data\n",
    "### All data ðŸ’ \n",
    "#### > 50 mod counts\n",
    "##### C8"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 55,
   "id": "003b2d4b-1297-4104-b49c-4adfb24d1eee",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Extract based on UMAP1 (x axis) & UMAP2 (y axis) values:\n",
    "mod_counts_UMAP_df_C8 = mod_counts_UMAP_df.loc[(mod_counts_UMAP_df['UMAP1']>= -3.5) & (mod_counts_UMAP_df['UMAP1']<= -2.5)\n",
    "    & (mod_counts_UMAP_df['UMAP2']>= -3) & (mod_counts_UMAP_df['UMAP2']<= -2)]\n",
    "\n",
    "# Check\n",
    "mod_counts_UMAP_df_C8.head()\n",
    "\n",
    "# Save those contig IDs\n",
    "mod_counts_UMAP_df_C8.index.to_list()\n",
    "\n",
    "df_mod_counts_C8 = df_mod_counts[df_mod_counts.index.isin(mod_counts_UMAP_df_C8.index)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 56,
   "id": "4345c91b-9bc7-4ed8-845e-6c100a82dc1d",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Save contig IDs\n",
    "mod_counts_C8_contigs = df_mod_counts_C8.index.to_list()\n",
    "\n",
    "directory = 'UMAP_HAMBI_above50'\n",
    "file_path = os.path.join(directory, 'mod_counts_C8_contigs.txt')\n",
    "\n",
    "with open(file_path, 'w') as file:\n",
    "    for item in mod_counts_C8_contigs:\n",
    "        file.write(f\"{item}\\n\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6401e1a6-c80b-4176-bfe2-f331ce998483",
   "metadata": {},
   "source": [
    "## Extract clusters and redraw UMAP with subset of data\n",
    "### All data ðŸ’ \n",
    "#### > 50 mod counts\n",
    "##### C9"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 57,
   "id": "b192743d-378c-44ab-b320-1e664539b56b",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Extract based on UMAP1 (x axis) & UMAP2 (y axis) values:\n",
    "mod_counts_UMAP_df_C9 = mod_counts_UMAP_df.loc[(mod_counts_UMAP_df['UMAP1']>= -1.5) & (mod_counts_UMAP_df['UMAP1']<= -1.2)\n",
    "    & (mod_counts_UMAP_df['UMAP2']>= -2.4) & (mod_counts_UMAP_df['UMAP2']<= -2.1)]\n",
    "\n",
    "# Check\n",
    "mod_counts_UMAP_df_C9.head()\n",
    "\n",
    "# Save those contig IDs\n",
    "mod_counts_UMAP_df_C9.index.to_list()\n",
    "\n",
    "df_mod_counts_C9 = df_mod_counts[df_mod_counts.index.isin(mod_counts_UMAP_df_C9.index)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 58,
   "id": "848c97ef-13ac-470b-807a-1beb85fbfbec",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Save contig IDs\n",
    "mod_counts_C9_contigs = df_mod_counts_C9.index.to_list()\n",
    "\n",
    "directory = 'UMAP_HAMBI_above50'\n",
    "file_path = os.path.join(directory, 'mod_counts_C9_contigs.txt')\n",
    "\n",
    "with open(file_path, 'w') as file:\n",
    "    for item in mod_counts_C9_contigs:\n",
    "        file.write(f\"{item}\\n\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b8b4a2cc-f29c-424e-ab50-e75ec48a7e5e",
   "metadata": {},
   "source": [
    "## Extract clusters and redraw UMAP with subset of data\n",
    "### All data ðŸ’ \n",
    "#### > 50 mod counts\n",
    "##### C10"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 59,
   "id": "80086fe1-8d99-49a2-831b-18928e5e8a61",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Extract based on UMAP1 (x axis) & UMAP2 (y axis) values:\n",
    "mod_counts_UMAP_df_C10 = mod_counts_UMAP_df.loc[(mod_counts_UMAP_df['UMAP1']>= 3) & (mod_counts_UMAP_df['UMAP1']<= 5)\n",
    "    & (mod_counts_UMAP_df['UMAP2']>= -4) & (mod_counts_UMAP_df['UMAP2']<= -1)]\n",
    "\n",
    "# Check\n",
    "mod_counts_UMAP_df_C10.head()\n",
    "\n",
    "# Save those contig IDs\n",
    "mod_counts_UMAP_df_C10.index.to_list()\n",
    "\n",
    "df_mod_counts_C10 = df_mod_counts[df_mod_counts.index.isin(mod_counts_UMAP_df_C10.index)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 60,
   "id": "a2b57eb3-c5a9-4911-9ee7-cfed589acb3f",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Save contig IDs\n",
    "mod_counts_C10_contigs = df_mod_counts_C10.index.to_list()\n",
    "\n",
    "directory = 'UMAP_HAMBI_above50'\n",
    "file_path = os.path.join(directory, 'mod_counts_C10_contigs.txt')\n",
    "\n",
    "with open(file_path, 'w') as file:\n",
    "    for item in mod_counts_C10_contigs:\n",
    "        file.write(f\"{item}\\n\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "af29cdce-8bab-4787-93ba-dade71036bfe",
   "metadata": {
    "jp-MarkdownHeadingCollapsed": true
   },
   "source": [
    "#### Draw with > 50 again without NA"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8b4051d6-88ba-4f00-a94c-4a176d89bb08",
   "metadata": {},
   "outputs": [],
   "source": [
    "df_mod_counts_no_NA = df_mod_counts[df_mod_counts['s'].notna()]\n",
    "print(df_mod_counts_no_NA)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5126c6d7-8405-4a4b-b8c5-3d76d11c4352",
   "metadata": {},
   "outputs": [],
   "source": [
    "n_neighbors = [20]\n",
    "min_dist = [0.5]\n",
    "\n",
    "# 50\n",
    "color_dict = [\"#00FF00\", \"#00FFFF\", \"#FF00FF\", \"#0000FF\", \"#800000\",\n",
    "    \"#FFFF00\", \"#000080\", \"#008000\", \"#808000\", \"#008080\", \"#800080\", \"#3bccf7\", \"#FF4500\", \"#4682B4\", \"#DAA520\"]\n",
    "\n",
    "\n",
    "custom_colors = color_dict\n",
    "for n in n_neighbors:\n",
    "    for m in min_dist:\n",
    "        reducer = umap.UMAP(n_neighbors=n, min_dist=m, random_state=seed)\n",
    "        embedding = reducer.fit_transform(df_mod_counts_no_NA.iloc[:, :-20])\n",
    "        UMAP_df_mod_counts_no_NA  = pd.DataFrame({\n",
    "            'UMAP1': embedding[:, 0],\n",
    "            'UMAP2': embedding[:, 1],\n",
    "            'Domain to species': df_mod_counts_no_NA['Domain to species'],\n",
    "            'Domain to genus': df_mod_counts_no_NA['Domain to genus'],\n",
    "            'element': df_mod_counts_no_NA['element'],\n",
    "            'contig': df_mod_counts_no_NA.index,\n",
    "            'contig_length_sqrt': df_mod_counts_no_NA['contig_length_sqrt']\n",
    "        })\n",
    "        \n",
    "        column = 'Domain to species'\n",
    "        UMAP_df_mod_counts_no_NA['Domain to species'] = UMAP_df_mod_counts_no_NA['Domain to species'].fillna('NA')\n",
    "\n",
    "        value_counts = UMAP_df_mod_counts_no_NA[column].value_counts().to_dict()\n",
    "\n",
    "        UMAP_df_mod_counts_no_NA[f'{column}_with_counts'] = UMAP_df_mod_counts_no_NA[column].apply(lambda x: f'{x} ({value_counts[x]})')\n",
    "\n",
    "        UMAP_df_mod_counts_no_NA['count'] = UMAP_df_mod_counts_no_NA[column].map(value_counts)\n",
    "        UMAP_df_mod_counts_no_NA = UMAP_df_mod_counts_no_NA.sort_values(by=['count'], ascending=False)\n",
    "\n",
    "        UMAP_df_mod_counts_no_NA[f'{column}_with_counts'] = pd.Categorical(\n",
    "            UMAP_df_mod_counts_no_NA[f'{column}_with_counts'],\n",
    "            categories=UMAP_df_mod_counts_no_NA[f'{column}_with_counts'].unique(),\n",
    "            ordered=True\n",
    "        )\n",
    "        \n",
    "        fig = px.scatter(UMAP_df_mod_counts_no_NA, \n",
    "                            x='UMAP1', \n",
    "                            y='UMAP2', \n",
    "                            labels={f'{column}_with_counts': column},\n",
    "                            color=f'{column}_with_counts', \n",
    "                            title=f' HAMBI - UMAP with n_neighbors={n}, min_dist={m}', \n",
    "                            color_discrete_sequence=custom_colors,\n",
    "                            hover_data={'Domain to species': True, 'Domain to genus': True,\n",
    "                                        'element': True, 'contig': True},\n",
    "                        size = 'contig_length_sqrt')\n",
    "        title = f' HAMBI - UMAP with n_neighbors={n}, min_dist={m}'\n",
    "        fig.update_layout(\n",
    "            height=1700,\n",
    "            width=1200,\n",
    "            title_text=title,\n",
    "            showlegend=True,\n",
    "            legend=dict(\n",
    "                x=0.5,\n",
    "                y=-0.1,\n",
    "                traceorder=\"normal\",\n",
    "                xanchor='center',\n",
    "                yanchor='top',\n",
    "                orientation='h'\n",
    "            ),\n",
    "            template='simple_white',\n",
    "            xaxis=dict(\n",
    "                showgrid=True,\n",
    "                gridcolor='lightgray',\n",
    "                zeroline=False,\n",
    "                showline=True,  \n",
    "                linecolor='black', \n",
    "                linewidth=1,\n",
    "                mirror=True\n",
    "            ),\n",
    "            yaxis=dict(\n",
    "                showgrid=True,\n",
    "                gridcolor='lightgray',\n",
    "                zeroline=False,\n",
    "                showline=True,\n",
    "                linecolor='black',\n",
    "                linewidth=1,\n",
    "                mirror=True\n",
    "            )\n",
    "        )\n",
    "        fig.show()\n",
    "        fig.write_image(f'UMAP_HAMBI/UMAP_{n}_{m}_filtered_by_50_mod_counts_no_NA.png')\n",
    "        fig.write_html(f'UMAP_HAMBI/UMAP_{n}_{m}_filtered_by_50_mod_counts_no_NA.html')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8d9e22bd-2da3-465b-92eb-37c369b383a9",
   "metadata": {
    "jp-MarkdownHeadingCollapsed": true
   },
   "source": [
    "## Extract clusters and redraw UMAP with subset of data\n",
    "### All data ðŸ’ \n",
    "#### > 100 mod counts\n",
    "##### C1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c9c88070-6edc-45cd-b41f-33f81f84032a",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Extract based on UMAP1 (x axis) & UMAP2 (y axis) values:\n",
    "mod_counts_UMAP_df_C1 = mod_counts_UMAP_df.loc[(mod_counts_UMAP_df['UMAP1']>= -2.2) & (mod_counts_UMAP_df['UMAP1']<= -1.6)\n",
    "    & (mod_counts_UMAP_df['UMAP2']>= 6.9) & (mod_counts_UMAP_df['UMAP2']<= 7.2)]\n",
    "\n",
    "# Check\n",
    "mod_counts_UMAP_df_C1.head()\n",
    "\n",
    "# Save those contig IDs\n",
    "mod_counts_UMAP_df_C1.index.to_list()\n",
    "\n",
    "df_mod_counts_C1 = df_mod_counts[df_mod_counts.index.isin(mod_counts_UMAP_df_C1.index)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a62fc291-863d-49d8-be33-0cbdeac51bee",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Save contig IDs\n",
    "mod_counts_C1_contigs = df_mod_counts_C1.index.to_list()\n",
    "\n",
    "directory = 'UMAP_HAMBI_above100'\n",
    "file_path = os.path.join(directory, 'mod_counts_C1_contigs.txt')\n",
    "\n",
    "with open(file_path, 'w') as file:\n",
    "    for item in mod_counts_C1_contigs:\n",
    "        file.write(f\"{item}\\n\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8e229ee2-fd92-4820-8702-58ba06d23141",
   "metadata": {
    "jp-MarkdownHeadingCollapsed": true
   },
   "source": [
    "## Extract clusters and redraw UMAP with subset of data\n",
    "### All data ðŸ’ \n",
    "#### > 100 mod counts\n",
    "##### C2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "52d87531-85d7-4219-a61b-7f02ec4ac3f0",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Extract based on UMAP1 (x axis) & UMAP2 (y axis) values:\n",
    "mod_counts_UMAP_df_C2 = mod_counts_UMAP_df.loc[(mod_counts_UMAP_df['UMAP1']>= 11.6) & (mod_counts_UMAP_df['UMAP1']<= 12)\n",
    "    & (mod_counts_UMAP_df['UMAP2']>= -0.8) & (mod_counts_UMAP_df['UMAP2']<= -0.6)]\n",
    "\n",
    "# Check\n",
    "mod_counts_UMAP_df_C2.head()\n",
    "\n",
    "# Save those contig IDs\n",
    "mod_counts_UMAP_df_C2.index.to_list()\n",
    "\n",
    "df_mod_counts_C2 = df_mod_counts[df_mod_counts.index.isin(mod_counts_UMAP_df_C2.index)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f73e5dfc-27ac-4560-82a2-f9939c134204",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Save contig IDs\n",
    "mod_counts_C2_contigs = df_mod_counts_C2.index.to_list()\n",
    "\n",
    "directory = 'UMAP_HAMBI_above100'\n",
    "file_path = os.path.join(directory, 'mod_counts_C2_contigs.txt')\n",
    "\n",
    "with open(file_path, 'w') as file:\n",
    "    for item in mod_counts_C2_contigs:\n",
    "        file.write(f\"{item}\\n\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "66e56956-bd94-4693-aff4-54eb5309be4f",
   "metadata": {
    "jp-MarkdownHeadingCollapsed": true
   },
   "source": [
    "## Extract clusters and redraw UMAP with subset of data\n",
    "### All data ðŸ’ \n",
    "#### > 100 mod counts\n",
    "##### C3"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "cda3dc65-363e-4dd9-92c2-e173037fd550",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Extract based on UMAP1 (x axis) & UMAP2 (y axis) values:\n",
    "mod_counts_UMAP_df_C3 = mod_counts_UMAP_df.loc[(mod_counts_UMAP_df['UMAP1']>= 13.05) & (mod_counts_UMAP_df['UMAP1']<= 13.8)\n",
    "    & (mod_counts_UMAP_df['UMAP2']>= -2.8) & (mod_counts_UMAP_df['UMAP2']<= -1.8)]\n",
    "\n",
    "# Check\n",
    "mod_counts_UMAP_df_C3.head()\n",
    "\n",
    "# Save those contig IDs\n",
    "mod_counts_UMAP_df_C3.index.to_list()\n",
    "\n",
    "df_mod_counts_C3 = df_mod_counts[df_mod_counts.index.isin(mod_counts_UMAP_df_C3.index)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5797ef55-5a94-4469-896b-c9992d7dd43e",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Save contig IDs\n",
    "mod_counts_C3_contigs = df_mod_counts_C3.index.to_list()\n",
    "\n",
    "directory = 'UMAP_HAMBI_above100'\n",
    "file_path = os.path.join(directory, 'mod_counts_C3_contigs.txt')\n",
    "\n",
    "with open(file_path, 'w') as file:\n",
    "    for item in mod_counts_C3_contigs:\n",
    "        file.write(f\"{item}\\n\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "580cfc89-9da6-4841-bfe7-8c7cca521e5e",
   "metadata": {
    "jp-MarkdownHeadingCollapsed": true
   },
   "source": [
    "## Extract clusters and redraw UMAP with subset of data\n",
    "### All data ðŸ’ \n",
    "#### > 100 mod counts\n",
    "##### C4"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "56922f62-6e2d-4dfb-acb7-209a152b3fd3",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Extract based on UMAP1 (x axis) & UMAP2 (y axis) values:\n",
    "mod_counts_UMAP_df_C4 = mod_counts_UMAP_df.loc[(mod_counts_UMAP_df['UMAP1']>= 9.9) & (mod_counts_UMAP_df['UMAP1']<= 12)\n",
    "    & (mod_counts_UMAP_df['UMAP2']>= -2.5) & (mod_counts_UMAP_df['UMAP2']<= -1)]\n",
    "\n",
    "# Check\n",
    "mod_counts_UMAP_df_C4.head()\n",
    "\n",
    "# Save those contig IDs\n",
    "mod_counts_UMAP_df_C4.index.to_list()\n",
    "\n",
    "df_mod_counts_C4 = df_mod_counts[df_mod_counts.index.isin(mod_counts_UMAP_df_C4.index)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f474d497-a59c-4861-ab0b-aea416cf9c6e",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Save contig IDs\n",
    "mod_counts_C4_contigs = df_mod_counts_C4.index.to_list()\n",
    "\n",
    "directory = 'UMAP_HAMBI_above100'\n",
    "file_path = os.path.join(directory, 'mod_counts_C4_contigs.txt')\n",
    "\n",
    "with open(file_path, 'w') as file:\n",
    "    for item in mod_counts_C4_contigs:\n",
    "        file.write(f\"{item}\\n\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ddeb2829-810e-45a7-818b-d6364df9f476",
   "metadata": {
    "jp-MarkdownHeadingCollapsed": true
   },
   "source": [
    "## Extract clusters and redraw UMAP with subset of data\n",
    "### All data ðŸ’ \n",
    "#### > 100 mod counts\n",
    "##### C5"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "035d156f-bba7-4f96-a4e1-185bace4f075",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Extract based on UMAP1 (x axis) & UMAP2 (y axis) values:\n",
    "mod_counts_UMAP_df_C5 = mod_counts_UMAP_df.loc[(mod_counts_UMAP_df['UMAP1']>= 1) & (mod_counts_UMAP_df['UMAP1']<= 4.1)\n",
    "    & (mod_counts_UMAP_df['UMAP2']>= 2) & (mod_counts_UMAP_df['UMAP2']<= 3.5)]\n",
    "\n",
    "# Check\n",
    "mod_counts_UMAP_df_C5.head()\n",
    "\n",
    "# Save those contig IDs\n",
    "mod_counts_UMAP_df_C5.index.to_list()\n",
    "\n",
    "df_mod_counts_C5 = df_mod_counts[df_mod_counts.index.isin(mod_counts_UMAP_df_C5.index)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0ef71465-b9ce-4cbe-9c51-aaa63800697d",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Save contig IDs\n",
    "mod_counts_C5_contigs = df_mod_counts_C5.index.to_list()\n",
    "\n",
    "directory = 'UMAP_HAMBI_above100'\n",
    "file_path = os.path.join(directory, 'mod_counts_C5_contigs.txt')\n",
    "\n",
    "with open(file_path, 'w') as file:\n",
    "    for item in mod_counts_C5_contigs:\n",
    "        file.write(f\"{item}\\n\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9d14ec9c-1e9e-4fc2-9f3f-b94951a3cc11",
   "metadata": {
    "jp-MarkdownHeadingCollapsed": true
   },
   "source": [
    "## Extract clusters and redraw UMAP with subset of data\n",
    "### All data ðŸ’ \n",
    "#### > 100 mod counts\n",
    "##### C6"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "85ace9c2-8e91-44c1-9caf-2717dc750e4d",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Extract based on UMAP1 (x axis) & UMAP2 (y axis) values:\n",
    "mod_counts_UMAP_df_C6 = mod_counts_UMAP_df.loc[(mod_counts_UMAP_df['UMAP1']>= 7.8) & (mod_counts_UMAP_df['UMAP1']<= 8.6)\n",
    "    & (mod_counts_UMAP_df['UMAP2']>= -4.4) & (mod_counts_UMAP_df['UMAP2']<= -3.4)]\n",
    "\n",
    "# Check\n",
    "mod_counts_UMAP_df_C6.head()\n",
    "\n",
    "# Save those contig IDs\n",
    "mod_counts_UMAP_df_C6.index.to_list()\n",
    "\n",
    "df_mod_counts_C6 = df_mod_counts[df_mod_counts.index.isin(mod_counts_UMAP_df_C6.index)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "97e73216-3fcd-45be-b903-7029d6ad9cbe",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Save contig IDs\n",
    "mod_counts_C6_contigs = df_mod_counts_C6.index.to_list()\n",
    "\n",
    "directory = 'UMAP_HAMBI_above100'\n",
    "file_path = os.path.join(directory, 'mod_counts_C6_contigs.txt')\n",
    "\n",
    "with open(file_path, 'w') as file:\n",
    "    for item in mod_counts_C6_contigs:\n",
    "        file.write(f\"{item}\\n\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8f59e61d-86bf-4aa9-9fed-452a15fb0aac",
   "metadata": {
    "jp-MarkdownHeadingCollapsed": true
   },
   "source": [
    "## Extract clusters and redraw UMAP with subset of data\n",
    "### All data ðŸ’ \n",
    "#### > 100 mod counts\n",
    "##### C7"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0ce4f08d-9be5-4f1a-8395-c5bf7fd8a38e",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Extract based on UMAP1 (x axis) & UMAP2 (y axis) values:\n",
    "mod_counts_UMAP_df_C7_1 = mod_counts_UMAP_df.loc[(mod_counts_UMAP_df['UMAP1']>= 6.8) & (mod_counts_UMAP_df['UMAP1']<= 7.5)\n",
    "    & (mod_counts_UMAP_df['UMAP2']>= -4) & (mod_counts_UMAP_df['UMAP2']<= -2.8)]\n",
    "\n",
    "mod_counts_UMAP_df_C7_2 = mod_counts_UMAP_df.loc[(mod_counts_UMAP_df['UMAP1']>= 5.35) & (mod_counts_UMAP_df['UMAP1']<= 6.5)\n",
    "    & (mod_counts_UMAP_df['UMAP2']>= -3) & (mod_counts_UMAP_df['UMAP2']<= -1.97)]\n",
    "\n",
    "mod_counts_UMAP_df_C7 = pd.concat([mod_counts_UMAP_df_C7_1, mod_counts_UMAP_df_C7_2], ignore_index=False)\n",
    "\n",
    "# Check\n",
    "mod_counts_UMAP_df_C7.head()\n",
    "\n",
    "# Save those contig IDs\n",
    "mod_counts_UMAP_df_C7.index.to_list()\n",
    "\n",
    "df_mod_counts_C7 = df_mod_counts[df_mod_counts.index.isin(mod_counts_UMAP_df_C7.index)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3575309a-9ca2-412d-bfdf-fc0acd551e3e",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Save contig IDs\n",
    "mod_counts_C7_contigs = df_mod_counts_C7.index.to_list()\n",
    "\n",
    "directory = 'UMAP_HAMBI_above100'\n",
    "file_path = os.path.join(directory, 'mod_counts_C7_contigs.txt')\n",
    "\n",
    "with open(file_path, 'w') as file:\n",
    "    for item in mod_counts_C7_contigs:\n",
    "        file.write(f\"{item}\\n\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7842df00-c287-4535-be8d-e6eae41bd34f",
   "metadata": {},
   "source": [
    "### Focus on distinct clusters"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e98fbc58-4504-4cda-835a-24f3b438c807",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Extract based on UMAP1 (x axis) & UMAP2 (y axis) values:\n",
    "mod_counts_UMAP_df_focused = mod_counts_UMAP_df.loc[(mod_counts_UMAP_df['UMAP1']>= 7) & (mod_counts_UMAP_df['UMAP1']<= 9.6)\n",
    "    & (mod_counts_UMAP_df['UMAP2']>= -6.1) & (mod_counts_UMAP_df['UMAP2']<= -3)]\n",
    "\n",
    "# Check\n",
    "mod_counts_UMAP_df_focused.head()\n",
    "\n",
    "mod_counts_data_focused = df_mod_counts[df_mod_counts.index.isin(mod_counts_UMAP_df_focused['contig'])]\n",
    "print(mod_counts_data_focused)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6a33aedc-6c2e-46c4-83cc-147e8f1f89f3",
   "metadata": {},
   "outputs": [],
   "source": [
    "#n_neighbors = [20, 30 ]\n",
    "#min_dist = [0.01, 0.1, 0.2]\n",
    "\n",
    "n_neighbors = [20]\n",
    "min_dist = [0.5]\n",
    "\n",
    "# 100\n",
    "color_dict = [\"#FF0000\", \"#008080\", \"#800000\", \"#3bccf7\", \"#0000FF\"]\n",
    "\n",
    "custom_colors = color_dict\n",
    "\n",
    "for n in n_neighbors:\n",
    "    for m in min_dist:\n",
    "        reducer = umap.UMAP(n_neighbors=n, min_dist=m, random_state=seed)\n",
    "        embedding = reducer.fit_transform(mod_counts_data_focused.iloc[:, :-20])\n",
    "        mod_counts_UMAP_df_focused  = pd.DataFrame({\n",
    "            'UMAP1': embedding[:, 0],\n",
    "            'UMAP2': embedding[:, 1],\n",
    "            'Domain to species': mod_counts_data_focused['Domain to species'],\n",
    "            'Domain to genus': mod_counts_data_focused['Domain to genus'],\n",
    "            'element': mod_counts_data_focused['element'],\n",
    "            'contig': mod_counts_data_focused.index,\n",
    "            'contig_length_sqrt': mod_counts_data_focused['contig_length_sqrt']\n",
    "        })\n",
    "        # clustering k defined choosing the k by looking at the UMAP \n",
    "        kmeans = KMeans(n_clusters=11, random_state=seed)\n",
    "        \n",
    "        k_labels = kmeans.fit_predict(embedding)\n",
    "        \n",
    "        mod_counts_UMAP_df_focused['cluster'] = k_labels\n",
    "        \n",
    "        column = 'Domain to species'\n",
    "        mod_counts_UMAP_df_focused['Domain to species'] = mod_counts_UMAP_df_focused['Domain to species'].fillna('NA')\n",
    "\n",
    "        value_counts = mod_counts_UMAP_df_focused[column].value_counts().to_dict()\n",
    "\n",
    "        mod_counts_UMAP_df_focused[f'{column}_with_counts'] = mod_counts_UMAP_df_focused[column].apply(lambda x: f'{x} ({value_counts[x]})')\n",
    "\n",
    "        mod_counts_UMAP_df_focused['count'] = mod_counts_UMAP_df_focused[column].map(value_counts)\n",
    "        mod_counts_UMAP_df_focused = mod_counts_UMAP_df_focused.sort_values(by=['count'], ascending=False)\n",
    "\n",
    "        mod_counts_UMAP_df_focused[f'{column}_with_counts'] = pd.Categorical(\n",
    "            mod_counts_UMAP_df_focused[f'{column}_with_counts'],\n",
    "            categories=mod_counts_UMAP_df_focused[f'{column}_with_counts'].unique(),\n",
    "            ordered=True\n",
    "        )\n",
    "        \n",
    "        fig = px.scatter(mod_counts_UMAP_df_focused, \n",
    "                            x='UMAP1', \n",
    "                            y='UMAP2', \n",
    "                            labels={f'{column}_with_counts': column},\n",
    "                            color=f'{column}_with_counts', \n",
    "                            title=f' HAMBI - UMAP with n_neighbors={n}, min_dist={m}', \n",
    "                            color_discrete_sequence=custom_colors,\n",
    "                            hover_data={'cluster': True, 'Domain to species': True, 'Domain to genus': True,\n",
    "                                        'element': True, 'contig': True},\n",
    "                        size = 'contig_length_sqrt'\n",
    "                        )\n",
    "        title = f' HAMBI - UMAP with n_neighbors={n}, min_dist={m}'\n",
    "        fig.update_layout(\n",
    "            height=1700,\n",
    "            width=1200,\n",
    "            title_text=title,\n",
    "            showlegend=True,\n",
    "            legend=dict(\n",
    "                x=0.5,\n",
    "                y=-0.1,\n",
    "                traceorder=\"normal\",\n",
    "                xanchor='center',\n",
    "                yanchor='top',\n",
    "                orientation='h'\n",
    "            ),\n",
    "            template='simple_white',\n",
    "            xaxis=dict(\n",
    "                showgrid=True,\n",
    "                gridcolor='lightgray',\n",
    "                zeroline=False,\n",
    "                showline=True,  \n",
    "                linecolor='black', \n",
    "                linewidth=1,\n",
    "                mirror=True\n",
    "            ),\n",
    "            yaxis=dict(\n",
    "                showgrid=True,\n",
    "                gridcolor='lightgray',\n",
    "                zeroline=False,\n",
    "                showline=True,\n",
    "                linecolor='black',\n",
    "                linewidth=1,\n",
    "                mirror=True\n",
    "            )\n",
    "        )\n",
    "        fig.show()\n",
    "        #fig.write_image(f'UMAP_HAMBI/UMAP_{n}_{m}_filtered_by_100_mod_counts_focused.png')\n",
    "        #fig.write_html(f'UMAP_HAMBI/UMAP_{n}_{m}_filtered_by_100_mod_counts_focused.html')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "98c98e66-e116-4caa-81d3-6779d3f44846",
   "metadata": {
    "jp-MarkdownHeadingCollapsed": true
   },
   "source": [
    "## Extract clusters and redraw UMAP with subset of data\n",
    "### All data ðŸ’ \n",
    "#### > 100 mod counts\n",
    "##### C8"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5c479717-10d2-480c-a2c6-aac09aebe4d9",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Extract based on UMAP1 (x axis) & UMAP2 (y axis) values:\n",
    "mod_counts_UMAP_df_focused_C8 = mod_counts_UMAP_df_focused.loc[(mod_counts_UMAP_df_focused['UMAP1']>= 10) & (mod_counts_UMAP_df_focused['UMAP1']<= 14)\n",
    "    & (mod_counts_UMAP_df_focused['UMAP2']>= 3) & (mod_counts_UMAP_df_focused['UMAP2']<= 6)]\n",
    "\n",
    "# Check\n",
    "mod_counts_UMAP_df_focused_C8.head()\n",
    "\n",
    "# Save those contig IDs\n",
    "mod_counts_UMAP_df_focused_C8.index.to_list()\n",
    "\n",
    "df_mod_counts_C8 = df_mod_counts[df_mod_counts.index.isin(mod_counts_UMAP_df_focused_C8.index)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3669f7c1-3138-4df0-b512-134f72150fb7",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Save contig IDs\n",
    "mod_counts_C8_contigs = df_mod_counts_C8.index.to_list()\n",
    "\n",
    "directory = 'UMAP_HAMBI_above100'\n",
    "file_path = os.path.join(directory, 'mod_counts_C8_contigs.txt')\n",
    "\n",
    "with open(file_path, 'w') as file:\n",
    "    for item in mod_counts_C8_contigs:\n",
    "        file.write(f\"{item}\\n\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7f79be19-cef2-442a-a3aa-516ce9479e42",
   "metadata": {
    "jp-MarkdownHeadingCollapsed": true
   },
   "source": [
    "## Extract clusters and redraw UMAP with subset of data\n",
    "### All data ðŸ’ \n",
    "#### > 100 mod counts\n",
    "##### C9"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9e265dc8-1d18-4c45-bd65-bd56206a125c",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Extract based on UMAP1 (x axis) & UMAP2 (y axis) values:\n",
    "mod_counts_UMAP_df_focused_C9 = mod_counts_UMAP_df_focused.loc[(mod_counts_UMAP_df_focused['UMAP1']>= 15) & (mod_counts_UMAP_df_focused['UMAP1']<= 16)\n",
    "    & (mod_counts_UMAP_df_focused['UMAP2']>= 2) & (mod_counts_UMAP_df_focused['UMAP2']<= 4)]\n",
    "\n",
    "# Check\n",
    "mod_counts_UMAP_df_focused_C9.head()\n",
    "\n",
    "# Save those contig IDs\n",
    "mod_counts_UMAP_df_focused_C9.index.to_list()\n",
    "\n",
    "df_mod_counts_C9 = df_mod_counts[df_mod_counts.index.isin(mod_counts_UMAP_df_focused_C9.index)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "513bfc4d-600f-4852-8f0e-22446b3e6c0d",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Save contig IDs\n",
    "mod_counts_C9_contigs = df_mod_counts_C9.index.to_list()\n",
    "\n",
    "directory = 'UMAP_HAMBI_above100'\n",
    "file_path = os.path.join(directory, 'mod_counts_C9_contigs.txt')\n",
    "\n",
    "with open(file_path, 'w') as file:\n",
    "    for item in mod_counts_C9_contigs:\n",
    "        file.write(f\"{item}\\n\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "cf2945f7-6a28-49f8-86d3-3ac177b7dd07",
   "metadata": {
    "jp-MarkdownHeadingCollapsed": true
   },
   "source": [
    "## Extract clusters and redraw UMAP with subset of data\n",
    "### All data ðŸ’ \n",
    "#### > 100 mod counts\n",
    "##### C10"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e751d4b4-bef7-440d-9fc6-fb3f669bf83b",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Extract based on UMAP1 (x axis) & UMAP2 (y axis) values:\n",
    "mod_counts_UMAP_df_focused_C10 = mod_counts_UMAP_df_focused.loc[(mod_counts_UMAP_df_focused['UMAP1']>= 16) & (mod_counts_UMAP_df_focused['UMAP1']<= 16.55)\n",
    "    & (mod_counts_UMAP_df_focused['UMAP2']>= 2.8) & (mod_counts_UMAP_df_focused['UMAP2']<= 4)]\n",
    "\n",
    "# Check\n",
    "mod_counts_UMAP_df_focused_C10.head()\n",
    "\n",
    "# Save those contig IDs\n",
    "mod_counts_UMAP_df_focused_C10.index.to_list()\n",
    "\n",
    "df_mod_counts_C10 = df_mod_counts[df_mod_counts.index.isin(mod_counts_UMAP_df_focused_C10.index)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0b85c89d-b281-4483-9a84-5ea973fb6372",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Save contig IDs\n",
    "mod_counts_C10_contigs = df_mod_counts_C10.index.to_list()\n",
    "\n",
    "directory = 'UMAP_HAMBI_above100'\n",
    "file_path = os.path.join(directory, 'mod_counts_C10_contigs.txt')\n",
    "\n",
    "with open(file_path, 'w') as file:\n",
    "    for item in mod_counts_C10_contigs:\n",
    "        file.write(f\"{item}\\n\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3c9b08a8-a1d6-45f9-b329-278f08f9ed30",
   "metadata": {
    "jp-MarkdownHeadingCollapsed": true
   },
   "source": [
    "## Extract clusters and redraw UMAP with subset of data\n",
    "### All data ðŸ’ \n",
    "#### > 100 mod counts\n",
    "##### C11"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7140f235-843f-4202-bcc1-ee97c2a1495e",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Extract based on UMAP1 (x axis) & UMAP2 (y axis) values:\n",
    "mod_counts_UMAP_df_focused_C11 = mod_counts_UMAP_df_focused.loc[(mod_counts_UMAP_df_focused['UMAP1']>= 16.9) & (mod_counts_UMAP_df_focused['UMAP1']<= 17.35)\n",
    "    & (mod_counts_UMAP_df_focused['UMAP2']>= 2.2) & (mod_counts_UMAP_df_focused['UMAP2']<= 3.2)]\n",
    "\n",
    "# Check\n",
    "mod_counts_UMAP_df_focused_C11.head()\n",
    "\n",
    "# Save those contig IDs\n",
    "mod_counts_UMAP_df_focused_C11.index.to_list()\n",
    "\n",
    "df_mod_counts_C11 = df_mod_counts[df_mod_counts.index.isin(mod_counts_UMAP_df_focused_C11.index)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "33db45c4-dc91-4f5d-b4b6-b9c04d4e1480",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Save contig IDs\n",
    "mod_counts_C11_contigs = df_mod_counts_C11.index.to_list()\n",
    "\n",
    "directory = 'UMAP_HAMBI_above100'\n",
    "file_path = os.path.join(directory, 'mod_counts_C11_contigs.txt')\n",
    "\n",
    "with open(file_path, 'w') as file:\n",
    "    for item in mod_counts_C11_contigs:\n",
    "        file.write(f\"{item}\\n\")"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.14"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
