{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "70b227aa-3598-459b-9c4c-6bb1c755791e",
   "metadata": {},
   "source": [
    "# UMAP clustering for the methylation signature Position Weight Matrices"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e929d811-8da6-47d5-aab2-aeea3493be08",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "import os \n",
    "import plotly.express as px\n",
    "import plotly.graph_objects as go\n",
    "from plotly.subplots import make_subplots\n",
    "import umap\n",
    "from sklearn.metrics import silhouette_score\n",
    "import seaborn as sns\n",
    "import os\n",
    "from PIL import Image, ImageFont\n",
    "from sklearn.cluster import KMeans\n",
    "\n",
    "seed = 98\n",
    "import matplotlib.pyplot as plt\n",
    "import seaborn as sns\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3e727344-51ac-43b6-ba2a-6fd8084b80d3",
   "metadata": {},
   "outputs": [],
   "source": [
    "path_to_images = '/scratch/project_2006608/Methylation/notebooks/UMAP_HAMBI/'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e8f02c19-29cf-42fb-a5c3-04073543ebbd",
   "metadata": {},
   "outputs": [],
   "source": [
    "file_path = '/scratch/project_2006608/Methylation/HAMBI_data/merged_data.tsv'\n",
    "df = pd.read_csv(file_path, sep='\\t', index_col=0, low_memory=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b82e9f92-98b6-4560-8532-ea58a8b26c3a",
   "metadata": {},
   "outputs": [],
   "source": [
    "print(df.shape)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b111d45d-7000-4b47-94f3-c6770c9b1ca6",
   "metadata": {},
   "source": [
    "## FILTERING\n",
    "### Removing the contigs that have no methyaltions (or under 20) detected for all of the three types. These are the rows where all values in the columns are 0s\n",
    "#### All data ðŸ’ "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "62810be5-9bed-4c7c-ad52-0711a3351cf1",
   "metadata": {},
   "outputs": [],
   "source": [
    "# remove rows where df.iloc[:, :-17] meaning no methylation detected or less than 20 detected since those are converted to 0 filled matrices\n",
    "df = df.loc[(df.iloc[:, :-17] != 0).any(axis=1)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e5cd2ad6-e9ee-465e-a8f1-930127b3ab82",
   "metadata": {},
   "outputs": [],
   "source": [
    "print(df.shape[0])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "08ec1a3c-55e2-42e7-b527-cd540a3da6c5",
   "metadata": {},
   "outputs": [],
   "source": [
    "df['Domain to species'].value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "16725f1f-ca6c-47db-8ebc-cfef787c7443",
   "metadata": {},
   "outputs": [],
   "source": [
    "df['Domain to species'].nunique(dropna = False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3a089cdd-f203-4a21-98cb-05a295e515c2",
   "metadata": {},
   "outputs": [],
   "source": [
    "df['Domain to genus'].nunique(dropna = False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1ee51b71-7823-4b25-8b55-fefdea5e060b",
   "metadata": {},
   "outputs": [],
   "source": [
    "df['Domain to family'].nunique(dropna = False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "38699591-4481-45a9-98cc-f38bbe863807",
   "metadata": {},
   "outputs": [],
   "source": [
    "df['Domain to order'].nunique(dropna = False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "58e19167-47f5-4298-9c5c-b37868ad893a",
   "metadata": {},
   "outputs": [],
   "source": [
    "df['Domain to class'].nunique(dropna = False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "85aef2be-0942-4946-bfe5-cdedb4647057",
   "metadata": {},
   "outputs": [],
   "source": [
    "df['Domain to phylum'].nunique(dropna = False)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "89c92161-db56-4689-9aed-b62b7b51b26e",
   "metadata": {},
   "source": [
    "### Number of contigs where missing the species information\n",
    "#### All data ðŸ’ "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5610762d-1520-4264-a73c-a04bafa2157e",
   "metadata": {},
   "outputs": [],
   "source": [
    "len(df[df['Domain to species'].isna()])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ba820e9a-5889-4c78-a650-81fb1d65ef34",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "n_neighbors = [20, 30 ]\n",
    "min_dist = [0.01, 0.1, 0.2]\n",
    "colors = sns.color_palette(\"hls\", 32).as_hex()\n",
    "\n",
    "unique_species = df['Domain to species'].unique()\n",
    "num_species = len(unique_species)\n",
    "color_dict = [\n",
    "    \"#FF0000\", \"#00FF00\", \"#0000FF\", \"#FFFF00\", \"#FF00FF\", \"#00FFFF\", \"#800000\", \"#008000\",\n",
    "    \"#000080\", \"#808000\", \"#800080\", \"#008080\", \"#FF4500\", \"#2E8B57\", \"#4682B4\", \"#DAA520\",\n",
    "    \"#FF1493\", \"#7FFF00\", \"#8A2BE2\", \"#5F9EA0\", \"#D2691E\", \"#DC143C\", \"#B8860B\", \"#32CD32\",\n",
    "    \"#FFD700\", \"#20B2AA\", \"#FF6347\", \"#4169E1\", \"#8B4513\", \"#6A5ACD\", \"#FF7F50\", \"#2F4F4F\"\n",
    "]\n",
    "custom_colors = color_dict\n",
    "\n",
    "\n",
    "for n in n_neighbors:\n",
    "    for m in min_dist:\n",
    "        \n",
    "        reducer = umap.UMAP(n_neighbors=n, min_dist=m, random_state=seed)\n",
    "        embedding = reducer.fit_transform(df.iloc[:, :-17])\n",
    "        UMAP_df  = pd.DataFrame({\n",
    "            'UMAP1': embedding[:, 0],\n",
    "            'UMAP2': embedding[:, 1],\n",
    "            'Domain to species': df['Domain to species'],\n",
    "            'Domain to genus': df['Domain to genus'],\n",
    "            'element': df['element']\n",
    "        })\n",
    "        # clustering\n",
    "        kmeans = KMeans(n_clusters=11, random_state=seed)\n",
    "        \n",
    "        k_labels = kmeans.fit_predict(embedding)\n",
    "        \n",
    "        UMAP_df['cluster'] = k_labels\n",
    "        UMAP_df['Domain to species'] = UMAP_df['Domain to species'].fillna('NA')\n",
    "        column = 'Domain to species'\n",
    "        value_counts = UMAP_df[column].value_counts().to_dict()\n",
    "\n",
    "        UMAP_df[f'{column}_with_counts'] = UMAP_df[column].apply(lambda x: f'{x} ({value_counts[x]})')\n",
    "\n",
    "        UMAP_df['count'] = UMAP_df[column].map(value_counts)\n",
    "        UMAP_df = UMAP_df.sort_values(by=['count'], ascending=False)\n",
    "\n",
    "        UMAP_df[f'{column}_with_counts'] = pd.Categorical(\n",
    "            UMAP_df[f'{column}_with_counts'],\n",
    "            categories=UMAP_df[f'{column}_with_counts'].unique(),\n",
    "            ordered=True\n",
    "        )\n",
    "        \n",
    "        fig = px.scatter(UMAP_df, \n",
    "                            x='UMAP1', \n",
    "                            y='UMAP2', \n",
    "                            labels={f'{column}_with_counts': column},\n",
    "                            color=f'{column}_with_counts', \n",
    "                            title=f' HAMBI - UMAP with n_neighbors={n}, min_dist={m}', \n",
    "                            color_discrete_sequence=custom_colors,\n",
    "                            hover_data={'cluster': True, 'Domain to species': True, 'Domain to genus': True, 'element': True} )\n",
    "        title = f' HAMBI - UMAP with n_neighbors={n}, min_dist={m}'\n",
    "        fig.update_layout(\n",
    "            height=1500,\n",
    "            width=1200,\n",
    "            title_text=title,\n",
    "            showlegend=True,\n",
    "            legend=dict(\n",
    "                x=0.5,\n",
    "                y=-0.1,\n",
    "                traceorder=\"normal\",\n",
    "                xanchor='center',\n",
    "                yanchor='top',\n",
    "                orientation='h'\n",
    "            ),\n",
    "            template='simple_white',\n",
    "            xaxis=dict(\n",
    "                showgrid=True,\n",
    "                gridcolor='lightgray',\n",
    "                zeroline=False,\n",
    "                showline=True,  \n",
    "                linecolor='black', \n",
    "                linewidth=1,\n",
    "                mirror=True\n",
    "            ),\n",
    "            yaxis=dict(\n",
    "                showgrid=True,\n",
    "                gridcolor='lightgray',\n",
    "                zeroline=False,\n",
    "                showline=True,\n",
    "                linecolor='black',\n",
    "                linewidth=1,\n",
    "                mirror=True\n",
    "            )\n",
    "        )\n",
    "        fig.show()\n",
    "        #fig.write_image(f'UMAP_HAMBI/UMAP_{n}_{m}.png')\n",
    "        #fig.write_html(f'UMAP_HAMBI/UMAP_{n}_{m}.html')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9e8f2470-d522-4888-9544-05a309383af7",
   "metadata": {},
   "source": [
    "### Using the best n_neighbours and min_dist to get clear clusters\n",
    "#### All data ðŸ’ "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "97ab916d-9527-4d8d-9de4-c36910abd44e",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "n_neighbors = [30]\n",
    "min_dist = [0.1]\n",
    "color_dict = [\n",
    "    \"#FF0000\", \"#00FF00\", \"#0000FF\", \"#FFFF00\", \"#FF00FF\", \"#00FFFF\", \"#800000\", \"#008000\",\n",
    "    \"#000080\", \"#808000\", \"#800080\", \"#008080\", \"#FF4500\", \"#2E8B57\", \"#4682B4\", \"#DAA520\",\n",
    "    \"#FF1493\", \"#9ACD32\", \"#8A2BE2\", \"#5F9EA0\", \"#D2691E\", \"#DC143C\", \"#B8860B\", \"#32CD32\",\n",
    "    \"#FFD700\", \"#20B2AA\", \"#FF6347\", \"#4169E1\", \"#8B4513\", \"#6A5ACD\", \"#FF7F50\", \"#2F4F4F\"\n",
    "]\n",
    "\n",
    "\n",
    "custom_colors = color_dict\n",
    "for n in n_neighbors:\n",
    "    for m in min_dist:\n",
    "        reducer = umap.UMAP(n_neighbors=n, min_dist=m, random_state=seed)\n",
    "        embedding = reducer.fit_transform(df.iloc[:, :-17])\n",
    "        UMAP_df  = pd.DataFrame({\n",
    "            'UMAP1': embedding[:, 0],\n",
    "            'UMAP2': embedding[:, 1],\n",
    "            'Domain to species': df['Domain to species'],\n",
    "            'Domain to genus': df['Domain to genus'],\n",
    "            'element': df['element']\n",
    "        })\n",
    "        # clustering k defined choosing the k by looking at the UMAP \n",
    "        kmeans = KMeans(n_clusters=11, random_state=seed)\n",
    "        \n",
    "        k_labels = kmeans.fit_predict(embedding)\n",
    "        \n",
    "        UMAP_df['cluster'] = k_labels\n",
    "        \n",
    "        column = 'Domain to species'\n",
    "        UMAP_df['Domain to species'] = UMAP_df['Domain to species'].fillna('NA')\n",
    "\n",
    "        value_counts = UMAP_df[column].value_counts().to_dict()\n",
    "\n",
    "        UMAP_df[f'{column}_with_counts'] = UMAP_df[column].apply(lambda x: f'{x} ({value_counts[x]})')\n",
    "\n",
    "        UMAP_df['count'] = UMAP_df[column].map(value_counts)\n",
    "        UMAP_df = UMAP_df.sort_values(by=['count'], ascending=False)\n",
    "\n",
    "        UMAP_df[f'{column}_with_counts'] = pd.Categorical(\n",
    "            UMAP_df[f'{column}_with_counts'],\n",
    "            categories=UMAP_df[f'{column}_with_counts'].unique(),\n",
    "            ordered=True\n",
    "        )\n",
    "        \n",
    "        fig = px.scatter(UMAP_df, \n",
    "                            x='UMAP1', \n",
    "                            y='UMAP2', \n",
    "                            labels={f'{column}_with_counts': column},\n",
    "                            color=f'{column}_with_counts', \n",
    "                            title=f' HAMBI - UMAP with n_neighbors={n}, min_dist={m}', \n",
    "                            color_discrete_sequence=custom_colors,\n",
    "                            hover_data={'cluster': True, 'Domain to species': True, 'Domain to genus': True, 'element': True} )\n",
    "        title = f' HAMBI - UMAP with n_neighbors={n}, min_dist={m}'\n",
    "        fig.update_layout(\n",
    "            height=1700,\n",
    "            width=1200,\n",
    "            title_text=title,\n",
    "            showlegend=True,\n",
    "            legend=dict(\n",
    "                x=0.5,\n",
    "                y=-0.1,\n",
    "                traceorder=\"normal\",\n",
    "                xanchor='center',\n",
    "                yanchor='top',\n",
    "                orientation='h'\n",
    "            ),\n",
    "            template='simple_white',\n",
    "            xaxis=dict(\n",
    "                showgrid=True,\n",
    "                gridcolor='lightgray',\n",
    "                zeroline=False,\n",
    "                showline=True,  \n",
    "                linecolor='black', \n",
    "                linewidth=1,\n",
    "                mirror=True\n",
    "            ),\n",
    "            yaxis=dict(\n",
    "                showgrid=True,\n",
    "                gridcolor='lightgray',\n",
    "                zeroline=False,\n",
    "                showline=True,\n",
    "                linecolor='black',\n",
    "                linewidth=1,\n",
    "                mirror=True\n",
    "            )\n",
    "        )\n",
    "        fig.show()\n",
    "        #fig.write_image(f'UMAP_HAMBI/UMAP_{n}_{m}.png')\n",
    "        #fig.write_html(f'UMAP_HAMBI/UMAP_{n}_{m}.html')\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ffced132-8998-4b89-8c9b-0adc094cccff",
   "metadata": {},
   "outputs": [],
   "source": [
    "custom_colors = ['#1f77b4', '#ff7f0e', '#2ca02c', '#d62728', \n",
    "                 '#9467bd', '#8c564b', '#e377c2', '#7f7f7f', \n",
    "                 '#bcbd22', '#17becf', '#ff8da1']\n",
    "\n",
    "# Convert 'kmeans' to a categorical variable with explicit order from 0 to 10\n",
    "UMAP_df = UMAP_df.sort_values(by=['cluster'], ascending=True)\n",
    "UMAP_df['clusters'] = pd.Categorical(UMAP_df['cluster'].astype(int), categories=range(0, 11), ordered=True)\n",
    "\n",
    "\n",
    "# Plot UMAP with the kmeans clusters\n",
    "fig = px.scatter(UMAP_df, \n",
    "                    x='UMAP1', \n",
    "                    y='UMAP2', \n",
    "                    labels={f'{column}_with_counts': column},\n",
    "                    color='clusters', \n",
    "                    title=f' HAMBI - UMAP with n_neighbors={n}, min_dist={m}', \n",
    "                    color_discrete_sequence=custom_colors,\n",
    "                    hover_data={'cluster': True, 'Domain to species': True, 'Domain to genus': True, 'element': True} )\n",
    "title = f' HAMBI - UMAP with n_neighbors={n}, min_dist={m} with kmeans clusters'\n",
    "\n",
    "# Update the layout of the figure\n",
    "fig.update_layout(\n",
    "    height=1300,\n",
    "    width=1200,\n",
    "    title_text=title,\n",
    "    showlegend=True,\n",
    "    legend=dict(\n",
    "        x=0.5,\n",
    "        y=-0.1,\n",
    "        traceorder=\"normal\",\n",
    "        xanchor='center',\n",
    "        yanchor='top',\n",
    "        orientation='h'\n",
    "    ),\n",
    "    template='simple_white',\n",
    "    xaxis=dict(\n",
    "        showgrid=True,\n",
    "        gridcolor='lightgray',\n",
    "        zeroline=False,\n",
    "        showline=True,  \n",
    "        linecolor='black', \n",
    "        linewidth=1,\n",
    "        mirror=True\n",
    "    ),\n",
    "    yaxis=dict(\n",
    "        showgrid=True,\n",
    "        gridcolor='lightgray',\n",
    "        zeroline=False,\n",
    "        showline=True,\n",
    "        linecolor='black',\n",
    "        linewidth=1,\n",
    "        mirror=True\n",
    "    )\n",
    ")\n",
    "\n",
    "# Show the figure\n",
    "fig.show()\n",
    "fig.write_image(f'UMAP_HAMBI/UMAP_{n}_{m}_kmeans.png')\n",
    "fig.write_html(f'UMAP_HAMBI/UMAP_{n}_{m}_kmeans.html')\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "36dbafab-61bb-46d5-89e5-7540a4b3c197",
   "metadata": {},
   "outputs": [],
   "source": [
    "len(UMAP_df['Domain to species'].value_counts())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e2f416e2-1ec2-4da8-8014-4ac540367dbc",
   "metadata": {},
   "outputs": [],
   "source": [
    "species_cluster_count = UMAP_df.groupby(['Domain to species', 'cluster']).size().reset_index(name='Count')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b34b164a-fcdf-423f-a161-22edc7dbad82",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Pivot the table to make species as rows and clusters as columns\n",
    "species_cluster_counts = UMAP_df.groupby(['Domain to species', 'cluster']).size().reset_index(name='Count')\n",
    "\n",
    "# Pivot the table to make species as rows and clusters as columns\n",
    "species_cluster_pivot = species_cluster_counts.pivot(index='Domain to species', columns='cluster', values='Count').fillna(0)\n",
    "\n",
    "# Convert all columns to numeric to ensure proper summation\n",
    "species_cluster_pivot = species_cluster_pivot.apply(pd.to_numeric)\n",
    "\n",
    "# Calculate the row sum (total contigs per species) and add it as a new column\n",
    "species_cluster_pivot['SUM'] = species_cluster_pivot.sum(axis=1)\n",
    "species_cluster_pivot.sort_values(by='SUM', ascending=False, inplace =True)\n",
    "species_cluster_pivot"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a54d161b-0e1f-420c-a0f1-d52af0c817b1",
   "metadata": {},
   "source": [
    "### Save to file\n",
    "#### All data ðŸ’ "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8535156d-4658-4b11-b00e-09a2ad452bff",
   "metadata": {},
   "outputs": [],
   "source": [
    "species_cluster_pivot.to_csv('UMAP_HAMBI/HAMBI_species_cluster_pivot.csv', index=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c225668e-89e5-4807-81ed-9a68ba45bc73",
   "metadata": {},
   "outputs": [],
   "source": [
    "def calculate_shannon_entropy(pivot_table):\n",
    "    cluster_columns = pivot_table.columns[:-1]  \n",
    "    total_samples = pivot_table['SUM'].sum() \n",
    "\n",
    "    total_weighted_entropy = 0\n",
    "\n",
    "    for cluster in cluster_columns:\n",
    "        total_in_cluster = pivot_table[cluster].sum()\n",
    "        \n",
    "        if total_in_cluster == 0:\n",
    "            continue  \n",
    "        \n",
    "        # the proportion of each species in this cluster\n",
    "        proportions = pivot_table[cluster] / total_in_cluster\n",
    "        \n",
    "        # the entropy for this cluster\n",
    "        entropy = -np.sum(proportions * np.log2(proportions + 1e-10))  # Small value to avoid log(0)\n",
    "        \n",
    "        # Weight the entropy by the proportion of the cluster size relative to the total\n",
    "        weighted_entropy = (total_in_cluster / total_samples) * entropy\n",
    "        total_weighted_entropy += weighted_entropy\n",
    "\n",
    "        print(f\"Entropy for cluster {cluster}: {entropy:.4f}\")\n",
    "    \n",
    "    print(f\"Total Weighted Entropy: {total_weighted_entropy:.4f}\")\n",
    "    return total_weighted_entropy\n",
    "\n",
    "calculate_shannon_entropy(species_cluster_pivot)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "00605bc0-6626-459e-801f-b4b2e73215f8",
   "metadata": {},
   "source": [
    "### Connecting clusters to taxa\n",
    "#### All data ðŸ’ "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7af5aa68-358c-4308-b458-3a15c77300d6",
   "metadata": {},
   "outputs": [],
   "source": [
    "def return_index(cluster:int, df:pd.DataFrame) -> tuple:\n",
    "    \"\"\"Returns the index of the cluster in the dataframe\n",
    "    \"\"\"\n",
    "    return df[df['cluster'] == cluster]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0b8b74b2-81c3-47eb-aafd-f39dc87233e4",
   "metadata": {},
   "outputs": [],
   "source": [
    "UMAP_df['cluster'].value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "efa9e1ac-9584-4ed8-ad53-cc5813619f0b",
   "metadata": {},
   "outputs": [],
   "source": [
    "cluster_data = {}\n",
    "for cluster in UMAP_df['cluster'].unique():\n",
    "    cluster_data[cluster] = UMAP_df[UMAP_df['cluster'] == cluster].index.tolist()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "056ed069-d284-4493-8d79-a2daf93079c1",
   "metadata": {},
   "outputs": [],
   "source": [
    "cluster_data.keys()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4fdab501-4ffd-476a-8951-528f5c043a78",
   "metadata": {},
   "source": [
    "### Plot the number of each species in each cluster\n",
    "#### All data ðŸ’ "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0f8ad5fe-03d4-485f-bbe2-598dcbd46ca5",
   "metadata": {},
   "outputs": [],
   "source": [
    "for cluster in cluster_data.keys():\n",
    "    # Select the subset of UMAP_df for the current cluster\n",
    "    cluster_df = UMAP_df.loc[cluster_data[cluster]]\n",
    "    \n",
    "    plt.figure(figsize=(20, 10))\n",
    "    sns.countplot(x='Domain to species', data=cluster_df, order=cluster_df['Domain to species'].value_counts().index)\n",
    "    plt.xticks(rotation=90)\n",
    "    plt.title(f'Cluster {cluster}')\n",
    "    plt.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "47f7113c-b9c8-44d7-8ea4-2df17214c2ff",
   "metadata": {},
   "source": [
    "### Plot to how many clusters one species has been assigned by the clustering algoritms\n",
    "#### All data ðŸ’ "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "75601958-659e-4a72-9a72-de62476e39e7",
   "metadata": {},
   "outputs": [],
   "source": [
    "species_kmeans_counts = UMAP_df.groupby('Domain to species')['cluster'].nunique()\n",
    "\n",
    "species_kmeans_counts.plot(kind='bar', color='lightgreen')\n",
    "plt.title('KMEANS')\n",
    "plt.xlabel('Species')\n",
    "plt.ylabel('Number of Unique KMeans Clusters')\n",
    "\n",
    "plt.xticks(rotation=90)\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0eb3a7d2-7b62-413e-a913-a93ea1645d94",
   "metadata": {},
   "source": [
    "### Create concatenated images of the sequencing logos for each cluster and species\n",
    "#### All data ðŸ’ "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a418e8f3-98d4-43cf-bac0-8fe21c07e3c4",
   "metadata": {},
   "outputs": [],
   "source": [
    "def search_images_in_subfolders(path_to_folders, index, image_types):\n",
    "    \"\"\"Search for images by index and type from subfolders, adding placeholders if necessary.\"\"\"\n",
    "    images = {}\n",
    "    \n",
    "    subfolders = [f for f in os.listdir(path_to_folders) if os.path.isdir(os.path.join(path_to_folders, f))]\n",
    "    for img_type in image_types:\n",
    "        image_found = False\n",
    "        for subfolder in subfolders:\n",
    "            image_path = os.path.join(path_to_folders, subfolder, 'logos', f\"{index}_{img_type}.png\")            \n",
    "            if os.path.exists(image_path):\n",
    "                try:\n",
    "                    images[img_type] = Image.open(image_path)\n",
    "                    image_found = True\n",
    "                    break\n",
    "                except Exception as e:\n",
    "                    print(f\"Failed to open image {image_path}: {e}\")\n",
    "                    images[img_type] = Image.new('RGB', (1200, 300), 'white')\n",
    "                    image_found = True\n",
    "                    break\n",
    "        if not image_found:\n",
    "            images[img_type] = Image.new('RGB', (1200, 300), 'white')\n",
    "            print(f\"No image found for {img_type}, using placeholder.\")\n",
    "    \n",
    "    return images\n",
    "\n",
    "def standardize_image_sizes(images, size=(1200, 300)):\n",
    "    \"\"\"Resize all images to a standard size.\"\"\"\n",
    "    resized_images = {}\n",
    "    for key, img in images.items():\n",
    "        resized_images[key] = img.resize(size)\n",
    "    return resized_images\n",
    "\n",
    "def combine_images_horizontally(images):\n",
    "    \"\"\"Combine images horizontally.\"\"\"\n",
    "    if not images:\n",
    "        return None\n",
    "    images = standardize_image_sizes(images)\n",
    "    \n",
    "    total_width = sum(img.width for img in images.values())\n",
    "    max_height = max(img.height for img in images.values())\n",
    "\n",
    "    new_im = Image.new('RGB', (total_width, max_height), 'white')\n",
    "    x_offset = 0\n",
    "    for img_type, img in images.items():\n",
    "        new_im.paste(img, (x_offset, 0))\n",
    "        x_offset += img.width\n",
    "    return new_im\n",
    "\n",
    "def combine_images_vertically(image_list):\n",
    "    \"\"\"Combine images vertically to form a single cluster image.\"\"\"\n",
    "    if not image_list or all(img is None for img in image_list):\n",
    "        return None\n",
    "    max_width = max(img.width for img in image_list)\n",
    "    total_height = sum(img.height for img in image_list)\n",
    "    \n",
    "    new_im = Image.new('RGB', (max_width, total_height), 'white')\n",
    "    y_offset = 0\n",
    "    for img in image_list:\n",
    "        new_im.paste(img, (0, y_offset))\n",
    "        y_offset += img.height\n",
    "    return new_im\n",
    "\n",
    "\n",
    "def display_and_save_image(image, title, save_path=None):\n",
    "    \"\"\"Display an image using matplotlib and optionally save it to a file.\"\"\"\n",
    "    if image is None:\n",
    "        print(\"No image to display or save.\")\n",
    "        return\n",
    "    \n",
    "    plt.figure(figsize=(10, 5))\n",
    "    plt.imshow(image)\n",
    "    plt.title(title)\n",
    "    plt.axis('off')\n",
    "\n",
    "    if save_path:\n",
    "        try:\n",
    "            if image.width == 0 or image.height == 0:\n",
    "                print(f\"Image {title} has invalid dimensions and cannot be saved.\")\n",
    "            else:\n",
    "                image.save(save_path)\n",
    "                print(f\"Saved image to {save_path}\")\n",
    "        except Exception as e:\n",
    "            print(f\"Failed to save the image {title} at {save_path}: {e}\")\n",
    "    plt.close()\n",
    "\n",
    "def process_clusters_and_species_images(UMAP_df, path_to_folders, image_types, outputpath):\n",
    "    \"\"\"Process images based on K-means clusters and species in the DataFrame.\"\"\"\n",
    "    cluster_col = 'cluster'  # Use only the K-means clustering column\n",
    "\n",
    "    for cluster in UMAP_df[cluster_col].unique():\n",
    "        print(f\"Processing K-means cluster {cluster}\")\n",
    "        cluster_df = UMAP_df[UMAP_df[cluster_col] == cluster]\n",
    "\n",
    "        for species in cluster_df['Domain to species'].unique():\n",
    "            print(f\"Processing species {species} in K-means cluster {cluster}\")\n",
    "            species_images = []\n",
    "            indexes = cluster_df[cluster_df['Domain to species'] == species].index\n",
    "\n",
    "            for index in indexes:\n",
    "                images = search_images_in_subfolders(path_to_folders, index, image_types)\n",
    "                combined_horizontal_image = combine_images_horizontally(images)\n",
    "                if combined_horizontal_image:\n",
    "                    species_images.append(combined_horizontal_image)\n",
    "\n",
    "            final_image = combine_images_vertically(species_images)\n",
    "\n",
    "            if final_image:\n",
    "                filename = f\"kmeans_cluster_{cluster}_{species}.png\"\n",
    "                display_path = os.path.join(outputpath, filename)\n",
    "                display_and_save_image(final_image, f\"K-means Cluster {cluster} - {species}\", display_path)\n",
    "    print(\"Done\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3be9fbb5-c2b0-479e-a1ae-b3becd8e849c",
   "metadata": {},
   "outputs": [],
   "source": [
    "path_to_folders = '/scratch/project_2006608/Methylation/HAMBI_data/'\n",
    "image_types = ['m4C_logo', 'm6A_logo', 'modified_base_logo']\n",
    "outputpath = '/scratch/project_2006608/Methylation/notebooks/logos_HAMBI/'\n",
    "\n",
    "process_clusters_and_species_images(UMAP_df, path_to_folders, image_types, outputpath)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8c44c586-ee3f-4ed7-881b-ab115a7d993d",
   "metadata": {
    "jp-MarkdownHeadingCollapsed": true
   },
   "source": [
    "**************************************************************************\n",
    "### UMAP clustering for the methylation signature Position Weight Matrices\n",
    "#### Filtered data ðŸ”³"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e23299fc-a7da-4bbe-9fa5-5e646cf8bf4e",
   "metadata": {},
   "source": [
    "##### Create data set without those which have the label 'NA'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a4aa219a-d1aa-4d9a-8fdc-9868e3996650",
   "metadata": {},
   "outputs": [],
   "source": [
    "path_to_images = '/scratch/project_2006608/Methylation/notebooks/UMAP_HAMBI_filt/'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "be687bad-0196-43d0-8741-5239cdaffa25",
   "metadata": {},
   "outputs": [],
   "source": [
    "file_path = '/scratch/project_2006608/Methylation/HAMBI_data/merged_data.tsv'\n",
    "df = pd.read_csv(file_path, sep='\\t', index_col=0, low_memory=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7f261bd1-e552-4d7b-a4d6-9af42ee65a38",
   "metadata": {},
   "outputs": [],
   "source": [
    "print(df.shape)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "52f66040-6968-44e7-abc4-18ca9ab79666",
   "metadata": {},
   "outputs": [],
   "source": [
    "filt_df = df.dropna()\n",
    "filt_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8795bcf2-205d-4145-b2f6-1b920b234d3c",
   "metadata": {},
   "outputs": [],
   "source": [
    "has_nan = filt_df['All'].isna().any()\n",
    "print(has_nan)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b059a411-7586-4c21-ae91-196f52999ef6",
   "metadata": {},
   "outputs": [],
   "source": [
    "filt_df['Domain to species'].value_counts(dropna = False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "590ad973-f729-464d-876a-c36e9f8d0093",
   "metadata": {},
   "outputs": [],
   "source": [
    "len(filt_df['Domain to species'].value_counts(dropna = False))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9b6298da-bb6a-412d-8031-3269d943878e",
   "metadata": {},
   "source": [
    "## FILTERING\n",
    "### Removing the contigs that have no methyaltions (or under 20) detected for all of the three types. These are the rows where all values in the columns are 0s\n",
    "#### Filtered data ðŸ”³"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b4497316-7bda-4957-9d12-ff5da13cbf98",
   "metadata": {},
   "outputs": [],
   "source": [
    "# remove rows where df.iloc[:, :-17] meaning no methylation detected or less than 20 detected since those are converted to 0 filled matrices\n",
    "filt_df = filt_df.loc[(filt_df.iloc[:, :-17] != 0).any(axis=1)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "18a45d43-6463-4a98-8b8f-ff4a1baa504e",
   "metadata": {},
   "outputs": [],
   "source": [
    "print(filt_df.shape[0])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9f790623-1160-4084-ac62-0cde197e7884",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "filt_df['Domain to species'].value_counts(dropna = False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a4cb9c04-4d75-4913-9db8-73245ebe5fcc",
   "metadata": {},
   "outputs": [],
   "source": [
    "filt_df['Domain to species'].nunique(dropna = False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3bdc7678-1ade-4a61-9a94-42cf0bde6241",
   "metadata": {},
   "outputs": [],
   "source": [
    "filt_df['Domain to genus'].nunique(dropna = False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "edac7cd5-bbfc-4770-974f-61e2a3c297c8",
   "metadata": {},
   "outputs": [],
   "source": [
    "filt_df['Domain to family'].nunique(dropna = False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "91625c64-6890-4fba-854a-06bf714d0468",
   "metadata": {},
   "outputs": [],
   "source": [
    "filt_df['Domain to order'].nunique(dropna = False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e642bdfe-f3fa-450b-8aa1-b5ebde6591f6",
   "metadata": {},
   "outputs": [],
   "source": [
    "filt_df['Domain to class'].nunique(dropna = False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d1a8e8db-ebb8-4347-8a9e-95d70842e100",
   "metadata": {},
   "outputs": [],
   "source": [
    "filt_df['Domain to phylum'].nunique(dropna = False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "fdfb8b42-8651-43c5-b7f7-2b756298f1ce",
   "metadata": {},
   "outputs": [],
   "source": [
    "len(filt_df[filt_df['Domain to species'].isna()])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "bd23a09f-ff47-4274-916c-2b0882d24f35",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "n_neighbors = [20, 30 ]\n",
    "min_dist = [0.01, 0.1, 0.2]\n",
    "colors = sns.color_palette(\"hls\", 32).as_hex()\n",
    "\n",
    "unique_species = filt_df['Domain to species'].unique()\n",
    "num_species = len(unique_species)\n",
    "color_dict = [\n",
    "    \"#FF0000\", \"#00FF00\", \"#0000FF\", \"#FFFF00\", \"#FF00FF\", \"#00FFFF\", \"#800000\", \"#008000\",\n",
    "    \"#000080\", \"#808000\", \"#800080\", \"#008080\", \"#FF4500\", \"#2E8B57\", \"#4682B4\", \"#DAA520\",\n",
    "    \"#FF1493\", \"#7FFF00\", \"#8A2BE2\", \"#5F9EA0\", \"#D2691E\", \"#DC143C\", \"#B8860B\", \"#32CD32\",\n",
    "    \"#FFD700\", \"#20B2AA\", \"#FF6347\", \"#4169E1\", \"#8B4513\", \"#6A5ACD\", \"#FF7F50\", \"#2F4F4F\"\n",
    "]\n",
    "custom_colors = color_dict\n",
    "\n",
    "\n",
    "for n in n_neighbors:\n",
    "    for m in min_dist:\n",
    "        \n",
    "        reducer = umap.UMAP(n_neighbors=n, min_dist=m, random_state=seed)\n",
    "        embedding = reducer.fit_transform(filt_df.iloc[:, :-17])\n",
    "        filt_UMAP_df  = pd.DataFrame({\n",
    "            'UMAP1': embedding[:, 0],\n",
    "            'UMAP2': embedding[:, 1],\n",
    "            'Domain to species': filt_df['Domain to species'],\n",
    "            'Domain to genus': filt_df['Domain to genus'],\n",
    "            'element': filt_df['element']\n",
    "        })\n",
    "        # clustering\n",
    "        kmeans = KMeans(n_clusters=11, random_state=seed)\n",
    "        \n",
    "        k_labels = kmeans.fit_predict(embedding)\n",
    "        \n",
    "        filt_UMAP_df['cluster'] = k_labels\n",
    "        filt_UMAP_df['Domain to species'] = filt_UMAP_df['Domain to species'].fillna('NA')\n",
    "        column = 'Domain to species'\n",
    "        value_counts = filt_UMAP_df[column].value_counts().to_dict()\n",
    "\n",
    "        filt_UMAP_df[f'{column}_with_counts'] = filt_UMAP_df[column].apply(lambda x: f'{x} ({value_counts[x]})')\n",
    "\n",
    "        filt_UMAP_df['count'] = filt_UMAP_df[column].map(value_counts)\n",
    "        filt_UMAP_df = filt_UMAP_df.sort_values(by=['count'], ascending=False)\n",
    "\n",
    "        filt_UMAP_df[f'{column}_with_counts'] = pd.Categorical(\n",
    "            filt_UMAP_df[f'{column}_with_counts'],\n",
    "            categories=filt_UMAP_df[f'{column}_with_counts'].unique(),\n",
    "            ordered=True\n",
    "        )\n",
    "        \n",
    "        fig = px.scatter(filt_UMAP_df, \n",
    "                            x='UMAP1', \n",
    "                            y='UMAP2', \n",
    "                            labels={f'{column}_with_counts': column},\n",
    "                            color=f'{column}_with_counts', \n",
    "                            title=f' HAMBI - UMAP with n_neighbors={n}, min_dist={m}', \n",
    "                            color_discrete_sequence=custom_colors,\n",
    "                            hover_data={'cluster': True, 'Domain to species': True, 'Domain to genus': True, 'element': True} )\n",
    "        title = f' HAMBI - UMAP with n_neighbors={n}, min_dist={m}'\n",
    "        fig.update_layout(\n",
    "            height=1500,\n",
    "            width=1200,\n",
    "            title_text=title,\n",
    "            showlegend=True,\n",
    "            legend=dict(\n",
    "                x=0.5,\n",
    "                y=-0.1,\n",
    "                traceorder=\"normal\",\n",
    "                xanchor='center',\n",
    "                yanchor='top',\n",
    "                orientation='h'\n",
    "            ),\n",
    "            template='simple_white',\n",
    "            xaxis=dict(\n",
    "                showgrid=True,\n",
    "                gridcolor='lightgray',\n",
    "                zeroline=False,\n",
    "                showline=True,  \n",
    "                linecolor='black', \n",
    "                linewidth=1,\n",
    "                mirror=True\n",
    "            ),\n",
    "            yaxis=dict(\n",
    "                showgrid=True,\n",
    "                gridcolor='lightgray',\n",
    "                zeroline=False,\n",
    "                showline=True,\n",
    "                linecolor='black',\n",
    "                linewidth=1,\n",
    "                mirror=True\n",
    "            )\n",
    "        )\n",
    "        fig.show()\n",
    "        #fig.write_image(f'UMAP_HAMBI_filt/UMAP_{n}_{m}.png')\n",
    "        #fig.write_html(f'UMAP_HAMBI_filt/UMAP_{n}_{m}.html')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "532d373e-229e-4f2d-a458-2733b7b7311a",
   "metadata": {},
   "source": [
    "### Using the best n_neighbours and min_dist to get clear clusters\n",
    "#### Filtered data ðŸ”³"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "03ace6c1-bc24-4568-b371-e1126df15ed0",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "n_neighbors = [30]\n",
    "min_dist = [0.1]\n",
    "color_dict = [\n",
    "    \"#FF0000\", \"#00FF00\", \"#0000FF\", \"#FFFF00\", \"#FF00FF\", \"#00FFFF\", \"#800000\", \"#008000\",\n",
    "    \"#000080\", \"#808000\", \"#800080\", \"#008080\", \"#FF4500\", \"#2E8B57\", \"#4682B4\", \"#DAA520\",\n",
    "    \"#FF1493\", \"#9ACD32\", \"#8A2BE2\", \"#5F9EA0\", \"#D2691E\", \"#DC143C\", \"#B8860B\", \"#32CD32\",\n",
    "    \"#FFD700\", \"#20B2AA\", \"#FF6347\", \"#4169E1\", \"#8B4513\", \"#6A5ACD\", \"#FF7F50\", \"#2F4F4F\"\n",
    "]\n",
    "\n",
    "\n",
    "custom_colors = color_dict\n",
    "for n in n_neighbors:\n",
    "    for m in min_dist:\n",
    "        reducer = umap.UMAP(n_neighbors=n, min_dist=m, random_state=seed)\n",
    "        embedding = reducer.fit_transform(filt_df.iloc[:, :-17])\n",
    "        filt_UMAP_df  = pd.DataFrame({\n",
    "            'UMAP1': embedding[:, 0],\n",
    "            'UMAP2': embedding[:, 1],\n",
    "            'Domain to species': filt_df['Domain to species'],\n",
    "            'Domain to genus': filt_df['Domain to genus'],\n",
    "            'element': filt_df['element']\n",
    "        })\n",
    "        # clustering k defined choosing the k by looking at the UMAP \n",
    "        kmeans = KMeans(n_clusters=11, random_state=seed)\n",
    "        \n",
    "        k_labels = kmeans.fit_predict(embedding)\n",
    "        \n",
    "        filt_UMAP_df['cluster'] = k_labels\n",
    "        \n",
    "        column = 'Domain to species'\n",
    "        filt_UMAP_df['Domain to species'] = filt_UMAP_df['Domain to species'].fillna('NA')\n",
    "\n",
    "        value_counts = filt_UMAP_df[column].value_counts().to_dict()\n",
    "\n",
    "        filt_UMAP_df[f'{column}_with_counts'] = filt_UMAP_df[column].apply(lambda x: f'{x} ({value_counts[x]})')\n",
    "\n",
    "        filt_UMAP_df['count'] = filt_UMAP_df[column].map(value_counts)\n",
    "        filt_UMAP_df = filt_UMAP_df.sort_values(by=['count'], ascending=False)\n",
    "\n",
    "        filt_UMAP_df[f'{column}_with_counts'] = pd.Categorical(\n",
    "            filt_UMAP_df[f'{column}_with_counts'],\n",
    "            categories=filt_UMAP_df[f'{column}_with_counts'].unique(),\n",
    "            ordered=True\n",
    "        )\n",
    "        \n",
    "        fig = px.scatter(filt_UMAP_df, \n",
    "                            x='UMAP1', \n",
    "                            y='UMAP2', \n",
    "                            labels={f'{column}_with_counts': column},\n",
    "                            color=f'{column}_with_counts', \n",
    "                            title=f' HAMBI - UMAP with n_neighbors={n}, min_dist={m}', \n",
    "                            color_discrete_sequence=custom_colors,\n",
    "                            hover_data={'cluster': True, 'Domain to species': True, 'Domain to genus': True, 'element': True} )\n",
    "        title = f' HAMBI - UMAP with n_neighbors={n}, min_dist={m}'\n",
    "        fig.update_layout(\n",
    "            height=1700,\n",
    "            width=1200,\n",
    "            title_text=title,\n",
    "            showlegend=True,\n",
    "            legend=dict(\n",
    "                x=0.5,\n",
    "                y=-0.1,\n",
    "                traceorder=\"normal\",\n",
    "                xanchor='center',\n",
    "                yanchor='top',\n",
    "                orientation='h'\n",
    "            ),\n",
    "            template='simple_white',\n",
    "            xaxis=dict(\n",
    "                showgrid=True,\n",
    "                gridcolor='lightgray',\n",
    "                zeroline=False,\n",
    "                showline=True,  \n",
    "                linecolor='black', \n",
    "                linewidth=1,\n",
    "                mirror=True\n",
    "            ),\n",
    "            yaxis=dict(\n",
    "                showgrid=True,\n",
    "                gridcolor='lightgray',\n",
    "                zeroline=False,\n",
    "                showline=True,\n",
    "                linecolor='black',\n",
    "                linewidth=1,\n",
    "                mirror=True\n",
    "            )\n",
    "        )\n",
    "        fig.show()\n",
    "        fig.write_image(f'UMAP_HAMBI_filt/UMAP_{n}_{m}.png')\n",
    "        fig.write_html(f'UMAP_HAMBI_filt/UMAP_{n}_{m}.html')\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6f50f704-0220-483a-92ad-fa7822805de2",
   "metadata": {},
   "outputs": [],
   "source": [
    "custom_colors = ['#1f77b4', '#ff7f0e', '#2ca02c', '#d62728', \n",
    "                 '#9467bd', '#8c564b', '#e377c2', '#7f7f7f', \n",
    "                 '#bcbd22', '#17becf', '#ff8da1']\n",
    "\n",
    "# Convert 'kmeans' to a categorical variable with explicit order from 0 to 10\n",
    "filt_UMAP_df = filt_UMAP_df.sort_values(by=['cluster'], ascending=True)\n",
    "filt_UMAP_df['clusters'] = pd.Categorical(filt_UMAP_df['cluster'].astype(int), categories=range(0, 11), ordered=True)\n",
    "\n",
    "\n",
    "# Plot UMAP with the kmeans clusters\n",
    "fig = px.scatter(filt_UMAP_df, \n",
    "                    x='UMAP1', \n",
    "                    y='UMAP2', \n",
    "                    labels={f'{column}_with_counts': column},\n",
    "                    color='clusters', \n",
    "                    title=f' HAMBI - UMAP with n_neighbors={n}, min_dist={m}', \n",
    "                    color_discrete_sequence=custom_colors,\n",
    "                    hover_data={'cluster': True, 'Domain to species': True, 'Domain to genus': True, 'element': True} )\n",
    "title = f' HAMBI - UMAP with n_neighbors={n}, min_dist={m} with kmeans clusters'\n",
    "\n",
    "# Update the layout of the figure\n",
    "fig.update_layout(\n",
    "    height=1300,\n",
    "    width=1200,\n",
    "    title_text=title,\n",
    "    showlegend=True,\n",
    "    legend=dict(\n",
    "        x=0.5,\n",
    "        y=-0.1,\n",
    "        traceorder=\"normal\",\n",
    "        xanchor='center',\n",
    "        yanchor='top',\n",
    "        orientation='h'\n",
    "    ),\n",
    "    template='simple_white',\n",
    "    xaxis=dict(\n",
    "        showgrid=True,\n",
    "        gridcolor='lightgray',\n",
    "        zeroline=False,\n",
    "        showline=True,  \n",
    "        linecolor='black', \n",
    "        linewidth=1,\n",
    "        mirror=True\n",
    "    ),\n",
    "    yaxis=dict(\n",
    "        showgrid=True,\n",
    "        gridcolor='lightgray',\n",
    "        zeroline=False,\n",
    "        showline=True,\n",
    "        linecolor='black',\n",
    "        linewidth=1,\n",
    "        mirror=True\n",
    "    )\n",
    ")\n",
    "\n",
    "# Show the figure\n",
    "fig.show()\n",
    "fig.write_image(f'UMAP_HAMBI_filt/UMAP_{n}_{m}_kmeans.png')\n",
    "fig.write_html(f'UMAP_HAMBI_filt/UMAP_{n}_{m}_kmeans.html')\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8104dc40-fed7-4fb3-9b2a-7c7ab86650c0",
   "metadata": {},
   "outputs": [],
   "source": [
    "species_cluster_count = filt_UMAP_df.groupby(['Domain to species', 'cluster']).size().reset_index(name='Count')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "fec3eed1-068f-4e61-9c79-87a6fd9ad01b",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "# Pivot the table to make species as rows and clusters as columns\n",
    "species_cluster_counts = filt_UMAP_df.groupby(['Domain to species', 'cluster']).size().reset_index(name='Count')\n",
    "\n",
    "# Pivot the table to make species as rows and clusters as columns\n",
    "species_cluster_pivot = species_cluster_counts.pivot(index='Domain to species', columns='cluster', values='Count').fillna(0)\n",
    "\n",
    "# Convert all columns to numeric to ensure proper summation\n",
    "species_cluster_pivot = species_cluster_pivot.apply(pd.to_numeric)\n",
    "\n",
    "# Calculate the row sum (total contigs per species) and add it as a new column\n",
    "species_cluster_pivot['SUM'] = species_cluster_pivot.sum(axis=1)\n",
    "species_cluster_pivot.sort_values(by='SUM', ascending=False, inplace =True)\n",
    "species_cluster_pivot"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9f366bbf-e425-4578-bc59-25beff0545b7",
   "metadata": {},
   "source": [
    "### Save to file\n",
    "#### Filtered data ðŸ”³"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e669801e-8478-4753-bede-74f1993ca553",
   "metadata": {},
   "outputs": [],
   "source": [
    "species_cluster_pivot.to_csv('UMAP_HAMBI_filt/HAMBI_species_cluster_pivot_filt.csv', index=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "eb8e4e07-5679-4ea8-a62e-d29c0d073e9b",
   "metadata": {},
   "outputs": [],
   "source": [
    "def calculate_shannon_entropy(pivot_table):\n",
    "    cluster_columns = pivot_table.columns[:-1]  \n",
    "    total_samples = pivot_table['SUM'].sum() \n",
    "\n",
    "    total_weighted_entropy = 0\n",
    "\n",
    "    for cluster in cluster_columns:\n",
    "        total_in_cluster = pivot_table[cluster].sum()\n",
    "        \n",
    "        if total_in_cluster == 0:\n",
    "            continue  \n",
    "        \n",
    "        # the proportion of each species in this cluster\n",
    "        proportions = pivot_table[cluster] / total_in_cluster\n",
    "        \n",
    "        # the entropy for this cluster\n",
    "        entropy = -np.sum(proportions * np.log2(proportions + 1e-10))  # Small value to avoid log(0)\n",
    "        \n",
    "        # Weight the entropy by the proportion of the cluster size relative to the total\n",
    "        weighted_entropy = (total_in_cluster / total_samples) * entropy\n",
    "        total_weighted_entropy += weighted_entropy\n",
    "\n",
    "        print(f\"Entropy for cluster {cluster}: {entropy:.4f}\")\n",
    "    \n",
    "    print(f\"Total Weighted Entropy: {total_weighted_entropy:.4f}\")\n",
    "    return total_weighted_entropy\n",
    "\n",
    "calculate_shannon_entropy(species_cluster_pivot)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "cfe12460-30fc-4e68-90f3-30506b7bfc7a",
   "metadata": {},
   "source": [
    "### Connecting clusters to taxa\n",
    "#### Filtered data ðŸ”³"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "dfcb7add-33b5-4a87-8cc1-cb6c3cdf3354",
   "metadata": {},
   "outputs": [],
   "source": [
    "def return_index(cluster:int, df:pd.DataFrame) -> tuple:\n",
    "    \"\"\"Returns the index of the cluster in the dataframe\n",
    "    \"\"\"\n",
    "    return filt_df[filt_df['cluster'] == cluster]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "49637db9-7cc0-4c04-b9f1-bddbc7c50aa2",
   "metadata": {},
   "outputs": [],
   "source": [
    "filt_UMAP_df['cluster'].value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "bc16dbab-e9cd-4a2f-955e-a439432f19fa",
   "metadata": {},
   "outputs": [],
   "source": [
    "cluster_data = {}\n",
    "for cluster in filt_UMAP_df['cluster'].unique():\n",
    "    cluster_data[cluster] = filt_UMAP_df[filt_UMAP_df['cluster'] == cluster].index.tolist()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "594841e0-8642-478a-8005-ee9ca3afde4d",
   "metadata": {},
   "outputs": [],
   "source": [
    "cluster_data.keys()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "77ecebe9-4e0f-4ce8-adf9-3263d59b8ec3",
   "metadata": {},
   "source": [
    "### Plot the number of each species in each cluster\n",
    "#### Filtered data ðŸ”³"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8e2d4405-c523-42c0-81c4-1e0f4b48fc6f",
   "metadata": {},
   "outputs": [],
   "source": [
    "for cluster in cluster_data.keys():\n",
    "    # Select the subset of filt_UMAP_df for the current cluster\n",
    "    cluster_df = filt_UMAP_df.loc[cluster_data[cluster]]\n",
    "    \n",
    "    plt.figure(figsize=(20, 10))\n",
    "    sns.countplot(x='Domain to species', data=cluster_df, order=cluster_df['Domain to species'].value_counts().index)\n",
    "    plt.xticks(rotation=90)\n",
    "    plt.title(f'Cluster {cluster}')\n",
    "    plt.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d9787353-94c6-4926-ba7f-45192e9ca4c1",
   "metadata": {},
   "source": [
    "### Plot to how many clusters one species has been assigned by the clustering algoritms\n",
    "#### Filtered data ðŸ”³"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c649add3-5c92-47a0-b3e2-1dbecbc5ac11",
   "metadata": {},
   "outputs": [],
   "source": [
    "species_kmeans_counts = filt_UMAP_df.groupby('Domain to species')['cluster'].nunique()\n",
    "\n",
    "species_kmeans_counts.plot(kind='bar', color='lightgreen')\n",
    "plt.title('KMEANS')\n",
    "plt.xlabel('Species')\n",
    "plt.ylabel('Number of Unique KMeans Clusters')\n",
    "\n",
    "plt.xticks(rotation=90)\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f93b8b9d-f699-4c23-8dfa-c2e74b7b2a10",
   "metadata": {},
   "source": [
    "**************************************************************************\n",
    "### UMAP clustering for the methylation signature Position Weight Matrices\n",
    "#### Top feature data ðŸ”"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8f463e50-7b9c-4e58-855d-2023d29d0f2e",
   "metadata": {},
   "source": [
    "##### Create data set with only top features (70)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "818f58a8-70f7-41eb-963d-f550d5954955",
   "metadata": {},
   "outputs": [],
   "source": [
    "path_to_images = '/scratch/project_2006608/Methylation/notebooks/UMAP_HAMBI_top_features/'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f9b5931a-61db-464f-a73b-a45da5431caf",
   "metadata": {},
   "outputs": [],
   "source": [
    "file_path = '/scratch/project_2006608/Methylation/HAMBI_data/merged_data.tsv'\n",
    "df = pd.read_csv(file_path, sep='\\t', index_col=0, low_memory=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1f4568fb-ef9d-4ce3-949a-2fddfa4adae5",
   "metadata": {},
   "outputs": [],
   "source": [
    "print(df.shape)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f61a43fa-4c87-4245-9132-1feaf54c3b3c",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Filter based on the features in 'RF_HAMBI/Domain_to_species_top_features.csv'\n",
    "## Load file\n",
    "file_path = '/scratch/project_2006608/Methylation/notebooks/RF_HAMBI/Domain_to_species_top_features.csv'\n",
    "feature_list = pd.read_csv(file_path, sep=',')\n",
    "feature_list.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e29225ef-d691-437c-bbc1-59ed2b4a3742",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Extract identifiers into list\n",
    "taxa_headers = df.columns[493:510]\n",
    "column_headers = pd.DataFrame(taxa_headers.tolist(), columns=['Feature'])\n",
    "print(column_headers.head())\n",
    "\n",
    "# Append the column_headers to it\n",
    "feature_list = pd.concat([feature_list, column_headers], ignore_index=True)\n",
    "\n",
    "print(feature_list)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8f987c9a-e25f-4c25-8a5c-c40d99604201",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Extract columns\n",
    "columns_to_keep = feature_list['Feature'].tolist()\n",
    "df_top70 = df[columns_to_keep]\n",
    "df_top70.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "176ace0c-15e3-4034-9a76-485634e56dc8",
   "metadata": {},
   "outputs": [],
   "source": [
    "df_top70['Domain to species'].value_counts(dropna = False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f3ad4400-89d5-40a9-86e5-392332787a56",
   "metadata": {},
   "outputs": [],
   "source": [
    "row_names = df_top70.index\n",
    "print(row_names)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "98c9ab88-143a-45fe-8c77-e8b064b54224",
   "metadata": {},
   "outputs": [],
   "source": [
    "len(df_top70['Domain to species'].value_counts(dropna = False))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "10727b87-df47-469f-8f5b-518f217f25d7",
   "metadata": {},
   "source": [
    "## FILTERING\n",
    "### Removing the contigs that have no methyaltions (or under 20) detected for all of the three types. These are the rows where all values in the columns are 0s\n",
    "#### Top feature data ðŸ”"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "77e816b7-7e11-46eb-ae0b-8bd608d4277a",
   "metadata": {},
   "outputs": [],
   "source": [
    "# remove rows where df.iloc[:, :-17] meaning no methylation detected or less than 20 detected since those are converted to 0 filled matrices\n",
    "df_top70 = df_top70.loc[(df_top70.iloc[:, :-17] != 0).any(axis=1)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3dfdcf54-52a1-4a09-92e9-efb9dbe44c76",
   "metadata": {},
   "outputs": [],
   "source": [
    "print(df_top70.shape[0])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "be3b2961-6c90-4ef2-9e3a-1d3f5f038951",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "df_top70['Domain to species'].value_counts(dropna = False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "01fb0c7f-cab1-4955-9e1d-a2b3c78945f5",
   "metadata": {},
   "outputs": [],
   "source": [
    "df_top70['Domain to species'].nunique(dropna = False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "99d53ffc-22b2-4023-9f67-2ed408fd0483",
   "metadata": {},
   "outputs": [],
   "source": [
    "df_top70['Domain to genus'].nunique(dropna = False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c35fd948-72c8-4019-beb8-c7917af2bd79",
   "metadata": {},
   "outputs": [],
   "source": [
    "df_top70['Domain to family'].nunique(dropna = False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b9ab47f5-3de5-4f5f-a3d0-944307a76285",
   "metadata": {},
   "outputs": [],
   "source": [
    "df_top70['Domain to order'].nunique(dropna = False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5a4dd626-9483-40ef-8d78-5a898384e686",
   "metadata": {},
   "outputs": [],
   "source": [
    "df_top70['Domain to class'].nunique(dropna = False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "06ecbd11-7749-4815-8b1a-4a385ea0e6b7",
   "metadata": {},
   "outputs": [],
   "source": [
    "df_top70['Domain to phylum'].nunique(dropna = False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f970d98a-94ec-4224-9f13-fb2789e3485f",
   "metadata": {},
   "outputs": [],
   "source": [
    "len(df_top70[df_top70['Domain to species'].isna()])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5c06a52e-bcf9-4075-8d1f-f517d81b9150",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "n_neighbors = [20, 30 ]\n",
    "min_dist = [0.01, 0.1, 0.2]\n",
    "colors = sns.color_palette(\"hls\", 32).as_hex()\n",
    "\n",
    "unique_species = df_top70['Domain to species'].unique()\n",
    "num_species = len(unique_species)\n",
    "color_dict = [\n",
    "    \"#FF0000\", \"#00FF00\", \"#0000FF\", \"#FFFF00\", \"#FF00FF\", \"#00FFFF\", \"#800000\", \"#008000\",\n",
    "    \"#000080\", \"#808000\", \"#800080\", \"#008080\", \"#FF4500\", \"#2E8B57\", \"#4682B4\", \"#DAA520\",\n",
    "    \"#FF1493\", \"#7FFF00\", \"#8A2BE2\", \"#5F9EA0\", \"#D2691E\", \"#DC143C\", \"#B8860B\", \"#32CD32\",\n",
    "    \"#FFD700\", \"#20B2AA\", \"#FF6347\", \"#4169E1\", \"#8B4513\", \"#6A5ACD\", \"#FF7F50\", \"#2F4F4F\"\n",
    "]\n",
    "custom_colors = color_dict\n",
    "\n",
    "\n",
    "for n in n_neighbors:\n",
    "    for m in min_dist:\n",
    "        \n",
    "        reducer = umap.UMAP(n_neighbors=n, min_dist=m, random_state=seed)\n",
    "        embedding = reducer.fit_transform(df_top70.iloc[:, :-17])\n",
    "        top70_UMAP_df  = pd.DataFrame({\n",
    "            'UMAP1': embedding[:, 0],\n",
    "            'UMAP2': embedding[:, 1],\n",
    "            'Domain to species': df_top70['Domain to species'],\n",
    "            'Domain to genus': df_top70['Domain to genus'],\n",
    "            'element': df_top70['element'],\n",
    "            'contig': df_top70.index,\n",
    "        })\n",
    "        # clustering\n",
    "        kmeans = KMeans(n_clusters=11, random_state=seed)\n",
    "        \n",
    "        k_labels = kmeans.fit_predict(embedding)\n",
    "        \n",
    "        top70_UMAP_df['cluster'] = k_labels\n",
    "        top70_UMAP_df['Domain to species'] = top70_UMAP_df['Domain to species'].fillna('NA')\n",
    "        column = 'Domain to species'\n",
    "        value_counts = top70_UMAP_df[column].value_counts().to_dict()\n",
    "\n",
    "        top70_UMAP_df[f'{column}_with_counts'] = top70_UMAP_df[column].apply(lambda x: f'{x} ({value_counts[x]})')\n",
    "\n",
    "        top70_UMAP_df['count'] = top70_UMAP_df[column].map(value_counts)\n",
    "        top70_UMAP_df = top70_UMAP_df.sort_values(by=['count'], ascending=False)\n",
    "\n",
    "        top70_UMAP_df[f'{column}_with_counts'] = pd.Categorical(\n",
    "            top70_UMAP_df[f'{column}_with_counts'],\n",
    "            categories=top70_UMAP_df[f'{column}_with_counts'].unique(),\n",
    "            ordered=True\n",
    "        )\n",
    "        \n",
    "        fig = px.scatter(top70_UMAP_df, \n",
    "                            x='UMAP1', \n",
    "                            y='UMAP2', \n",
    "                            labels={f'{column}_with_counts': column},\n",
    "                            color=f'{column}_with_counts', \n",
    "                            title=f' HAMBI - UMAP with n_neighbors={n}, min_dist={m}', \n",
    "                            color_discrete_sequence=custom_colors,\n",
    "                            hover_data={'cluster': True, 'Domain to species': True, 'Domain to genus': True, 'element': True, 'contig': True} )\n",
    "        title = f' HAMBI - UMAP with n_neighbors={n}, min_dist={m}'\n",
    "        fig.update_layout(\n",
    "            height=1500,\n",
    "            width=1200,\n",
    "            title_text=title,\n",
    "            showlegend=True,\n",
    "            legend=dict(\n",
    "                x=0.5,\n",
    "                y=-0.1,\n",
    "                traceorder=\"normal\",\n",
    "                xanchor='center',\n",
    "                yanchor='top',\n",
    "                orientation='h'\n",
    "            ),\n",
    "            template='simple_white',\n",
    "            xaxis=dict(\n",
    "                showgrid=True,\n",
    "                gridcolor='lightgray',\n",
    "                zeroline=False,\n",
    "                showline=True,  \n",
    "                linecolor='black', \n",
    "                linewidth=1,\n",
    "                mirror=True\n",
    "            ),\n",
    "            yaxis=dict(\n",
    "                showgrid=True,\n",
    "                gridcolor='lightgray',\n",
    "                zeroline=False,\n",
    "                showline=True,\n",
    "                linecolor='black',\n",
    "                linewidth=1,\n",
    "                mirror=True\n",
    "            )\n",
    "        )\n",
    "        fig.show()\n",
    "        #fig.write_image(f'UMAP_HAMBI_top_features/UMAP_{n}_{m}.png')\n",
    "        #fig.write_html(f'UMAP_HAMBI_top_features/UMAP_{n}_{m}.html')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f44e9551-cdbc-44c4-b560-1959579535ea",
   "metadata": {},
   "source": [
    "### Using the best n_neighbours and min_dist to get clear clusters\n",
    "#### Top feature data ðŸ”"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "730b44d1-5fca-4e1f-bec7-dfafc867b8a6",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "n_neighbors = [30]\n",
    "min_dist = [0.1]\n",
    "color_dict = [\n",
    "    \"#FF0000\", \"#00FF00\", \"#0000FF\", \"#FFFF00\", \"#FF00FF\", \"#00FFFF\", \"#800000\", \"#008000\",\n",
    "    \"#000080\", \"#808000\", \"#800080\", \"#008080\", \"#FF4500\", \"#2E8B57\", \"#4682B4\", \"#DAA520\",\n",
    "    \"#FF1493\", \"#9ACD32\", \"#8A2BE2\", \"#5F9EA0\", \"#D2691E\", \"#DC143C\", \"#B8860B\", \"#32CD32\",\n",
    "    \"#FFD700\", \"#20B2AA\", \"#FF6347\", \"#4169E1\", \"#8B4513\", \"#6A5ACD\", \"#FF7F50\", \"#2F4F4F\"\n",
    "]\n",
    "\n",
    "\n",
    "custom_colors = color_dict\n",
    "for n in n_neighbors:\n",
    "    for m in min_dist:\n",
    "        reducer = umap.UMAP(n_neighbors=n, min_dist=m, random_state=seed)\n",
    "        embedding = reducer.fit_transform(df_top70.iloc[:, :-17])\n",
    "        top70_UMAP_df  = pd.DataFrame({\n",
    "            'UMAP1': embedding[:, 0],\n",
    "            'UMAP2': embedding[:, 1],\n",
    "            'Domain to species': df_top70['Domain to species'],\n",
    "            'Domain to genus': df_top70['Domain to genus'],\n",
    "            'element': df_top70['element'],\n",
    "            'contig': df_top70.index,\n",
    "        })\n",
    "        # clustering k defined choosing the k by looking at the UMAP \n",
    "        kmeans = KMeans(n_clusters=11, random_state=seed)\n",
    "        \n",
    "        k_labels = kmeans.fit_predict(embedding)\n",
    "        \n",
    "        top70_UMAP_df['cluster'] = k_labels\n",
    "        \n",
    "        column = 'Domain to species'\n",
    "        top70_UMAP_df['Domain to species'] = top70_UMAP_df['Domain to species'].fillna('NA')\n",
    "\n",
    "        value_counts = top70_UMAP_df[column].value_counts().to_dict()\n",
    "\n",
    "        top70_UMAP_df[f'{column}_with_counts'] = top70_UMAP_df[column].apply(lambda x: f'{x} ({value_counts[x]})')\n",
    "\n",
    "        top70_UMAP_df['count'] = top70_UMAP_df[column].map(value_counts)\n",
    "        top70_UMAP_df = top70_UMAP_df.sort_values(by=['count'], ascending=False)\n",
    "\n",
    "        top70_UMAP_df[f'{column}_with_counts'] = pd.Categorical(\n",
    "            top70_UMAP_df[f'{column}_with_counts'],\n",
    "            categories=top70_UMAP_df[f'{column}_with_counts'].unique(),\n",
    "            ordered=True\n",
    "        )\n",
    "        \n",
    "        fig = px.scatter(top70_UMAP_df, \n",
    "                            x='UMAP1', \n",
    "                            y='UMAP2', \n",
    "                            labels={f'{column}_with_counts': column},\n",
    "                            color=f'{column}_with_counts', \n",
    "                            title=f' HAMBI - UMAP with n_neighbors={n}, min_dist={m}', \n",
    "                            color_discrete_sequence=custom_colors,\n",
    "                            hover_data={'cluster': True, 'Domain to species': True, 'Domain to genus': True, 'element': True, 'contig': True} )\n",
    "        title = f' HAMBI - UMAP with n_neighbors={n}, min_dist={m}'\n",
    "        fig.update_layout(\n",
    "            height=1700,\n",
    "            width=1200,\n",
    "            title_text=title,\n",
    "            showlegend=True,\n",
    "            legend=dict(\n",
    "                x=0.5,\n",
    "                y=-0.1,\n",
    "                traceorder=\"normal\",\n",
    "                xanchor='center',\n",
    "                yanchor='top',\n",
    "                orientation='h'\n",
    "            ),\n",
    "            template='simple_white',\n",
    "            xaxis=dict(\n",
    "                showgrid=True,\n",
    "                gridcolor='lightgray',\n",
    "                zeroline=False,\n",
    "                showline=True,  \n",
    "                linecolor='black', \n",
    "                linewidth=1,\n",
    "                mirror=True\n",
    "            ),\n",
    "            yaxis=dict(\n",
    "                showgrid=True,\n",
    "                gridcolor='lightgray',\n",
    "                zeroline=False,\n",
    "                showline=True,\n",
    "                linecolor='black',\n",
    "                linewidth=1,\n",
    "                mirror=True\n",
    "            )\n",
    "        )\n",
    "        fig.show()\n",
    "        #fig.write_image(f'UMAP_HAMBI_top_features/UMAP_{n}_{m}.png')\n",
    "        #fig.write_html(f'UMAP_HAMBI_top_features/UMAP_{n}_{m}.html')\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5347b4e9-518d-4fdb-b38e-d3c3d8364dcf",
   "metadata": {},
   "outputs": [],
   "source": [
    "custom_colors = ['#1f77b4', '#ff7f0e', '#2ca02c', '#d62728', \n",
    "                 '#9467bd', '#8c564b', '#e377c2', '#7f7f7f', \n",
    "                 '#bcbd22', '#17becf', '#ff8da1']\n",
    "\n",
    "# Convert 'kmeans' to a categorical variable with explicit order from 0 to 10\n",
    "top70_UMAP_df = top70_UMAP_df.sort_values(by=['cluster'], ascending=True)\n",
    "top70_UMAP_df['clusters'] = pd.Categorical(top70_UMAP_df['cluster'].astype(int), categories=range(0, 11), ordered=True)\n",
    "\n",
    "\n",
    "# Plot UMAP with the kmeans clusters\n",
    "fig = px.scatter(top70_UMAP_df, \n",
    "                    x='UMAP1', \n",
    "                    y='UMAP2', \n",
    "                    labels={f'{column}_with_counts': column},\n",
    "                    color='clusters', \n",
    "                    title=f' HAMBI - UMAP with n_neighbors={n}, min_dist={m}', \n",
    "                    color_discrete_sequence=custom_colors,\n",
    "                    hover_data={'cluster': True, 'Domain to species': True, 'Domain to genus': True, 'element': True} )\n",
    "title = f' HAMBI - UMAP with n_neighbors={n}, min_dist={m} with kmeans clusters'\n",
    "\n",
    "# Update the layout of the figure\n",
    "fig.update_layout(\n",
    "    height=1300,\n",
    "    width=1200,\n",
    "    title_text=title,\n",
    "    showlegend=True,\n",
    "    legend=dict(\n",
    "        x=0.5,\n",
    "        y=-0.1,\n",
    "        traceorder=\"normal\",\n",
    "        xanchor='center',\n",
    "        yanchor='top',\n",
    "        orientation='h'\n",
    "    ),\n",
    "    template='simple_white',\n",
    "    xaxis=dict(\n",
    "        showgrid=True,\n",
    "        gridcolor='lightgray',\n",
    "        zeroline=False,\n",
    "        showline=True,  \n",
    "        linecolor='black', \n",
    "        linewidth=1,\n",
    "        mirror=True\n",
    "    ),\n",
    "    yaxis=dict(\n",
    "        showgrid=True,\n",
    "        gridcolor='lightgray',\n",
    "        zeroline=False,\n",
    "        showline=True,\n",
    "        linecolor='black',\n",
    "        linewidth=1,\n",
    "        mirror=True\n",
    "    )\n",
    ")\n",
    "\n",
    "# Show the figure\n",
    "fig.show()\n",
    "fig.write_image(f'UMAP_HAMBI_top_features/UMAP_{n}_{m}_kmeans.png')\n",
    "fig.write_html(f'UMAP_HAMBI_top_features/UMAP_{n}_{m}_kmeans.html')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "dd75a9c0-6130-4b62-8889-4028a7a76184",
   "metadata": {},
   "outputs": [],
   "source": [
    "species_cluster_count = top70_UMAP_df.groupby(['Domain to species', 'cluster']).size().reset_index(name='Count')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "cdfa14e8-f06d-40e4-9811-56941890854f",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "# Pivot the table to make species as rows and clusters as columns\n",
    "species_cluster_counts = top70_UMAP_df.groupby(['Domain to species', 'cluster']).size().reset_index(name='Count')\n",
    "\n",
    "# Pivot the table to make species as rows and clusters as columns\n",
    "species_cluster_pivot = species_cluster_counts.pivot(index='Domain to species', columns='cluster', values='Count').fillna(0)\n",
    "\n",
    "# Convert all columns to numeric to ensure proper summation\n",
    "species_cluster_pivot = species_cluster_pivot.apply(pd.to_numeric)\n",
    "\n",
    "# Calculate the row sum (total contigs per species) and add it as a new column\n",
    "species_cluster_pivot['SUM'] = species_cluster_pivot.sum(axis=1)\n",
    "species_cluster_pivot.sort_values(by='SUM', ascending=False, inplace =True)\n",
    "species_cluster_pivot"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "80e50fed-c37c-434a-a41a-2b13dd8538f0",
   "metadata": {},
   "source": [
    "### Save to file\n",
    "#### Top feature data ðŸ”"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c3142ad9-d70a-486d-a200-915647e4d66f",
   "metadata": {},
   "outputs": [],
   "source": [
    "species_cluster_pivot.to_csv('UMAP_HAMBI_top_features/HAMBI_species_cluster_pivot_top70_features.csv', index=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "85949862-b40d-4007-ba69-8bf062bb83db",
   "metadata": {},
   "outputs": [],
   "source": [
    "def calculate_shannon_entropy(pivot_table):\n",
    "    cluster_columns = pivot_table.columns[:-1]  \n",
    "    total_samples = pivot_table['SUM'].sum() \n",
    "\n",
    "    total_weighted_entropy = 0\n",
    "\n",
    "    for cluster in cluster_columns:\n",
    "        total_in_cluster = pivot_table[cluster].sum()\n",
    "        \n",
    "        if total_in_cluster == 0:\n",
    "            continue  \n",
    "        \n",
    "        # the proportion of each species in this cluster\n",
    "        proportions = pivot_table[cluster] / total_in_cluster\n",
    "        \n",
    "        # the entropy for this cluster\n",
    "        entropy = -np.sum(proportions * np.log2(proportions + 1e-10))  # Small value to avoid log(0)\n",
    "        \n",
    "        # Weight the entropy by the proportion of the cluster size relative to the total\n",
    "        weighted_entropy = (total_in_cluster / total_samples) * entropy\n",
    "        total_weighted_entropy += weighted_entropy\n",
    "\n",
    "        print(f\"Entropy for cluster {cluster}: {entropy:.4f}\")\n",
    "    \n",
    "    print(f\"Total Weighted Entropy: {total_weighted_entropy:.4f}\")\n",
    "    return total_weighted_entropy\n",
    "\n",
    "calculate_shannon_entropy(species_cluster_pivot)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6f9b50a7-87b4-4b2f-abe8-9187625b8c68",
   "metadata": {},
   "source": [
    "### Connecting clusters to taxa\n",
    "#### Top feature data ðŸ”"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0c5a9b41-1e6d-4421-941b-4e4eed7ab8a8",
   "metadata": {},
   "outputs": [],
   "source": [
    "def return_index(cluster:int, df:pd.DataFrame) -> tuple:\n",
    "    \"\"\"Returns the index of the cluster in the dataframe\n",
    "    \"\"\"\n",
    "    return top70_df[top70_df['cluster'] == cluster]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "dd9496ef-1e33-45f4-b01e-9acf4910bb41",
   "metadata": {},
   "outputs": [],
   "source": [
    "top70_UMAP_df['cluster'].value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "fc4a56da-d437-4121-b2fb-68e8fbd55e77",
   "metadata": {},
   "outputs": [],
   "source": [
    "cluster_data = {}\n",
    "for cluster in top70_UMAP_df['cluster'].unique():\n",
    "    cluster_data[cluster] = top70_UMAP_df[top70_UMAP_df['cluster'] == cluster].index.tolist()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "951ee3c4-9cd4-4d2a-b069-ddde08163a40",
   "metadata": {},
   "outputs": [],
   "source": [
    "cluster_data.keys()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5edf00f3-58cc-47d0-96aa-a98389259403",
   "metadata": {},
   "source": [
    "### Plot the number of each species in each cluster\n",
    "#### Top feature data ðŸ”"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8868b7de-046a-4ef9-931e-b0711fd8c0d4",
   "metadata": {},
   "outputs": [],
   "source": [
    "for cluster in cluster_data.keys():\n",
    "    # Select the subset of filt_UMAP_df for the current cluster\n",
    "    cluster_df = top70_UMAP_df.loc[cluster_data[cluster]]\n",
    "    \n",
    "    plt.figure(figsize=(20, 10))\n",
    "    sns.countplot(x='Domain to species', data=cluster_df, order=cluster_df['Domain to species'].value_counts().index)\n",
    "    plt.xticks(rotation=90)\n",
    "    plt.title(f'Cluster {cluster}')\n",
    "    plt.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d4950dda-275d-4fae-baf4-d68274dc8212",
   "metadata": {},
   "source": [
    "### Plot to how many clusters one species has been assigned by the clustering algoritms\n",
    "#### Top feature data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3b7c548e-1957-4ed0-af72-a735440e6db1",
   "metadata": {},
   "outputs": [],
   "source": [
    "species_kmeans_counts = top70_UMAP_df.groupby('Domain to species')['cluster'].nunique()\n",
    "\n",
    "species_kmeans_counts.plot(kind='bar', color='lightgreen')\n",
    "plt.title('KMEANS')\n",
    "plt.xlabel('Species')\n",
    "plt.ylabel('Number of Unique KMeans Clusters')\n",
    "\n",
    "plt.xticks(rotation=90)\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d567a0f3-a6a7-40f9-a13a-61de8c3885a6",
   "metadata": {},
   "source": [
    "### Extract clusters and redraw UMAP with subset of data\n",
    "#### Top feature data ðŸ”\n",
    "##### Cluster1: *Pseudomonas_E putida*"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d99cc844-dcfa-470f-9647-79aa3a315a54",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Extract based on UMAP1 & UMAP2 values:\n",
    "top70_UMAP_df_C1 = top70_UMAP_df.loc[(top70_UMAP_df['UMAP1']>= 10) & (top70_UMAP_df['UMAP1']<= 15)\n",
    "    & (top70_UMAP_df['UMAP2']>= -2) & (top70_UMAP_df['UMAP2']<= 0)]\n",
    "\n",
    "# Check\n",
    "top70_UMAP_df_C1.head()\n",
    "\n",
    "# Save those contig IDs to extract from the 'df_top70'\n",
    "top70_UMAP_df_C1.index.to_list()\n",
    "\n",
    "df_top70_C1 = df_top70[df_top70.index.isin(top70_UMAP_df_C1.index)]\n",
    "print(top70_UMAP_df_C1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4e46b398-ac7e-45eb-9e48-e037f4d1226f",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Save contig IDs\n",
    "top70_C1_contigs = df_top70_C1.index.to_list()\n",
    "\n",
    "directory = 'UMAP_HAMBI_top_features'\n",
    "file_path = os.path.join(directory, 'top70_C1_contigs.txt')\n",
    "\n",
    "with open(file_path, 'w') as file:\n",
    "    for item in top70_C1_contigs:\n",
    "        file.write(f\"{item}\\n\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2dcaa65d-9c2c-4057-9366-56e1f60b975a",
   "metadata": {},
   "source": [
    "### Extract clusters and redraw UMAP with subset of data\n",
    "#### Top feature data ðŸ”\n",
    "##### Cluster1A: *Pseudomonas_E putida*"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a2ca88ab-7c8a-4ae7-8337-dc1ef2395a52",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Extract based on UMAP1 & UMAP2 values:\n",
    "top70_UMAP_df_C1A = top70_UMAP_df.loc[(top70_UMAP_df['UMAP1']>= 12) & (top70_UMAP_df['UMAP1']<= 12.71)\n",
    "    & (top70_UMAP_df['UMAP2']>= -0.8510) & (top70_UMAP_df['UMAP2']<= -0.6000)]\n",
    "\n",
    "# Check\n",
    "top70_UMAP_df_C1A.head()\n",
    "\n",
    "# Save those contig IDs to extract from the 'df_top70'\n",
    "top70_UMAP_df_C1A.index.to_list()\n",
    "\n",
    "df_top70_C1A = df_top70[df_top70.index.isin(top70_UMAP_df_C1A.index)]\n",
    "print(top70_UMAP_df_C1A)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7c9e5bff-172d-4dc9-bc5c-344b92309bc6",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Save contig IDs\n",
    "top70_C1A_contigs = df_top70_C1A.index.to_list()\n",
    "\n",
    "directory = 'UMAP_HAMBI_top_features'\n",
    "file_path = os.path.join(directory, 'top70_C1A_contigs.txt')\n",
    "\n",
    "with open(file_path, 'w') as file:\n",
    "    for item in top70_C1A_contigs:\n",
    "        file.write(f\"{item}\\n\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "dd207376-fb2a-44d5-a52d-d3ed0ed6c42e",
   "metadata": {},
   "source": [
    "### Extract clusters and redraw UMAP with subset of data\n",
    "#### Top feature data ðŸ”\n",
    "##### Cluster1B: *Pseudomonas_E putida*"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "06540caa-e957-4891-becf-d6372f2404c3",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Extract based on UMAP1 & UMAP2 values:\n",
    "top70_UMAP_df_C1B = top70_UMAP_df.loc[(top70_UMAP_df['UMAP1']>= 12.5) & (top70_UMAP_df['UMAP1']<= 13)\n",
    "    & (top70_UMAP_df['UMAP2']>= -1.53) & (top70_UMAP_df['UMAP2']<= -1.12)]\n",
    "\n",
    "# Check\n",
    "top70_UMAP_df_C1B.head()\n",
    "\n",
    "# Save those contig IDs to extract from the 'df_top70'\n",
    "top70_UMAP_df_C1B.index.to_list()\n",
    "\n",
    "df_top70_C1B = df_top70[df_top70.index.isin(top70_UMAP_df_C1B.index)]\n",
    "print(top70_UMAP_df_C1B)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "78f3c46a-23af-4ccb-a4b6-3278756264bc",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Save contig IDs\n",
    "top70_C1B_contigs = df_top70_C1B.index.to_list()\n",
    "\n",
    "directory = 'UMAP_HAMBI_top_features'\n",
    "file_path = os.path.join(directory, 'top70_C1B_contigs.txt')\n",
    "\n",
    "with open(file_path, 'w') as file:\n",
    "    for item in top70_C1B_contigs:\n",
    "        file.write(f\"{item}\\n\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "68d09e4d-e305-45be-b9eb-36b5fa1d2c2d",
   "metadata": {},
   "source": [
    "### Extract clusters and redraw UMAP with subset of data\n",
    "#### Top feature data ðŸ”\n",
    "##### Cluster1C: *Pseudomonas_E putida*"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0b0204d5-dfba-4346-a5db-27f6073dfe8f",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Extract based on UMAP1 & UMAP2 values:\n",
    "top70_UMAP_df_C1C = top70_UMAP_df.loc[(top70_UMAP_df['UMAP1']>= 12.9) & (top70_UMAP_df['UMAP1']<= 13.31)\n",
    "    & (top70_UMAP_df['UMAP2']>= -1.124) & (top70_UMAP_df['UMAP2']<= -0.534)]\n",
    "\n",
    "# Check\n",
    "top70_UMAP_df_C1C.head()\n",
    "\n",
    "# Save those contig IDs to extract from the 'df_top70'\n",
    "top70_UMAP_df_C1C.index.to_list()\n",
    "\n",
    "df_top70_C1C = df_top70[df_top70.index.isin(top70_UMAP_df_C1C.index)]\n",
    "print(top70_UMAP_df_C1C)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f4466014-3daf-4a2a-8f06-a0678f4e82da",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Save contig IDs\n",
    "top70_C1C_contigs = df_top70_C1C.index.to_list()\n",
    "\n",
    "directory = 'UMAP_HAMBI_top_features'\n",
    "file_path = os.path.join(directory, 'top70_C1C_contigs.txt')\n",
    "\n",
    "with open(file_path, 'w') as file:\n",
    "    for item in top70_C1C_contigs:\n",
    "        file.write(f\"{item}\\n\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5709cdf5-3612-4dc6-9d95-405406aeebfa",
   "metadata": {},
   "source": [
    "### Extract clusters\n",
    "#### Top feature data ðŸ”\n",
    "##### Cluster2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "fc91401e-f65e-4767-af42-f8417fa3a76f",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Extract based on UMAP1 & UMAP2 values:\n",
    "top70_UMAP_df_C2 = top70_UMAP_df.loc[(top70_UMAP_df['UMAP1']>= 13) & (top70_UMAP_df['UMAP1']<= 15)\n",
    "    & (top70_UMAP_df['UMAP2']>= 6) & (top70_UMAP_df['UMAP2']<= 8)]\n",
    "\n",
    "# Check\n",
    "top70_UMAP_df_C2.head()\n",
    "\n",
    "# Save those contig IDs to extract from the 'df_top70'\n",
    "top70_UMAP_df_C2.index.to_list()\n",
    "\n",
    "df_top70_C2 = df_top70[df_top70.index.isin(top70_UMAP_df_C2.index)]\n",
    "print(top70_UMAP_df_C2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1de86101-8fbb-4542-9202-5f4b2514591d",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Save contig IDs\n",
    "top70_C2_contigs = df_top70_C2.index.to_list()\n",
    "\n",
    "directory = 'UMAP_HAMBI_top_features'\n",
    "file_path = os.path.join(directory, 'top70_C2_contigs.txt')\n",
    "\n",
    "with open(file_path, 'w') as file:\n",
    "    for item in top70_C2_contigs:\n",
    "        file.write(f\"{item}\\n\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "684ce0a8-b26a-4e42-ba69-1adb21a98fd5",
   "metadata": {},
   "source": [
    "### Extract clusters\n",
    "#### Top feature data ðŸ”\n",
    "##### Cluster3"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "734bacfe-3f87-4077-8b2b-74466402bdc6",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Extract based on UMAP1 & UMAP2 values:\n",
    "top70_UMAP_df_C3 = top70_UMAP_df.loc[(top70_UMAP_df['UMAP1']>= 5.10) & (top70_UMAP_df['UMAP1']<= 6.70)\n",
    "    & (top70_UMAP_df['UMAP2']>= 8.30) & (top70_UMAP_df['UMAP2']<= 10.75)]\n",
    "\n",
    "# Check\n",
    "top70_UMAP_df_C3.head()\n",
    "\n",
    "# Save those contig IDs to extract from the 'df_top70'\n",
    "top70_UMAP_df_C3.index.to_list()\n",
    "\n",
    "df_top70_C3 = df_top70[df_top70.index.isin(top70_UMAP_df_C3.index)]\n",
    "print(top70_UMAP_df_C3)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d358380c-5503-4e66-8847-9d22474bb0d5",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Save contig IDs\n",
    "top70_C3_contigs = df_top70_C3.index.to_list()\n",
    "\n",
    "directory = 'UMAP_HAMBI_top_features'\n",
    "file_path = os.path.join(directory, 'top70_C3_contigs.txt')\n",
    "\n",
    "with open(file_path, 'w') as file:\n",
    "    for item in top70_C3_contigs:\n",
    "        file.write(f\"{item}\\n\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "693fbf2c-f460-4698-93a4-59dfa87a964f",
   "metadata": {},
   "source": [
    "### Redraw subste of data\n",
    "#### Top feature data ðŸ”\n",
    "##### Cluster3"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "cb69ac08-a696-4527-a5c4-5a0ab42e79f3",
   "metadata": {},
   "outputs": [],
   "source": [
    "n_neighbors = [30]\n",
    "min_dist = [0.1]\n",
    "color_dict = [\n",
    "    \"#FF0000\", \"#00FF00\", \"#0000FF\", \"#FFFF00\", \"#FF00FF\", \"#00FFFF\", \"#800000\", \"#008000\",\n",
    "    \"#000080\", \"#808000\", \"#800080\", \"#008080\", \"#FF4500\", \"#2E8B57\", \"#4682B4\", \"#DAA520\",\n",
    "    \"#FF1493\", \"#9ACD32\", \"#8A2BE2\", \"#5F9EA0\", \"#D2691E\", \"#DC143C\", \"#B8860B\", \"#32CD32\",\n",
    "    \"#FFD700\", \"#20B2AA\", \"#FF6347\", \"#4169E1\", \"#8B4513\", \"#6A5ACD\", \"#FF7F50\", \"#2F4F4F\"\n",
    "]\n",
    "\n",
    "\n",
    "custom_colors = color_dict\n",
    "for n in n_neighbors:\n",
    "    for m in min_dist:\n",
    "        reducer = umap.UMAP(n_neighbors=n, min_dist=m, random_state=seed)\n",
    "        embedding = reducer.fit_transform(df_top70_C3.iloc[:, :-17])\n",
    "        top70_UMAP_df_C3  = pd.DataFrame({\n",
    "            'UMAP1': embedding[:, 0],\n",
    "            'UMAP2': embedding[:, 1],\n",
    "            'Domain to species': df_top70_C3['Domain to species'],\n",
    "            'Domain to genus': df_top70_C3['Domain to genus'],\n",
    "            'element': df_top70_C3['element'],\n",
    "            'contig': df_top70_C3.index,\n",
    "        })\n",
    "        # clustering k defined choosing the k by looking at the UMAP \n",
    "        kmeans = KMeans(n_clusters=11, random_state=seed)\n",
    "        \n",
    "        k_labels = kmeans.fit_predict(embedding)\n",
    "        \n",
    "        top70_UMAP_df_C3['cluster'] = k_labels\n",
    "        \n",
    "        column = 'Domain to species'\n",
    "        top70_UMAP_df_C3['Domain to species'] = top70_UMAP_df_C3['Domain to species'].fillna('NA')\n",
    "\n",
    "        value_counts = top70_UMAP_df[column].value_counts().to_dict()\n",
    "\n",
    "        top70_UMAP_df_C3[f'{column}_with_counts'] = top70_UMAP_df_C3[column].apply(lambda x: f'{x} ({value_counts[x]})')\n",
    "\n",
    "        top70_UMAP_df_C3['count'] = top70_UMAP_df_C3[column].map(value_counts)\n",
    "        top70_UMAP_df_C3 = top70_UMAP_df_C3.sort_values(by=['count'], ascending=False)\n",
    "\n",
    "        top70_UMAP_df_C3[f'{column}_with_counts'] = pd.Categorical(\n",
    "            top70_UMAP_df_C3[f'{column}_with_counts'],\n",
    "            categories=top70_UMAP_df_C3[f'{column}_with_counts'].unique(),\n",
    "            ordered=True\n",
    "        )\n",
    "        \n",
    "        fig = px.scatter(top70_UMAP_df_C3, \n",
    "                            x='UMAP1', \n",
    "                            y='UMAP2', \n",
    "                            labels={f'{column}_with_counts': column},\n",
    "                            color=f'{column}_with_counts', \n",
    "                            title=f' HAMBI - UMAP with n_neighbors={n}, min_dist={m}', \n",
    "                            color_discrete_sequence=custom_colors,\n",
    "                            hover_data={'cluster': True, 'Domain to species': True, 'Domain to genus': True, 'element': True, 'contig': True} )\n",
    "        title = f' HAMBI - UMAP with n_neighbors={n}, min_dist={m}'\n",
    "        fig.update_traces(marker=dict(size=10))\n",
    "        fig.update_layout(\n",
    "            height=1700,\n",
    "            width=1200,\n",
    "            title_text=title,\n",
    "            showlegend=True,\n",
    "            legend=dict(\n",
    "                x=0.5,\n",
    "                y=-0.1,\n",
    "                traceorder=\"normal\",\n",
    "                xanchor='center',\n",
    "                yanchor='top',\n",
    "                orientation='h'\n",
    "            ),\n",
    "            template='simple_white',\n",
    "            xaxis=dict(\n",
    "                showgrid=True,\n",
    "                gridcolor='lightgray',\n",
    "                zeroline=False,\n",
    "                showline=True,  \n",
    "                linecolor='black', \n",
    "                linewidth=1,\n",
    "                mirror=True\n",
    "            ),\n",
    "            yaxis=dict(\n",
    "                showgrid=True,\n",
    "                gridcolor='lightgray',\n",
    "                zeroline=False,\n",
    "                showline=True,\n",
    "                linecolor='black',\n",
    "                linewidth=1,\n",
    "                mirror=True\n",
    "            )\n",
    "        )\n",
    "        fig.show()\n",
    "        fig.write_image(f'UMAP_HAMBI_top_features/UMAP_{n}_{m}_C3.png')\n",
    "        fig.write_html(f'UMAP_HAMBI_top_features/UMAP_{n}_{m}_C3.html')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c590353c-26e5-41d4-9ec7-e5a836ff7ee6",
   "metadata": {
    "scrolled": true
   },
   "source": [
    "#### Check the numbers of plasmids in each datasets"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9c46fbfe-0713-4c6e-931a-12cc825c3a10",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "UMAP_df[UMAP_df['element'].str.contains('plas', case=False, na=False)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3670b673-12a8-4b33-bd27-d08d389418ab",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "filt_UMAP_df[filt_UMAP_df['element'].str.contains('plas', case=False, na=False)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7de31205-1821-4abd-84b2-6bbbe9c7da5f",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "top70_UMAP_df[top70_UMAP_df['element'].str.contains('plas', case=False, na=False)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a7335a32-7e7b-40f4-9aec-c8e603ee873f",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "top70_UMAP_df_C1[top70_UMAP_df_C1['element'].str.contains('plas', case=False, na=False)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f5db2e52-fc77-4d63-ab08-6edc640932ca",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "top70_UMAP_df_C2[top70_UMAP_df_C2['element'].str.contains('plas', case=False, na=False)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "05c92a56-792e-4174-bbca-a6632b2c1217",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "top70_UMAP_df_C3[top70_UMAP_df_C3['element'].str.contains('plas', case=False, na=False)]"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.14"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
